--- a/net/minecraft/data/tags/DamageTypeTagsProvider.java
+++ b/net/minecraft/data/tags/DamageTypeTagsProvider.java
@@ -13,6 +_,10 @@
         super(output, Registries.DAMAGE_TYPE, lookupProvider);
     }
 
+    public DamageTypeTagsProvider(final PackOutput output, final CompletableFuture<HolderLookup.Provider> lookupProvider, String modId, @org.jetbrains.annotations.Nullable net.minecraftforge.common.data.ExistingFileHelper existingFileHelper) {
+        super(output, Registries.DAMAGE_TYPE, lookupProvider, modId, existingFileHelper);
+    }
+
     @Override
     protected void addTags(final HolderLookup.Provider registries) {
         this.tag(DamageTypeTags.DAMAGES_HELMET).add(DamageTypes.FALLING_ANVIL, DamageTypes.FALLING_BLOCK, DamageTypes.FALLING_STALACTITE);
