--- a/net/minecraft/data/tags/FlatLevelGeneratorPresetTagsProvider.java
+++ b/net/minecraft/data/tags/FlatLevelGeneratorPresetTagsProvider.java
@@ -13,6 +_,10 @@
         super(output, Registries.FLAT_LEVEL_GENERATOR_PRESET, lookupProvider);
     }
 
+    public FlatLevelGeneratorPresetTagsProvider(final PackOutput output, final CompletableFuture<HolderLookup.Provider> lookupProvider, String modId, @org.jetbrains.annotations.Nullable net.minecraftforge.common.data.ExistingFileHelper existingFileHelper) {
+        super(output, Registries.FLAT_LEVEL_GENERATOR_PRESET, lookupProvider, modId, existingFileHelper);
+    }
+
     @Override
     protected void addTags(final HolderLookup.Provider registries) {
         this.tag(FlatLevelGeneratorPresetTags.VISIBLE)
