--- a/net/minecraft/data/tags/InstrumentTagsProvider.java
+++ b/net/minecraft/data/tags/InstrumentTagsProvider.java
@@ -13,6 +_,10 @@
         super(output, Registries.INSTRUMENT, lookupProvider);
     }
 
+    public InstrumentTagsProvider(final PackOutput output, final CompletableFuture<HolderLookup.Provider> lookupProvider, String modId, @org.jetbrains.annotations.Nullable net.minecraftforge.common.data.ExistingFileHelper existingFileHelper) {
+        super(output, Registries.INSTRUMENT, lookupProvider, modId, existingFileHelper);
+    }
+
     @Override
     protected void addTags(final HolderLookup.Provider registries) {
         this.tag(InstrumentTags.REGULAR_GOAT_HORNS)
