--- a/net/minecraft/data/tags/IntrinsicHolderTagsProvider.java
+++ b/net/minecraft/data/tags/IntrinsicHolderTagsProvider.java
@@ -26,6 +_,18 @@
         final PackOutput output,
         final ResourceKey<? extends Registry<T>> registryKey,
         final CompletableFuture<HolderLookup.Provider> lookupProvider,
+        final Function<T, ResourceKey<T>> keyExtractor,
+        final String modid,
+        final @org.jetbrains.annotations.Nullable net.minecraftforge.common.data.ExistingFileHelper existingFileHelper
+    ) {
+        super(output, registryKey, lookupProvider, modid, existingFileHelper);
+        this.keyExtractor = keyExtractor;
+    }
+
+    public IntrinsicHolderTagsProvider(
+        final PackOutput output,
+        final ResourceKey<? extends Registry<T>> registryKey,
+        final CompletableFuture<HolderLookup.Provider> lookupProvider,
         final CompletableFuture<TagsProvider.TagLookup<T>> parentProvider,
         final Function<T, ResourceKey<T>> keyExtractor
     ) {
@@ -33,8 +_,21 @@
         this.keyExtractor = keyExtractor;
     }
 
+    public IntrinsicHolderTagsProvider(
+        final PackOutput output,
+        final ResourceKey<? extends Registry<T>> registryKey,
+        final CompletableFuture<HolderLookup.Provider> lookupProvider,
+        final CompletableFuture<TagsProvider.TagLookup<T>> parentProvider,
+        final Function<T, ResourceKey<T>> keyExtractor,
+        final String modid,
+        final @org.jetbrains.annotations.Nullable net.minecraftforge.common.data.ExistingFileHelper existingFileHelper
+    ) {
+        super(output, registryKey, lookupProvider, parentProvider, modid, existingFileHelper);
+        this.keyExtractor = keyExtractor;
+    }
+
     protected TagAppender<T, T> tag(final TagKey<T> tag) {
         TagBuilder builder = this.getOrCreateRawBuilder(tag);
-        return TagAppender.<T>forBuilder(builder).map(this.keyExtractor);
+        return TagAppender.<T>forBuilder(builder, this.modId).map(this.keyExtractor);
     }
 }
