--- a/net/minecraft/data/tags/KeyTagProvider.java
+++ b/net/minecraft/data/tags/KeyTagProvider.java
@@ -20,9 +_,13 @@
         return TagAppender.forBuilder(builder);
     }
 
+    protected KeyTagProvider(final PackOutput output, final ResourceKey<? extends Registry<T>> registryKey, final CompletableFuture<HolderLookup.Provider> lookupProvider, String modId, @org.jetbrains.annotations.Nullable net.minecraftforge.common.data.ExistingFileHelper existingFileHelper) {
+        super(output, registryKey, lookupProvider, modId, existingFileHelper);
+    }
+
     protected TagAppender<ResourceKey<T>, T> tag(final TagKey<T> tag, final boolean replace) {
         TagBuilder builder = this.getOrCreateRawBuilder(tag);
         builder.setReplace(replace);
-        return TagAppender.forBuilder(builder);
+        return TagAppender.forBuilder(builder, this.modId);
     }
 }
