--- a/net/minecraft/data/tags/TagAppender.java
+++ b/net/minecraft/data/tags/TagAppender.java
@@ -9,7 +_,7 @@
 import net.minecraft.tags.TagBuilder;
 import net.minecraft.tags.TagKey;
 
-public interface TagAppender<E, T> {
+public interface TagAppender<E, T> extends net.minecraftforge.common.extensions.IForgeTagAppender<E, T> {
     TagAppender<E, T> add(E element);
 
     default TagAppender<E, T> add(final E... elements) {
@@ -33,6 +_,10 @@
     TagAppender<E, T> addOptionalTag(TagKey<T> tag);
 
     static <T> TagAppender<ResourceKey<T>, T> forBuilder(final TagBuilder builder) {
+        return forBuilder(builder, "unknown");
+    }
+
+    static <T> TagAppender<ResourceKey<T>, T> forBuilder(final TagBuilder builder, String source) {
         return new TagAppender<ResourceKey<T>, T>() {
             public TagAppender<ResourceKey<T>, T> add(final ResourceKey<T> element) {
                 builder.addElement(element.identifier());
@@ -55,6 +_,21 @@
                 builder.addOptionalTag(tag.location());
                 return this;
             }
+
+            @Override
+            public TagBuilder getInternalBuilder() {
+                return builder;
+            }
+
+            @Override
+            public String getSourceName() {
+                return source;
+            }
+
+            @Override
+            public TagAppender<ResourceKey<T>, T> remove(ResourceKey<T> value) {
+                return this.remove(value.identifier());
+            }
         };
     }
 
@@ -86,6 +_,22 @@
             @Override
             public TagAppender<U, T> addOptionalTag(final TagKey<T> tag) {
                 original.addOptionalTag(tag);
+                return this;
+            }
+
+            @Override
+            public TagBuilder getInternalBuilder() {
+                return original.getInternalBuilder();
+            }
+
+            @Override
+            public String getSourceName() {
+                return original.getSourceName();
+            }
+
+            @Override
+            public TagAppender<U, T> remove(U value) {
+                original.remove(converter.apply(value));
                 return this;
             }
         };
