--- a/net/minecraft/data/tags/VanillaBlockTagsProvider.java
+++ b/net/minecraft/data/tags/VanillaBlockTagsProvider.java
@@ -17,6 +_,10 @@
         super(output, Registries.BLOCK, lookupProvider, e -> e.builtInRegistryHolder().key());
     }
 
+    public VanillaBlockTagsProvider(final PackOutput output, final CompletableFuture<HolderLookup.Provider> lookupProvider, String modId, @org.jetbrains.annotations.Nullable net.minecraftforge.common.data.ExistingFileHelper existingFileHelper) {
+        super(output, Registries.BLOCK, lookupProvider, e -> e.builtInRegistryHolder().key(), modId, existingFileHelper);
+    }
+
     @Override
     protected void addTags(final HolderLookup.Provider registries) {
         (new BlockItemTagsProvider() {
