--- a/net/minecraft/data/tags/VanillaEnchantmentTagsProvider.java
+++ b/net/minecraft/data/tags/VanillaEnchantmentTagsProvider.java
@@ -11,6 +_,10 @@
         super(output, lookupProvider);
     }
 
+    public VanillaEnchantmentTagsProvider(final PackOutput output, final CompletableFuture<HolderLookup.Provider> lookupProvider, String modId, @org.jetbrains.annotations.Nullable net.minecraftforge.common.data.ExistingFileHelper existingFileHelper) {
+        super(output, lookupProvider, modId, existingFileHelper);
+    }
+
     @Override
     protected void addTags(final HolderLookup.Provider registries) {
         this.tooltipOrder(
