--- a/net/minecraft/data/tags/WorldPresetTagsProvider.java
+++ b/net/minecraft/data/tags/WorldPresetTagsProvider.java
@@ -13,6 +_,10 @@
         super(output, Registries.WORLD_PRESET, lookupProvider);
     }
 
+    public WorldPresetTagsProvider(final PackOutput output, final CompletableFuture<HolderLookup.Provider> lookupProvider, String modId, @org.jetbrains.annotations.Nullable net.minecraftforge.common.data.ExistingFileHelper existingFileHelper) {
+       super(output, Registries.WORLD_PRESET, lookupProvider, modId, existingFileHelper);
+    }
+
     @Override
     protected void addTags(final HolderLookup.Provider registries) {
         this.tag(WorldPresetTags.NORMAL)
