--- a/net/minecraft/client/renderer/debug/EntityHitboxDebugRenderer.java
+++ b/net/minecraft/client/renderer/debug/EntityHitboxDebugRenderer.java
@@ -100,8 +_,8 @@
             );
         }
 
-        if (p_450643_ instanceof EnderDragon enderdragon) {
-            for (EnderDragonPart enderdragonpart : enderdragon.getSubEntities()) {
+        if (p_450643_.isMultipartEntity()) {
+            for (var enderdragonpart : p_450643_.getParts()) {
                 Vec3 vec34 = enderdragonpart.position();
                 Vec3 vec35 = enderdragonpart.getPosition(p_456396_);
                 Vec3 vec36 = vec35.subtract(vec34);
