--- a/net/minecraft/world/level/material/FlowingFluid.java
+++ b/net/minecraft/world/level/material/FlowingFluid.java
@@ -172,7 +_,7 @@
             BlockState blockstate = level.getBlockState(blockpos);
             FluidState fluidstate = blockstate.getFluidState();
             if (fluidstate.getType().isSame(this) && canPassThroughWall(direction, level, pos, state, blockpos, blockstate)) {
-                if (fluidstate.isSource()) {
+                if (fluidstate.isSource() && net.minecraftforge.event.ForgeEventFactory.canCreateFluidSource(level, blockpos, blockstate, fluidstate.canConvertToSource(level, blockpos))) {
                     j++;
                 }
 
@@ -180,7 +_,7 @@
             }
         }
 
-        if (j >= 2 && this.canConvertToSource(level)) {
+        if (j >= 2) {
             BlockState blockstate1 = level.getBlockState(blockpos$mutableblockpos.setWithOffset(pos, Direction.DOWN));
             FluidState fluidstate1 = blockstate1.getFluidState();
             if (blockstate1.isSolid() || this.isSourceBlockOfThisType(fluidstate1)) {
@@ -265,6 +_,15 @@
         return this.getSource().defaultFluidState().setValue(FALLING, falling);
     }
 
+    @Override
+    public boolean canConvertToSource(FluidState state, ServerLevel level, BlockPos pos) {
+       return this.canConvertToSource(level);
+    }
+
+    /**
+     * @deprecated Forge: Use {@link #canConvertToSource(FluidState, Level, BlockPos)} instead.
+     */
+    @Deprecated
     protected abstract boolean canConvertToSource(ServerLevel level);
 
     protected void spreadTo(final LevelAccessor level, final BlockPos pos, final BlockState state, final Direction direction, final FluidState target) {
