--- a/net/minecraft/world/level/material/LavaFluid.java
+++ b/net/minecraft/world/level/material/LavaFluid.java
@@ -93,7 +_,7 @@
                     BlockState blockstate = level.getBlockState(blockpos);
                     if (blockstate.isAir()) {
                         if (this.hasFlammableNeighbours(level, blockpos)) {
-                            level.setBlockAndUpdate(blockpos, BaseFireBlock.getState(level, blockpos));
+                            level.setBlockAndUpdate(blockpos, net.minecraftforge.event.ForgeEventFactory.fireFluidPlaceBlockEvent(level, blockpos, pos, Blocks.FIRE.defaultBlockState()));
                             return;
                         }
                     } else if (blockstate.blocksMotion()) {
@@ -107,8 +_,8 @@
                         return;
                     }
 
-                    if (level.isEmptyBlock(blockpos1.above()) && this.isFlammable(level, blockpos1)) {
-                        level.setBlockAndUpdate(blockpos1.above(), BaseFireBlock.getState(level, blockpos1));
+                    if (level.isEmptyBlock(blockpos1.above()) && this.isFlammable(level, blockpos1, Direction.UP)) {
+                        level.setBlockAndUpdate(blockpos1.above(), net.minecraftforge.event.ForgeEventFactory.fireFluidPlaceBlockEvent(level, blockpos1.above(), pos, Blocks.FIRE.defaultBlockState()));
                     }
                 }
             }
@@ -124,7 +_,7 @@
 
     private boolean hasFlammableNeighbours(final LevelReader level, final BlockPos pos) {
         for (Direction direction : Direction.values()) {
-            if (this.isFlammable(level, pos.relative(direction))) {
+            if (this.isFlammable(level, pos.relative(direction), direction.getOpposite())) {
                 return true;
             }
         }
@@ -132,10 +_,17 @@
         return false;
     }
 
+    /** @deprecated Forge: use {@link LavaFluid#isFlammable(LevelReader,BlockPos,Direction)} instead */
     private boolean isFlammable(final LevelReader level, final BlockPos pos) {
         return level.isInsideBuildHeight(pos.getY()) && !level.hasChunkAt(pos) ? false : level.getBlockState(pos).ignitedByLava();
     }
 
+    private boolean isFlammable(final LevelReader level, final BlockPos pos, Direction face) {
+        return level.isInsideBuildHeight(pos.getY()) && !level.hasChunkAt(pos)
+            ? false
+            : level.getBlockState(pos).ignitedByLava() && level.getBlockState(pos).isFlammable(level, pos, face);
+     }
+
     @Override
     public @Nullable ParticleOptions getDripParticle() {
         return ParticleTypes.DRIPPING_LAVA;
@@ -206,7 +_,8 @@
             FluidState fluidstate = level.getFluidState(pos);
             if (this.is(FluidTags.LAVA) && fluidstate.is(FluidTags.WATER)) {
                 if (state.getBlock() instanceof LiquidBlock) {
-                    level.setBlock(pos, Blocks.STONE.defaultBlockState(), 3);
+                    level.setBlock(pos, net.minecraftforge.event.ForgeEventFactory.fireFluidPlaceBlockEvent(level, pos, pos, Blocks.STONE.defaultBlockState()), 3);
+
                 }
 
                 this.fizz(level, pos);
