--- a/net/minecraft/server/network/EventLoopGroupHolder.java
+++ b/net/minecraft/server/network/EventLoopGroupHolder.java
@@ -52,6 +_,7 @@
     private final Class<? extends Channel> channelCls;
     private final Class<? extends ServerChannel> serverChannelCls;
     private volatile @Nullable EventLoopGroup group;
+    private volatile @Nullable EventLoopGroup groupClient;
 
     public static EventLoopGroupHolder remote(final boolean allowNativeTransport) {
         if (allowNativeTransport) {
@@ -77,24 +_,31 @@
         this.serverChannelCls = serverChannelCls;
     }
 
-    private ThreadFactory createThreadFactory() {
-        return new ThreadFactoryBuilder().setNameFormat("Netty " + this.type + " IO #%d").setDaemon(true).build();
+    private ThreadFactory createThreadFactory(boolean client) {
+        return new ThreadFactoryBuilder().setNameFormat("Netty " + this.type + (client ? " Client" : " Server") + " IO #%d").setDaemon(true).setThreadFactory(net.minecraftforge.fml.util.thread.SidedThreadGroups.get(client)).build();
     }
 
     protected abstract IoHandlerFactory ioHandlerFactory();
 
-    private EventLoopGroup createEventLoopGroup() {
-        return new MultiThreadIoEventLoopGroup(this.createThreadFactory(), this.ioHandlerFactory());
+    private EventLoopGroup createEventLoopGroup(boolean client) {
+        return new MultiThreadIoEventLoopGroup(this.createThreadFactory(client), this.ioHandlerFactory());
     }
 
     public EventLoopGroup eventLoopGroup() {
-        EventLoopGroup result = this.group;
+        return eventLoopGroup(false);
+    }
+
+    public EventLoopGroup eventLoopGroup(boolean client) {
+        EventLoopGroup result = client ? this.groupClient : this.group;
         if (result == null) {
             synchronized (this) {
-                result = this.group;
+                result = client ? this.groupClient : this.group;
                 if (result == null) {
-                    result = this.createEventLoopGroup();
-                    this.group = result;
+                    result = this.createEventLoopGroup(client);
+                    if (client)
+                        this.groupClient = result;
+                    else
+                        this.group = result;
                 }
             }
         }
