--- a/net/minecraft/server/network/EventLoopGroupHolder.java
+++ b/net/minecraft/server/network/EventLoopGroupHolder.java
@@ -48,6 +_,12 @@
             return LocalIoHandler.newFactory();
         }
     };
+    private static final EventLoopGroupHolder LOCAL_CLIENT = new EventLoopGroupHolder("Local Client", LocalChannel.class, LocalServerChannel.class) {
+        @Override
+        protected IoHandlerFactory ioHandlerFactory() {
+            return LocalIoHandler.newFactory();
+        }
+    };
     private final String type;
     private final Class<? extends Channel> channelCls;
     private final Class<? extends ServerChannel> serverChannelCls;
@@ -71,6 +_,10 @@
         return LOCAL;
     }
 
+    public static EventLoopGroupHolder localClient() {
+        return LOCAL_CLIENT;
+    }
+
     EventLoopGroupHolder(String p_451976_, Class<? extends Channel> p_460581_, Class<? extends ServerChannel> p_455144_) {
         this.type = p_451976_;
         this.channelCls = p_460581_;
@@ -78,7 +_,7 @@
     }
 
     private ThreadFactory createThreadFactory() {
-        return new ThreadFactoryBuilder().setNameFormat("Netty " + this.type + " IO #%d").setDaemon(true).build();
+        return new ThreadFactoryBuilder().setNameFormat("Netty " + this.type + " IO #%d").setDaemon(true).setThreadFactory(net.minecraftforge.fml.util.thread.SidedThreadGroups.get(this == LOCAL_CLIENT)).build();
     }
 
     protected abstract IoHandlerFactory ioHandlerFactory();
