--- a/net/minecraft/server/network/ServerConnectionListener.java
+++ b/net/minecraft/server/network/ServerConnectionListener.java
@@ -38,6 +_,7 @@
 import org.slf4j.Logger;
 
 public class ServerConnectionListener {
+    private static final int READ_TIMEOUT = Integer.parseInt(System.getProperty("forge.readTimeout", "30"));
     private static final Logger LOGGER = LogUtils.getLogger();
     private final MinecraftServer server;
     public volatile boolean running;
@@ -50,6 +_,8 @@
     }
 
     public void startTcpServerListener(final @Nullable InetAddress address, final int port) throws IOException {
+        var fixedAddress = address != null ? address : new java.net.InetSocketAddress(port).getAddress();
+        net.minecraftforge.network.DualStackUtils.checkIPv6(fixedAddress);
         synchronized (this.channels) {
             EventLoopGroupHolder eventloopgroupholder = EventLoopGroupHolder.remote(this.server.useNativeTransport());
             this.channels.add(new ServerBootstrap().channel(eventloopgroupholder.serverChannelCls()).childHandler(new ChannelInitializer<Channel>() {
@@ -64,7 +_,7 @@
                     } catch (ChannelException channelexception) {
                     }
 
-                    ChannelPipeline channelpipeline = channel.pipeline().addLast("timeout", new ReadTimeoutHandler(30));
+                    ChannelPipeline channelpipeline = channel.pipeline().addLast("timeout", new ReadTimeoutHandler(READ_TIMEOUT));
                     if (ServerConnectionListener.this.server.repliesToStatus()) {
                         channelpipeline.addLast("legacy_query", new LegacyQueryHandler(ServerConnectionListener.this.getServer()));
                     }
@@ -76,7 +_,7 @@
                     connection.configurePacketHandler(channelpipeline);
                     connection.setListenerForServerboundHandshake(new ServerHandshakePacketListenerImpl(ServerConnectionListener.this.server, connection));
                 }
-            }).group(eventloopgroupholder.eventLoopGroup()).localAddress(address, port).bind().syncUninterruptibly());
+            }).group(eventloopgroupholder.eventLoopGroup()).localAddress(fixedAddress, port).bind().syncUninterruptibly());
         }
     }
 
