--- a/net/minecraft/util/SpawnUtil.java
+++ b/net/minecraft/util/SpawnUtil.java
@@ -46,7 +_,7 @@
                 )) {
                 T t = (T)entityType.create(level, null, blockpos$mutableblockpos, spawnReason, false, false);
                 if (t != null) {
-                    if (t.checkSpawnRules(level, spawnReason) && t.checkSpawnObstruction(level)) {
+                    if (net.minecraftforge.event.ForgeEventFactory.checkSpawnPosition(t, level, spawnReason)) {
                         level.addFreshEntityWithPassengers(t);
                         t.playAmbientSound();
                         return Optional.of(t);
