--- a/net/minecraft/world/item/enchantment/effects/ReplaceDisk.java
+++ b/net/minecraft/world/item/enchantment/effects/ReplaceDisk.java
@@ -48,6 +_,7 @@
         for (BlockPos blockpos1 : BlockPos.betweenClosed(blockpos.offset(-i, 0, -i), blockpos.offset(i, Math.min(j - 1, 0), i))) {
             if (blockpos1.distToCenterSqr(position.x(), blockpos1.getY() + 0.5, position.z()) < Mth.square(i)
                 && this.predicate.map(p -> p.test(serverLevel, blockpos1)).orElse(true)
+                && !net.minecraftforge.event.ForgeEventFactory.onBlockPlace(entity, net.minecraftforge.common.util.BlockSnapshot.create(serverLevel.dimension(), serverLevel, blockpos), net.minecraft.core.Direction.UP)
                 && serverLevel.setBlockAndUpdate(blockpos1, this.blockState.getState(serverLevel, randomsource, blockpos1))) {
                 this.triggerGameEvent.ifPresent(event -> serverLevel.gameEvent(entity, (Holder<GameEvent>)event, blockpos1));
             }
