--- a/net/minecraft/world/level/entity/PersistentEntitySectionManager.java
+++ b/net/minecraft/world/level/entity/PersistentEntitySectionManager.java
@@ -72,7 +_,18 @@
         return this.addEntity(entity, false);
     }
 
+    public boolean addNewEntityWithoutEvent(T entity) {
+       return this.addEntityWithoutEvent(entity, false);
+    }
+
     private boolean addEntity(final T entity, final boolean loaded) {
+        if (entity instanceof Entity && net.minecraftforge.event.ForgeEventFactory.onEntityJoinLevel((Entity)entity, ((Entity)entity).level(), loaded)) {
+            return false;
+        }
+        return addEntityWithoutEvent(entity, loaded);
+    }
+
+    private boolean addEntityWithoutEvent(final T entity, final boolean loaded) {
         if (!this.addEntityUuid(entity)) {
             return false;
         } else {
@@ -93,6 +_,10 @@
                 this.startTicking(entity);
             }
 
+            if (entity instanceof Entity) {
+                ((Entity)entity).onAddedToLevel();
+            }
+
             return true;
         }
     }
@@ -370,6 +_,7 @@
 
     private class Callback implements EntityInLevelCallback {
         private final T entity;
+        private final Entity realEntity;
         private long currentSectionKey;
         private EntitySection<T> currentSection;
 
@@ -377,6 +_,7 @@
             Objects.requireNonNull(PersistentEntitySectionManager.this);
             super();
             this.entity = entity;
+            this.realEntity = entity instanceof Entity e ? e : null;
             this.currentSectionKey = currentSectionKey;
             this.currentSection = currentSection;
         }
@@ -395,9 +_,13 @@
                 PersistentEntitySectionManager.this.removeSectionIfEmpty(this.currentSectionKey, this.currentSection);
                 EntitySection<T> newSection = PersistentEntitySectionManager.this.sectionStorage.getOrCreateSection(newSectionPos);
                 newSection.add(this.entity);
+                long oldSectionKey = currentSectionKey;
                 this.currentSection = newSection;
                 this.currentSectionKey = newSectionPos;
                 this.updateStatus(previousStatus, newSection.getStatus());
+                if (this.realEntity != null) {
+                    net.minecraftforge.event.ForgeEventFactory.onEntityEnterSection(this.realEntity, oldSectionKey, newSectionPos);
+                }
             }
         }
 
