--- a/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
@@ -64,6 +_,7 @@
     private static final int DEFAULT_LIT_TIME_REMAINING = 0;
     private static final int DEFAULT_LIT_TOTAL_TIME = 0;
     private static final float DEFAULT_SPEED_MULTIPLIER = 1.0F;
+    private final RecipeType<? extends AbstractCookingRecipe> recipeType;
     protected NonNullList<ItemStack> items = NonNullList.withSize(3, ItemStack.EMPTY);
     private int litTimeRemaining;
     private int litTotalTime;
@@ -117,6 +_,7 @@
     ) {
         super(type, worldPosition, blockState);
         this.quickCheck = RecipeManager.createCheck(recipeType);
+        this.recipeType = recipeType;
     }
 
     @Override
@@ -169,7 +_,7 @@
                 if (recipe != null) {
                     int maxStackSize = entity.getMaxStackSize();
                     ItemStack burnResult = recipe.value().assemble(input);
-                    if (!burnResult.isEmpty() && canBurn(entity.items, maxStackSize, burnResult)) {
+                    if (!burnResult.isEmpty() && entity.canBurn(entity.items, maxStackSize, burnResult)) {
                         if (!isLit) {
                             int newLitTime = entity.getBurnDuration(level, fuel);
                             float newSpeedMultiplier = entity.getSpeedMultiplier(level, fuel);
@@ -183,7 +_,7 @@
                             }
 
                             if (newLitTime > 0) {
-                                consumeFuel(level, pos, entity.items, fuel);
+                                entity.consumeFuel(level, pos, entity.items, fuel);
                                 isLit = true;
                                 changed = true;
                             }
@@ -194,7 +_,7 @@
                             if (entity.cookingTimer >= entity.cookingTotalTime) {
                                 entity.cookingTimer = 0;
                                 entity.cookingTotalTime = getTotalCookTime(recipe, entity);
-                                burn(entity.items, ingredient, burnResult);
+                                entity.burn(entity.items, ingredient, burnResult);
                                 entity.setRecipeUsed(recipe);
                                 changed = true;
                             }
@@ -223,9 +_,8 @@
         }
     }
 
-    private static void consumeFuel(final ServerLevel level, final BlockPos pos, final NonNullList<ItemStack> items, final ItemStack fuel) {
-        Item fuelItem = fuel.getItem();
-        ItemStackTemplate remainder = fuelItem.getCraftingRemainder();
+    protected void consumeFuel(final ServerLevel level, final BlockPos pos, final NonNullList<ItemStack> items, final ItemStack fuel) {
+        ItemStackTemplate remainder = fuel.getCraftingRemainder();
         ItemStack newFuel = fuel;
         fuel.shrink(1);
         if (remainder != null) {
@@ -239,7 +_,7 @@
         items.set(1, newFuel);
     }
 
-    private static boolean canBurn(final NonNullList<ItemStack> items, final int maxStackSize, final ItemStack burnResult) {
+    protected boolean canBurn(final NonNullList<ItemStack> items, final int maxStackSize, final ItemStack burnResult) {
         ItemStack resultItemStack = items.get(2);
         if (resultItemStack.isEmpty()) {
             return true;
@@ -254,7 +_,7 @@
         return resultCount <= maxResultCount;
     }
 
-    private static void burn(final NonNullList<ItemStack> items, final ItemStack inputItemStack, final ItemStack result) {
+    protected void burn(final NonNullList<ItemStack> items, final ItemStack inputItemStack, final ItemStack result) {
         ItemStack resultItemStack = items.get(2);
         if (resultItemStack.isEmpty()) {
             items.set(2, result.copy());
@@ -270,7 +_,8 @@
     }
 
     protected int getBurnDuration(final ServerLevel level, final ItemStack fuelItem) {
-        return ResolvableInt.getFromItem(fuelItem, DataComponents.COOKING_FUEL, CookingFuel::burnTime, this.getLootContext(level), 0);
+        var vanilla = ResolvableInt.getFromItem(fuelItem, DataComponents.COOKING_FUEL, CookingFuel::burnTime, this.getLootContext(level), 0);
+        return net.minecraftforge.event.ForgeEventFactory.getItemBurnTime(fuelItem, vanilla, recipeType);
     }
 
     protected float getSpeedMultiplier(final ServerLevel level, final ItemStack fuelItem) {
@@ -348,7 +_,7 @@
         }
 
         ItemStack fuelSlot = this.items.get(1);
-        return itemStack.has(DataComponents.COOKING_FUEL) || itemStack.is(Items.BUCKET) && !fuelSlot.is(Items.BUCKET);
+        return itemStack.has(DataComponents.COOKING_FUEL) || (itemStack.is(Items.BUCKET) && !fuelSlot.is(Items.BUCKET)) || net.minecraftforge.event.ForgeEventFactory.getItemBurnTime(itemStack, 0, recipeType) > 0;
     }
 
     @Override
@@ -407,6 +_,35 @@
         for (ItemStack itemStack : this.items) {
             contents.accountStack(itemStack);
         }
+    }
+
+    net.minecraftforge.common.util.LazyOptional<? extends net.minecraftforge.items.IItemHandler>[] handlers =
+        net.minecraftforge.items.wrapper.SidedInvWrapper.create(this, Direction.UP, Direction.DOWN, Direction.NORTH);
+
+    @Override
+    public <T> net.minecraftforge.common.util.LazyOptional<T> getCapability(net.minecraftforge.common.capabilities.Capability<T> capability, @Nullable Direction facing) {
+        if (capability == net.minecraftforge.common.capabilities.ForgeCapabilities.ITEM_HANDLER && facing != null && !this.remove) {
+            return switch (facing) {
+                case UP -> handlers[0].cast();
+                case DOWN -> handlers[1].cast();
+                default -> handlers[2].cast();
+            };
+        }
+        return super.getCapability(capability, facing);
+    }
+
+    @Override
+    public void invalidateCaps() {
+        super.invalidateCaps();
+        for (int x = 0; x < handlers.length; x++) {
+            handlers[x].invalidate();
+        }
+    }
+
+    @Override
+    public void reviveCaps() {
+        super.reviveCaps();
+        this.handlers = net.minecraftforge.items.wrapper.SidedInvWrapper.create(this, Direction.UP, Direction.DOWN, Direction.NORTH);
     }
 
     @Override
