--- a/net/minecraft/world/level/block/entity/ConduitBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/ConduitBlockEntity.java
@@ -147,7 +_,7 @@
                      {
                         BlockPos testPos = worldPosition.offset(ox, oy, oz);
                         BlockState testBlock = level.getBlockState(testPos);
-                        if (testBlock.is(BlockTags.CONDUIT_EFFECT_BLOCK)) {
+                        if (testBlock.isConduitFrame(level, testPos, worldPosition)) {
                             effectBlocks.add(testPos);
                         }
                     }
