--- a/net/minecraft/world/level/block/entity/FuelValues.java
+++ b/net/minecraft/world/level/block/entity/FuelValues.java
@@ -25,7 +_,11 @@
     }
 
     public boolean isFuel(final ItemStack itemStack) {
-        return this.values.containsKey(itemStack.getItem());
+        return isFuel(itemStack, null);
+    }
+
+    public boolean isFuel(ItemStack stack, @org.jetbrains.annotations.Nullable net.minecraft.world.item.crafting.RecipeType<?> type) {
+        return burnDuration(stack, type) > 0;
     }
 
     public SequencedSet<Item> fuelItems() {
@@ -33,7 +_,16 @@
     }
 
     public int burnDuration(final ItemStack itemStack) {
-        return itemStack.isEmpty() ? 0 : this.values.getInt(itemStack.getItem());
+        return burnDuration(itemStack, null);
+    }
+
+    public int burnDuration(final ItemStack itemStack, final @org.jetbrains.annotations.Nullable net.minecraft.world.item.crafting.RecipeType<?> type) {
+        if (itemStack.isEmpty())
+            return 0;
+        int ret = itemStack.getBurnTime(type);
+        if (ret == -1)
+            ret = this.values.getInt(itemStack.getItem());
+        return net.minecraftforge.event.ForgeEventFactory.getItemBurnTime(itemStack, ret, type);
     }
 
     public static FuelValues vanillaBurnTimes(final HolderLookup.Provider registries, final FeatureFlagSet enabledFeatures) {
