--- a/net/minecraft/client/multiplayer/chat/ChatListener.java
+++ b/net/minecraft/client/multiplayer/chat/ChatListener.java
@@ -140,6 +_,8 @@
             LocalPlayer localplayer = this.minecraft.player;
             if (localplayer != null && localplayer.chatAbilities().canReceivePlayerMessages()) {
                 Component component = boundChatType.decorate(message);
+                component = net.minecraftforge.client.ForgeHooksClient.onClientChat(boundChatType, component, Util.NIL_UUID);
+                if (component == null) return false;
                 this.minecraft.gui.getChat().addPlayerMessage(component, null, GuiMessageTag.system());
                 this.narrateChatMessage(boundChatType, message);
                 this.logSystemMessage(component, instant);
@@ -169,12 +_,16 @@
                 MessageSignature messagesignature = message.signature();
                 FilterMask filtermask = message.filterMask();
                 if (filtermask.isEmpty()) {
-                    this.minecraft.gui.getChat().addPlayerMessage(decoratedMessage, messagesignature, guimessagetag);
+                    Component forgeComponent = net.minecraftforge.client.ForgeHooksClient.onClientPlayerChat(boundChatType, decoratedMessage, message, message.sender());
+                    if (forgeComponent == null) return false;
+                    this.minecraft.gui.getChat().addPlayerMessage(forgeComponent, messagesignature, guimessagetag);
                     this.narrateChatMessage(boundChatType, message.decoratedContent());
                 } else {
                     Component component = filtermask.applyWithFormatting(message.signedContent());
                     if (component != null) {
-                        this.minecraft.gui.getChat().addPlayerMessage(boundChatType.decorate(component), messagesignature, guimessagetag);
+                        Component forgeComponent = net.minecraftforge.client.ForgeHooksClient.onClientPlayerChat(boundChatType, boundChatType.decorate(component), message, message.sender());
+                        if (forgeComponent == null) return false;
+                        this.minecraft.gui.getChat().addPlayerMessage(forgeComponent, messagesignature, guimessagetag);
                         this.narrateChatMessage(boundChatType, component);
                     }
                 }
@@ -208,10 +_,12 @@
         chatlog.push(LoggedChatMessage.system(message, timeStamp));
     }
 
-    public void handleSystemMessage(final Component message, final boolean remote) {
+    public void handleSystemMessage(Component message, final boolean remote) {
         if (!this.minecraft.options.hideMatchedNames().get() || !this.minecraft.isBlocked(this.guessChatUUID(message))) {
             LocalPlayer localplayer = this.minecraft.player;
             if (localplayer != null && localplayer.chatAbilities().canReceiveSystemMessages()) {
+                message = net.minecraftforge.client.ForgeHooksClient.onClientSystemMessage(message, remote);
+                if (message == null) return;
                 if (remote) {
                     this.minecraft.gui.getChat().addServerSystemMessage(message);
                     this.logSystemMessage(message, Instant.now());
