--- a/com/mojang/renderpearl/frontend/FrontendGpuDevice.java
+++ b/com/mojang/renderpearl/frontend/FrontendGpuDevice.java
@@ -94,6 +_,21 @@
         return this.createTexture(this.isDebuggingEnabled() && label != null ? label.get() : null, usage, format, width, height, depthOrLayers, mipLevels);
     }
 
+    /** Forge: same as {@link #createTexture(Supplier, int, GpuFormat, int, int, int, int)} but with stencil support */
+    @Override
+    public GpuTexture createTexture(
+        final @Nullable Supplier<String> label,
+        final @GpuTexture.Usage int usage,
+        final GpuFormat format,
+        final int width,
+        final int height,
+        final int depthOrLayers,
+        final int mipLevels,
+        final boolean stencil
+    ) {
+        return this.createTexture(this.isDebuggingEnabled() && label != null ? label.get() : null, usage, format, width, height, depthOrLayers, mipLevels, stencil);
+    }
+
     @Override
     public GpuTexture createTexture(
         final @Nullable String label,
@@ -106,6 +_,22 @@
     ) {
         this.verifyTextureCreationArgs(usage, width, height, depthOrLayers, mipLevels);
         return this.backend.createTexture(label, usage, format, width, height, depthOrLayers, mipLevels);
+    }
+
+    /** Forge: same as {@link #createTexture(String, int, GpuFormat, int, int, int, int)} but with stencil support */
+    @Override
+    public GpuTexture createTexture(
+        final @Nullable String label,
+        final @GpuTexture.Usage int usage,
+        final GpuFormat format,
+        final int width,
+        final int height,
+        final int depthOrLayers,
+        final int mipLevels,
+        final boolean stencil
+    ) {
+        this.verifyTextureCreationArgs(usage, width, height, depthOrLayers, mipLevels);
+        return this.backend.createTexture(label, usage, format, width, height, depthOrLayers, mipLevels, stencil);
     }
 
     private void verifyTextureCreationArgs(final @GpuTexture.Usage int usage, final int width, final int height, final int depthOrLayers, final int mipLevels) {
