--- a/net/minecraft/world/entity/boss/enderdragon/EnderDragonPart.java
+++ b/net/minecraft/world/entity/boss/enderdragon/EnderDragonPart.java
@@ -14,13 +_,13 @@
 import net.minecraft.world.level.storage.ValueOutput;
 import org.jspecify.annotations.Nullable;
 
-public class EnderDragonPart extends Entity {
+public class EnderDragonPart extends net.minecraftforge.entity.PartEntity<EnderDragon> {
     public final EnderDragon parentMob;
     public final String name;
     private final EntityDimensions size;
 
     public EnderDragonPart(EnderDragon p_450227_, String p_453842_, float p_460816_, float p_454700_) {
-        super(p_450227_.getType(), p_450227_.level());
+        super(p_450227_);
         this.size = EntityDimensions.scalable(p_460816_, p_454700_);
         this.refreshDimensions();
         this.parentMob = p_450227_;
