--- a/net/minecraft/client/KeyboardHandler.java
+++ b/net/minecraft/client/KeyboardHandler.java
@@ -497,7 +_,7 @@
             if (screen != null) {
                 try {
                     if (p_90895_ != 1 && p_90895_ != 2) {
-                        if (p_90895_ == 0 && screen.keyReleased(p_423534_)) {
+                        if (p_90895_ == 0 && net.minecraftforge.client.ForgeHooksClient.onScreenKeyReleased(screen, p_423534_)) {
                             if (options.keyDebugModifier.matches(p_423534_)) {
                                 this.usedDebugKeyAsModifier = false;
                             }
@@ -506,7 +_,7 @@
                         }
                     } else {
                         screen.afterKeyboardAction();
-                        if (screen.keyPressed(p_423534_)) {
+                        if (net.minecraftforge.client.ForgeHooksClient.onScreenKeyPressed(screen, p_423534_)) {
                             if (this.minecraft.screen == null) {
                                 InputConstants.Key inputconstants$key = InputConstants.getKey(p_423534_);
                                 KeyMapping.set(inputconstants$key, false);
@@ -582,6 +_,7 @@
                     }
                 }
             }
+            net.minecraftforge.client.ForgeHooksClient.onKeyInput(p_423534_, p_90895_);
         }
     }
 
@@ -590,7 +_,7 @@
             Screen screen = this.minecraft.screen;
             if (screen != null && this.minecraft.getOverlay() == null) {
                 try {
-                    screen.charTyped(p_427531_);
+                    net.minecraftforge.client.ForgeHooksClient.onScreenCharTyped(screen, p_427531_);
                 } catch (Throwable throwable) {
                     CrashReport crashreport = CrashReport.forThrowable(throwable, "charTyped event handler");
                     screen.fillCrashDetails(crashreport);
