--- a/net/minecraft/client/KeyboardHandler.java
+++ b/net/minecraft/client/KeyboardHandler.java
@@ -514,7 +_,7 @@
             if (screen != null) {
                 try {
                     if (action != 1 && action != 2) {
-                        if (action == 0 && screen.keyReleased(event)) {
+                        if (action == 0 && net.minecraftforge.client.ForgeHooksClient.onScreenKeyReleased(screen, event)) {
                             if (options.keyDebugModifier.matches(event)) {
                                 this.usedDebugKeyAsModifier = false;
                             }
@@ -523,7 +_,7 @@
                         }
                     } else {
                         screen.afterKeyboardAction();
-                        if (screen.keyPressed(event)) {
+                        if (net.minecraftforge.client.ForgeHooksClient.onScreenKeyPressed(screen, event)) {
                             if (this.minecraft.screen == null) {
                                 InputConstants.Key key = InputConstants.getKey(event);
                                 KeyMapping.set(key, false);
@@ -595,6 +_,7 @@
                     }
                 }
             }
+            net.minecraftforge.client.ForgeHooksClient.onKeyInput(event, action);
         }
     }
 
@@ -603,7 +_,7 @@
             Screen screen = this.minecraft.screen;
             if (screen != null && this.minecraft.getOverlay() == null) {
                 try {
-                    screen.charTyped(event);
+                    net.minecraftforge.client.ForgeHooksClient.onScreenCharTyped(screen, event);
                 } catch (Throwable var8) {
                     CrashReport report = CrashReport.forThrowable(var8, "charTyped event handler");
                     screen.fillCrashDetails(report);
