--- a/net/minecraft/client/MouseHandler.java
+++ b/net/minecraft/client/MouseHandler.java
@@ -79,6 +_,7 @@
                 this.activeButton = null;
             }
 
+            if (net.minecraftforge.client.event.ForgeEventFactoryClient.onMouseButtonPre(rawButtonInfo, action)) return;
             if (this.minecraft.gui.overlay() == null) {
                 if (pressed
                     && this.minecraft
@@ -104,7 +_,7 @@
                                 && currentTime - this.lastClick.time() < 250L
                                 && this.lastClick.screen() == screen
                                 && this.lastClickButton == event.button();
-                            if (screen.mouseClicked(event, doubleClick)) {
+                            if (net.minecraftforge.client.event.ForgeEventFactoryClient.onScreenMouseClicked(screen, xm, ym, event, doubleClick)) {
                                 this.lastClick = new MouseHandler.LastClick(currentTime, screen);
                                 this.lastClickButton = rawButtonInfo.button();
                                 return;
@@ -118,7 +_,7 @@
                         }
                     } else {
                         try {
-                            if (screen.mouseReleased(event)) {
+                            if (net.minecraftforge.client.event.ForgeEventFactoryClient.onScreenMouseReleased(screen, xm, ym, event)) {
                                 return;
                             }
                         } catch (Throwable t) {
@@ -147,6 +_,7 @@
                     KeyMapping.click(mouseKey);
                 }
             }
+            net.minecraftforge.client.event.ForgeEventFactoryClient.onMouseButtonPost(rawButtonInfo, action);
         }
     }
 
@@ -182,7 +_,9 @@
                 if (this.minecraft.gui.screen() != null) {
                     double xm = this.getScaledXPos(this.minecraft.getWindow());
                     double ym = this.getScaledYPos(this.minecraft.getWindow());
-                    this.minecraft.gui.screen().mouseScrolled(xm, ym, scaledXOffset, scaledYOffset);
+                    if (net.minecraftforge.client.event.ForgeEventFactoryClient.onScreenMouseScrollPre(this.minecraft.gui.screen(), xm, ym, scaledXOffset, scaledYOffset)) return;
+                    if (this.minecraft.gui.screen().mouseScrolled(xm, ym, scaledXOffset, scaledYOffset)) return;
+                    net.minecraftforge.client.event.ForgeEventFactoryClient.onScreenMouseScrollPost(this.minecraft.gui.screen(), xm, ym, scaledXOffset, scaledYOffset);
                     this.minecraft.gui.screen().afterMouseAction();
                 } else if (this.minecraft.player != null) {
                     Vector2i wheelXY = this.scrollWheelHandler.onMouseScroll(scaledXOffset, scaledYOffset);
@@ -191,6 +_,7 @@
                     }
 
                     int wheel = wheelXY.y == 0 ? -wheelXY.x : wheelXY.y;
+                    if (net.minecraftforge.client.event.ForgeEventFactoryClient.onMouseScroll(this, scaledXOffset, scaledYOffset)) return;
                     if (this.minecraft.player.isSpectator()) {
                         if (this.minecraft.gui.hud.getSpectatorGui().isMenuActive()) {
                             this.minecraft.gui.hud.getSpectatorGui().onMouseScrolled(-wheel);
@@ -285,7 +_,7 @@
                     double dy = getScaledYPos(window, this.accumulatedDY);
 
                     try {
-                        screen.mouseDragged(new MouseButtonEvent(xm, ym, this.activeButton), dx, dy);
+                        net.minecraftforge.client.ForgeHooksClient.onScreenMouseDrag(screen, new MouseButtonEvent(xm, ym, this.activeButton), dx, dy);
                     } catch (Throwable t) {
                         CrashReport report = CrashReport.forThrowable(t, "mouseDragged event handler");
                         CrashReportCategory mouseDetails = report.addCategory("Mouse");
@@ -369,6 +_,14 @@
 
     public double ypos() {
         return this.ypos;
+    }
+
+    public double getXVelocity() {
+        return this.accumulatedDX;
+    }
+
+    public double getYVelocity() {
+        return this.accumulatedDY;
     }
 
     public void setIgnoreFirstMove() {
