--- a/net/minecraft/client/Screenshot.java
+++ b/net/minecraft/client/Screenshot.java
@@ -45,6 +_,13 @@
                     file2 = new File(file1, forceName);
                 }
 
+                var event = new net.minecraftforge.client.event.ScreenshotEvent(image, file2);
+                if (net.minecraftforge.client.event.ScreenshotEvent.BUS.post(event)) {
+                    callback.accept(event.getCancelMessage());
+                    return;
+                }
+                final File outputFile = event.getScreenshotFile();
+
                 Util.ioPool()
                     .execute(
                         () -> {
@@ -52,10 +_,13 @@
                                 NativeImage nativeimage = image;
 
                                 try {
-                                    image.writeToFile(file2);
-                                    Component component = Component.literal(file2.getName())
+                                    image.writeToFile(outputFile);
+                                    Component component = Component.literal(outputFile.getName())
                                         .withStyle(ChatFormatting.UNDERLINE)
                                         .withStyle(s -> s.withClickEvent(new ClickEvent.OpenFile(file2.getAbsoluteFile())));
+                                    if (event.getResultMessage() != null)
+                                        callback.accept(event.getResultMessage());
+                                    else
                                     callback.accept(Component.translatable("screenshot.success", component));
                                 } catch (Throwable throwable1) {
                                     if (image != null) {
