--- a/net/minecraft/client/resources/model/ModelDiscovery.java
+++ b/net/minecraft/client/resources/model/ModelDiscovery.java
@@ -177,6 +_,14 @@
             return this.parent;
         }
 
+        private net.minecraftforge.client.model.geometry.@Nullable ModelContext context = null;
+        @Override
+        public net.minecraftforge.client.model.geometry.IGeometryBakingContext getContext() {
+            if (context == null || context.parent() != parent)
+                context = new net.minecraftforge.client.model.geometry.ModelContext(this);
+            return context;
+        }
+
         @Override
         public String debugName() {
             return this.id.toString();
@@ -229,14 +_,14 @@
 
         private QuadCollection bakeDefaultState(final TextureSlots textureSlots, final ModelBaker baker, final ModelState state) {
             QuadCollection result = this.getSlot(KEY_DEFAULT_GEOMETRY);
-            return result != null ? result : this.updateSlot(KEY_DEFAULT_GEOMETRY, this.getTopGeometry().bake(textureSlots, baker, state, this));
+            return result != null ? result : this.updateSlot(KEY_DEFAULT_GEOMETRY, this.getTopGeometry().bake(textureSlots, baker, state, this, getContext()));
         }
 
         @Override
         public QuadCollection bakeTopGeometry(final TextureSlots textureSlots, final ModelBaker baker, final ModelState state) {
             return state == BlockModelRotation.IDENTITY ? this.bakeDefaultState(textureSlots, baker, state) : this.modelBakeCache.computeIfAbsent(state, s -> {
                 UnbakedGeometry topGeometry = this.getTopGeometry();
-                return topGeometry.bake(textureSlots, baker, s, this);
+                return topGeometry.bake(textureSlots, baker, s, this, getContext());
             });
         }
     }
