--- a/net/minecraft/world/level/biome/Biome.java
+++ b/net/minecraft/world/level/biome/Biome.java
@@ -71,10 +_,10 @@
     @Deprecated(forRemoval = true)
     public static final Noise BIOME_INFO_NOISE = new SimplexNoise(new WorldgenRandom(new LegacyRandomSource(2345L)), true);
     private static final int TEMPERATURE_CACHE_SIZE = 1024;
-    private final Biome.ClimateSettings climateSettings;
-    private final BiomeGenerationSettings generationSettings;
-    private final EnvironmentAttributeMap attributes;
-    private final BiomeSpecialEffects specialEffects;
+    private Biome.ClimateSettings climateSettings;
+    private BiomeGenerationSettings generationSettings;
+    private EnvironmentAttributeMap attributes;
+    private BiomeSpecialEffects specialEffects;
     private final ThreadLocal<Long2FloatLinkedOpenHashMap> temperatureCache = ThreadLocal.withInitial(() -> {
         Long2FloatLinkedOpenHashMap map = new Long2FloatLinkedOpenHashMap(1024, 0.25F) {
             @Override
@@ -84,6 +_,7 @@
         map.defaultReturnValue(Float.NaN);
         return map;
     });
+    private final net.minecraftforge.common.world.ModifiableBiomeInfo modifiableBiomeInfo;
 
     private Biome(
         final Biome.ClimateSettings climateSettings,
@@ -95,6 +_,15 @@
         this.attributes = attributes;
         this.specialEffects = specialEffects;
         this.generationSettings = generationSettings;
+        var info = new net.minecraftforge.common.world.ModifiableBiomeInfo.BiomeInfo(climateSettings, attributes, specialEffects, generationSettings);
+        this.modifiableBiomeInfo = new net.minecraftforge.common.world.ModifiableBiomeInfo(info, this::update);
+    }
+
+    private void update(net.minecraftforge.common.world.ModifiableBiomeInfo.BiomeInfo info) {
+        this.climateSettings = info.climateSettings();
+        this.attributes = info.attributes();
+        this.specialEffects = info.effects();
+        this.generationSettings = info.generationSettings();
     }
 
     public boolean hasPrecipitation() {
@@ -196,7 +_,7 @@
     }
 
     public BiomeGenerationSettings getGenerationSettings() {
-        return this.generationSettings;
+        return this.modifiableBiomeInfo().get().generationSettings();
     }
 
     public int getGrassColor(final double x, final double z) {
@@ -249,6 +_,31 @@
 
     public int getWaterColor() {
         return this.specialEffects.waterColor();
+    }
+
+    /**
+     * {@return Cache of original biome data and biome data modified by biome modifiers}
+     * Modified biome data is set by server after datapacks and serverconfigs load.
+     * Climate and effects field reads are coremodded to redirect to this.
+     **/
+    public net.minecraftforge.common.world.ModifiableBiomeInfo modifiableBiomeInfo() {
+        return this.modifiableBiomeInfo;
+    }
+
+    /**
+     * {@return The biome's climate settings, with modifications if called after modifiers are applied in server init.}
+     * Note: This method is called by our field_to_method coremod. Which redirects {@link #climateSettings} to this method at load time.
+     */
+    public ClimateSettings getModifiedClimateSettings() {
+        return this.modifiableBiomeInfo().get().climateSettings();
+    }
+
+    /**
+     * {@return The biome's client effects, with modifications if called after modifiers are applied in server init.}
+     * Note: This methods is called by our field_to_method coremod. Which redirects {@link #specialEffects} to this method at load time.
+     */
+    public BiomeSpecialEffects getModifiedSpecialEffects() {
+        return this.modifiableBiomeInfo().get().effects();
     }
 
     public static class BiomeBuilder {
