--- a/net/minecraft/world/entity/monster/CrossbowAttackMob.java
+++ b/net/minecraft/world/entity/monster/CrossbowAttackMob.java
@@ -16,9 +_,10 @@
     void onCrossbowAttackPerformed();
 
     default void performCrossbowAttack(final LivingEntity body, final float crossbowPower) {
-        InteractionHand interactionhand = ProjectileUtil.getWeaponHoldingHand(body, Items.CROSSBOW);
+        InteractionHand interactionhand = ProjectileUtil.getWeaponHoldingHand(body, item -> item instanceof CrossbowItem);
         ItemStack itemstack = body.getItemInHand(interactionhand);
-        if (itemstack.getItem() instanceof CrossbowItem crossbowitem) {
+        if (body.isHolding(is -> is.getItem() instanceof CrossbowItem)) {
+            var crossbowitem = (CrossbowItem) itemstack.getItem();
             crossbowitem.performShooting(
                 body.level(), body, interactionhand, itemstack, crossbowPower, 14 - body.level().getDifficulty().getId() * 4, this.getTarget()
             );
