--- a/net/minecraft/client/renderer/entity/player/AvatarRenderer.java
+++ b/net/minecraft/client/renderer/entity/player/AvatarRenderer.java
@@ -94,7 +_,7 @@
             return HumanoidModel.ArmPose.EMPTY;
         }
 
-        if (!avatar.isSwinging() && itemInHand.is(Items.CROSSBOW) && CrossbowItem.isCharged(itemInHand)) {
+        if (!avatar.isSwinging() && itemInHand.getItem() instanceof CrossbowItem && CrossbowItem.isCharged(itemInHand)) {
             return HumanoidModel.ArmPose.CROSSBOW_HOLD;
         }
 
@@ -135,7 +_,7 @@
 
         return HumanoidMobRenderer.usesSpearPose(itemInHand, hand.asArm(avatar.getMainArm()), avatar)
             ? HumanoidModel.ArmPose.SPEAR
-            : HumanoidModel.ArmPose.ITEM;
+            : net.minecraftforge.client.extensions.common.IClientItemExtensions.of(itemInHand).getArmPose(avatar, hand, itemInHand);
     }
 
     public Identifier getTextureLocation(final AvatarRenderState state) {
@@ -234,12 +_,14 @@
     public void renderRightHand(
         final PoseStack poseStack, final SubmitNodeCollector submitNodeCollector, final int lightCoords, final Identifier skinTexture, final boolean hasSleeve
     ) {
+        if (!net.minecraftforge.client.event.ForgeEventFactoryClient.onRenderArm(poseStack, submitNodeCollector, lightCoords, HumanoidArm.RIGHT))
         this.renderHand(poseStack, submitNodeCollector, lightCoords, skinTexture, this.model.rightArm, hasSleeve);
     }
 
     public void renderLeftHand(
         final PoseStack poseStack, final SubmitNodeCollector submitNodeCollector, final int lightCoords, final Identifier skinTexture, final boolean hasSleeve
     ) {
+        if (!net.minecraftforge.client.event.ForgeEventFactoryClient.onRenderArm(poseStack, submitNodeCollector, lightCoords, HumanoidArm.LEFT))
         this.renderHand(poseStack, submitNodeCollector, lightCoords, skinTexture, this.model.leftArm, hasSleeve);
     }
 
@@ -297,5 +_,12 @@
 
     public static boolean isPlayerUpsideDown(final Player player) {
         return isUpsideDownName(player.getGameProfile().name());
+    }
+
+    @Override
+    public void submit(AvatarRenderState state, PoseStack pose, SubmitNodeCollector source, CameraRenderState camera) {
+        if (net.minecraftforge.client.event.ForgeEventFactoryClient.onRenderAvatarPre(state, this, pose, source, camera)) return;
+        super.submit(state, pose, source, camera);
+        net.minecraftforge.client.event.ForgeEventFactoryClient.onRenderAvatarPost(state, this, pose, source, camera);
     }
 }
