--- a/net/minecraft/client/gui/screens/LoadingOverlay.java
+++ b/net/minecraft/client/gui/screens/LoadingOverlay.java
@@ -64,6 +_,8 @@
         return color & 16777215 | alpha << 24;
     }
 
+    protected boolean extractContentRenderState(final GuiGraphicsExtractor graphics, float alpha) { return true; }
+
     @Override
     public void extractRenderState(final GuiGraphicsExtractor graphics, final int mouseX, final int mouseY, final float a) {
         int i = graphics.guiWidth();
@@ -103,6 +_,7 @@
             f2 = 1.0F;
         }
 
+        if (extractContentRenderState(graphics, f2)) {
         int k2 = (int)(graphics.guiWidth() * 0.5);
         int i1 = (int)(graphics.guiHeight() * 0.5);
         double d0 = Math.min(graphics.guiWidth() * 0.75, (double)graphics.guiHeight()) * 0.25;
@@ -118,6 +_,7 @@
         if (f < 1.0F) {
             this.extractProgressBar(graphics, i / 2 - k1, i2 - 5, i / 2 + k1, i2 + 5, 1.0F - Mth.clamp(f, 0.0F, 1.0F));
         }
+        }
 
         if (f >= 2.0F) {
             this.minecraft.setOverlay(null);
@@ -127,6 +_,7 @@
     @Override
     public void tick() {
         if (this.fadeOutStart == -1L && this.reload.isDone() && this.isReadyToFadeOut()) {
+            this.fadeOutStart = Util.getMillis(); // Forge: Moved up to guard against inf loops caused by callback
             try {
                 this.reload.checkExceptions();
                 this.onFinish.accept(Optional.empty());
@@ -134,7 +_,6 @@
                 this.onFinish.accept(Optional.of(throwable));
             }
 
-            this.fadeOutStart = Util.getMillis();
             if (this.minecraft.screen != null) {
                 Window window = this.minecraft.getWindow();
                 this.minecraft.screen.init(window.getGuiScaledWidth(), window.getGuiScaledHeight());
