--- a/net/minecraft/data/loot/LootTableProvider.java
+++ b/net/minecraft/data/loot/LootTableProvider.java
@@ -60,7 +_,7 @@
     private CompletableFuture<?> run(final CachedOutput cache, final HolderLookup.Provider registries) {
         WritableRegistry<LootTable> writableregistry = new MappedRegistry<>(Registries.LOOT_TABLE, Lifecycle.experimental());
         Map<RandomSupport.Seed128bit, Identifier> map = new Object2ObjectOpenHashMap<>();
-        this.subProviders.forEach(subProvider -> subProvider.provider().apply(registries).generate((id, lootTable) -> {
+        this.getTables().forEach(subProvider -> subProvider.provider().apply(registries).generate((id, lootTable) -> {
             Identifier identifier = sequenceIdForLootTable(id);
             Identifier identifier1 = map.put(RandomSequence.seedForKey(identifier), identifier);
             if (identifier1 != null) {
@@ -76,11 +_,8 @@
         HolderGetter.Provider holdergetter$provider = new RegistryAccess.ImmutableRegistryAccess(List.of(writableregistry)).freeze();
         ValidationContextSource validationcontextsource = new ValidationContextSource(problemreporter$collector, holdergetter$provider);
 
-        for (ResourceKey<LootTable> resourcekey : Sets.difference(this.requiredTables, writableregistry.registryKeySet())) {
-            problemreporter$collector.report(new LootTableProvider.MissingTableProblem(resourcekey));
-        }
+        validate(writableregistry, validationcontextsource, problemreporter$collector);
 
-        LootDataType.TABLE.runValidation(validationcontextsource, writableregistry);
         if (!problemreporter$collector.isEmpty()) {
             problemreporter$collector.forEach((id, problem) -> LOGGER.warn("Found validation problem in {}: {}", id, problem.description()));
             throw new IllegalStateException("Failed to validate loot tables, see logs");
@@ -101,6 +_,18 @@
     @Override
     public final String getName() {
         return "Loot Tables";
+    }
+
+    public List<LootTableProvider.SubProviderEntry> getTables() {
+        return this.subProviders;
+    }
+
+    protected void validate(net.minecraft.core.Registry<LootTable> map, ValidationContextSource context, ProblemReporter report) {
+        for (ResourceKey<LootTable> resourcekey : Sets.difference(this.requiredTables, map.registryKeySet())) {
+            report.report(new LootTableProvider.MissingTableProblem(resourcekey));
+        }
+
+        LootDataType.TABLE.runValidation(context, map);
     }
 
     public record MissingTableProblem(ResourceKey<LootTable> id) implements ProblemReporter.Problem {
