--- a/net/minecraft/data/loot/LootTableProvider.java
+++ b/net/minecraft/data/loot/LootTableProvider.java
@@ -33,7 +_,7 @@
         Map<RandomSupport.Seed128bit, Identifier> randomSequenceSeeds = new Object2ObjectOpenHashMap<>();
         HolderGetter<LootTable> lootTables = context.lookup(Registries.LOOT_TABLE);
         this.requiredTables.forEach(lootTables::get);
-        this.subProviders.forEach(subProvider -> subProvider.bootstrap().create(new LootTableSubProvider.Context() {
+        this.getTables().forEach(subProvider -> subProvider.bootstrap().create(new LootTableSubProvider.Context() {
             @Override
             public Holder.Reference<LootTable> accept(final ResourceKey<LootTable> key, final LootTable.Builder lootTable) {
                 Identifier sequenceId = LootTableProvider.sequenceIdForLootTable(key);
@@ -61,6 +_,13 @@
 
     private static Identifier sequenceIdForLootTable(final ResourceKey<LootTable> id) {
         return id.identifier();
+    }
+
+    public Set<ResourceKey<LootTable>> getRequired() {
+        return this.requiredTables;
+    }
+    public List<LootTableProvider.SubProviderEntry> getTables() {
+        return this.subProviders;
     }
 
     public record SubProviderEntry(LootTableSubProvider.Factory bootstrap, ContextKeySet paramSet) {
