--- a/net/minecraft/server/packs/resources/FallbackResourceManager.java
+++ b/net/minecraft/server/packs/resources/FallbackResourceManager.java
@@ -103,8 +_,11 @@
 
         for (int i = this.fallbacks.size() - 1; i >= 0; i--) {
             FallbackResourceManager.PackEntry entry = this.fallbacks.get(i);
-            PackResources fileSource = entry.resources;
-            if (fileSource != null) {
+            PackResources pack = entry.resources;
+            if (pack != null) {
+                var children = pack.getChildren();
+                var packs = children == null ? List.of(pack) : children;
+                for (final PackResources fileSource : packs) {
                 IoSupplier<InputStream> resource = fileSource.getResource(this.type, location);
                 if (resource != null) {
                     IoSupplier<ResourceMetadata> metadataGetter;
@@ -118,6 +_,7 @@
                     }
 
                     result.add(new Resource(fileSource, resource, metadataGetter));
+                }
                 }
             }
 
