--- a/net/minecraft/server/packs/resources/SimpleJsonResourceReloadListener.java
+++ b/net/minecraft/server/packs/resources/SimpleJsonResourceReloadListener.java
@@ -27,6 +_,10 @@
         this(JsonOps.INSTANCE, codec, lister);
     }
 
+    protected SimpleJsonResourceReloadListener(final Codec<T> codec, final FileToIdConverter lister, net.minecraftforge.common.crafting.conditions.ICondition.IContext context) {
+        this(context.wrap(JsonOps.INSTANCE), codec, lister);
+    }
+
     protected SimpleJsonResourceReloadListener(final DynamicOps<JsonElement> ops, final Codec<T> codec, final FileToIdConverter lister) {
         this.ops = ops;
         this.codec = codec;
@@ -41,7 +_,15 @@
             Identifier id = this.lister.fileToId(location);
 
             try (Reader reader = entry.getValue().openAsReader()) {
-                this.codec.parse(this.ops, StrictJsonParser.parse(reader)).ifSuccess(parsed -> {
+                var json = StrictJsonParser.parse(reader);
+                json = net.minecraftforge.common.ForgeHooks.readConditional(this.ops, json);
+                if (json == null) {
+                    LOGGER.debug("Skipping loading {} as its conditions were not met", location);
+                    continue;
+                }
+                this.codec.parse(this.ops, json).ifSuccess(parsed -> {
+                    parsed = net.minecraftforge.common.ForgeHooks.onJsonDataParsed(this.codec, id, parsed);
+                    if (parsed == null) return;
                     if (result.putIfAbsent(id, (T)parsed) != null) {
                         throw new IllegalStateException("Duplicate data file ignored with ID " + id);
                     }
