--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -124,7 +_,7 @@
 import net.minecraft.world.scores.Team;
 import org.jspecify.annotations.Nullable;
 
-public abstract class Player extends Avatar implements ContainerUser {
+public abstract class Player extends Avatar implements ContainerUser, net.minecraftforge.common.extensions.IForgePlayer {
     public static final int MAX_HEALTH = 20;
     public static final int SLEEP_DURATION = 100;
     public static final int WAKE_UP_DURATION = 10;
@@ -172,6 +_,9 @@
     private Optional<GlobalPos> lastDeathLocation = Optional.empty();
     public @Nullable FishingHook fishing;
     protected float hurtDir;
+    private final java.util.Collection<MutableComponent> prefixes = new java.util.LinkedList<>();
+    private final java.util.Collection<MutableComponent> suffixes = new java.util.LinkedList<>();
+    @Nullable private Pose forcedPose;
 
     public Player(final Level level, final GameProfile gameProfile) {
         super(EntityType.PLAYER, level);
@@ -180,6 +_,17 @@
         this.inventory = new Inventory(this, this.equipment);
         this.inventoryMenu = new InventoryMenu(this.inventory, !level.isClientSide(), this);
         this.containerMenu = this.inventoryMenu;
+
+        // Forge: wrapped inventories moved to constructor as the inventory init was moved in here too
+        this.playerMainHandler = net.minecraftforge.common.util.LazyOptional.of(
+            () -> new net.minecraftforge.items.wrapper.PlayerMainInvWrapper(inventory));
+        this.playerEquipmentHandler = net.minecraftforge.common.util.LazyOptional.of(
+            () -> new net.minecraftforge.items.wrapper.PlayerEquipmentInvWrapper(inventory));
+        this.playerJoinedHandler = net.minecraftforge.common.util.LazyOptional.of(
+            () -> new net.minecraftforge.items.wrapper.CombinedInvWrapper(
+                new net.minecraftforge.items.wrapper.PlayerMainInvWrapper(inventory),
+                new net.minecraftforge.items.wrapper.PlayerEquipmentInvWrapper(inventory)
+            ));
     }
 
     @Override
@@ -213,7 +_,8 @@
             .add(Attributes.MINING_EFFICIENCY)
             .add(Attributes.SWEEPING_DAMAGE_RATIO)
             .add(Attributes.WAYPOINT_TRANSMIT_RANGE, 6.0E7)
-            .add(Attributes.WAYPOINT_RECEIVE_RANGE, 6.0E7);
+            .add(Attributes.WAYPOINT_RECEIVE_RANGE, 6.0E7)
+            .add(Attributes.ATTACK_KNOCKBACK);
     }
 
     @Override
@@ -227,6 +_,7 @@
 
     @Override
     public void tick() {
+        net.minecraftforge.event.ForgeEventFactory.onPlayerPreTick(this);
         this.noPhysics = this.isSpectator();
         if (this.isSpectator() || this.isPassenger()) {
             this.setOnGround(false);
@@ -243,7 +_,7 @@
             }
 
             if (!this.level().isClientSide()
-                && !this.level().environmentAttributes().getValue(EnvironmentAttributes.BED_RULE, this.position()).canSleep(this.level())) {
+                && !net.minecraftforge.event.ForgeEventFactory.onSleepingTimeCheck(this, getSleepingPos(), this.level().environmentAttributes().getValue(EnvironmentAttributes.BED_RULE, this.position()))) {
                 this.stopSleepInBed(false, true);
             }
         } else if (this.sleepCounter > 0) {
@@ -308,6 +_,7 @@
         if (!this.getAbilities().flying) {
             super.onAboveBubbleColumn(dragDown, pos);
         }
+        net.minecraftforge.event.ForgeEventFactory.onPlayerPostTick(this);
     }
 
     @Override
@@ -338,6 +_,10 @@
     }
 
     protected void updatePlayerPose() {
+        if (forcedPose != null) {
+            this.setPose(forcedPose);
+            return;
+        }
         if (this.canPlayerFitWithinBlocksAndEntitiesWhen(Pose.SWIMMING)) {
             Pose pose = this.getDesiredPose();
             Pose pose1;
@@ -522,6 +_,7 @@
 
     @Override
     public void die(final DamageSource source) {
+        if (net.minecraftforge.event.ForgeEventFactory.onLivingDeath(this, source)) return;
         super.die(source);
         this.reapplyPosition();
         if (!this.isSpectator() && this.level() instanceof ServerLevel serverlevel) {
@@ -578,10 +_,15 @@
     }
 
     public @Nullable ItemEntity drop(final ItemStack itemStack, final boolean thrownFromHand) {
-        return this.drop(itemStack, false, thrownFromHand);
+        return net.minecraftforge.common.ForgeHooks.onPlayerTossEvent(this, itemStack, thrownFromHand);
     }
 
+    /** @deprecated Use {@link #getDestroySpeed(BlockState,BlockPos) */
     public float getDestroySpeed(final BlockState state) {
+        return getDestroySpeed(state, null);
+    }
+
+    public float getDestroySpeed(BlockState state, @Nullable BlockPos pos) {
         float f = this.inventory.getSelectedItem().getDestroySpeed(state);
         if (f > 1.0F) {
             f += (float)this.getAttributeValue(Attributes.MINING_EFFICIENCY);
@@ -610,11 +_,14 @@
             f /= 5.0F;
         }
 
+        f = net.minecraftforge.event.ForgeEventFactory.getBreakSpeed(this, state, f, pos);
+
         return f;
     }
 
     public boolean hasCorrectToolForDrops(final BlockState state) {
-        return !state.requiresCorrectToolForDrops() || this.inventory.getSelectedItem().isCorrectToolForDrops(state);
+        var vanilla = !state.requiresCorrectToolForDrops() || this.inventory.getSelectedItem().isCorrectToolForDrops(state);
+        return net.minecraftforge.event.ForgeEventFactory.doPlayerHarvestCheck(this, state, vanilla);
     }
 
     @Override
@@ -675,6 +_,7 @@
 
     @Override
     public boolean hurtServer(final ServerLevel level, final DamageSource source, float damage) {
+        if (!net.minecraftforge.common.ForgeHooks.onPlayerAttack(this, source, damage)) return false;
         if (this.isInvulnerableTo(level, source)) {
             return false;
         } else if (this.abilities.invulnerable && !source.is(DamageTypeTags.BYPASSES_INVULNERABILITY)) {
@@ -743,10 +_,13 @@
     @Override
     protected void actuallyHurt(final ServerLevel level, final DamageSource source, float dmg) {
         if (!this.isInvulnerableTo(level, source)) {
+            dmg = net.minecraftforge.common.ForgeHooks.onLivingHurt(this, source, dmg);
+            if (dmg <= 0) return;
             dmg = this.getDamageAfterArmorAbsorb(source, dmg);
             dmg = this.getDamageAfterMagicAbsorb(source, dmg);
             float f1 = Math.max(dmg - this.getAbsorptionAmount(), 0.0F);
             this.setAbsorptionAmount(this.getAbsorptionAmount() - (dmg - f1));
+            f1 = net.minecraftforge.common.ForgeHooks.onLivingDamage(this, source, f1);
             float f = dmg - f1;
             if (f > 0.0F && f < 3.4028235E37F) {
                 this.awardStat(Stats.DAMAGE_ABSORBED, Math.round(f * 10.0F));
@@ -824,6 +_,10 @@
 
             return InteractionResult.PASS;
         } else {
+            var event = new net.minecraftforge.event.entity.player.PlayerInteractEvent.EntityInteractSpecific(this, hand, entity, location);
+            if (net.minecraftforge.event.entity.player.PlayerInteractEvent.EntityInteractSpecific.BUS.post(event)) {
+                return event.getCancellationResult();
+            }
             ItemStack itemstack = this.getItemInHand(hand);
             ItemStack itemstack1 = itemstack.copy();
             InteractionResult interactionresult = entity.interact(this, hand, location);
@@ -832,6 +_,10 @@
                     itemstack.setCount(itemstack1.getCount());
                 }
 
+                if (!this.abilities.instabuild && itemstack.isEmpty()) {
+                    net.minecraftforge.event.ForgeEventFactory.onPlayerDestroyItem(this, itemstack1, hand.asEquipmentSlot());
+                }
+
                 return interactionresult;
             } else {
                 if (!itemstack.isEmpty() && entity instanceof LivingEntity) {
@@ -843,6 +_,7 @@
                     if (interactionresult1.consumesAction()) {
                         this.level().gameEvent(GameEvent.ENTITY_INTERACT, entity.position(), GameEvent.Context.of(this));
                         if (itemstack.isEmpty() && !this.hasInfiniteMaterials()) {
+                            net.minecraftforge.event.ForgeEventFactory.onPlayerDestroyItem(this, itemstack1, hand.asEquipmentSlot());
                             this.setItemInHand(hand, ItemStack.EMPTY);
                         }
 
@@ -942,6 +_,7 @@
     }
 
     public void attack(final Entity entity) {
+        if (!net.minecraftforge.common.ForgeHooks.onPlayerAttackTarget(this, entity)) return;
         if (!this.cannotAttack(entity)) {
             float f = this.isAutoSpinAttack() ? this.autoSpinAttackDmg : (float)this.getAttributeValue(Attributes.ATTACK_DAMAGE);
             ItemStack itemstack = this.getWeaponItem();
@@ -963,8 +_,10 @@
 
                     f += itemstack.getItem().getAttackDamageBonus(entity, f, damagesource);
                     boolean flag2 = flag && this.canCriticalAttack(entity);
+                    var hitResult = net.minecraftforge.common.ForgeHooks.getCriticalHit(this, entity, flag2, flag2 ? 1.5F : 1.0F);
+                    flag2 = hitResult != null;
                     if (flag2) {
-                        f *= 1.5F;
+                        f *= hitResult.getDamageModifier();
                     }
 
                     float f3 = f + f2;
@@ -1036,7 +_,7 @@
             double d0 = this.getKnownMovement().horizontalDistanceSqr();
             double d1 = this.getSpeed() * 2.5;
             if (d0 < Mth.square(d1)) {
-                return this.getItemInHand(InteractionHand.MAIN_HAND).is(ItemTags.SWORDS);
+                return this.getItemInHand(InteractionHand.MAIN_HAND).canPerformAction(net.minecraftforge.common.ToolActions.SWORD_SWEEP);
             }
         }
 
@@ -1079,8 +_,8 @@
 
     private void itemAttackInteraction(final Entity entity, final ItemStack attackingItemStack, final DamageSource damageSource, final boolean applyToTarget) {
         Entity entityx = entity;
-        if (entity instanceof EnderDragonPart) {
-            entityx = ((EnderDragonPart)entity).parentMob;
+        if (entity instanceof net.minecraftforge.entity.PartEntity<?> pe) {
+            entityx = pe.getParent();
         }
 
         boolean flag = false;
@@ -1105,6 +_,7 @@
                 } else {
                     this.setItemInHand(InteractionHand.OFF_HAND, ItemStack.EMPTY);
                 }
+                this.resetAttackStrengthTicker(); // FORGE: Moved from beginning of attack() so that getAttackStrengthScale() returns an accurate value during all attack events
             }
         }
     }
@@ -1145,7 +_,7 @@
         if (this.level() instanceof ServerLevel serverlevel) {
             float f = 1.0F + (float)this.getAttributeValue(Attributes.SWEEPING_DAMAGE_RATIO) * baseDamage;
 
-            for (LivingEntity livingentity : this.level().getEntitiesOfClass(LivingEntity.class, entity.getBoundingBox().inflate(1.0, 0.25, 1.0))) {
+            for (LivingEntity livingentity : this.level().getEntitiesOfClass(LivingEntity.class, this.getItemInHand(InteractionHand.MAIN_HAND).getSweepHitBox(this, entity))) {
                 if (livingentity != this
                     && livingentity != entity
                     && !this.isAlliedTo(livingentity)
@@ -1311,6 +_,7 @@
     }
 
     public void stopSleepInBed(final boolean forcefulWakeUp, final boolean updateLevelList) {
+        net.minecraftforge.event.ForgeEventFactory.onPlayerWakeup(this, forcefulWakeUp, updateLevelList);
         super.stopSleeping();
         if (this.level() instanceof ServerLevel && updateLevelList) {
             ((ServerLevel)this.level()).updateSleepingPlayerList();
@@ -1420,6 +_,7 @@
     @Override
     public boolean causeFallDamage(final double fallDistance, final float damageModifier, final DamageSource damageSource) {
         if (this.abilities.mayfly) {
+            net.minecraftforge.event.ForgeEventFactory.onPlayerFall(this, fallDistance, damageModifier);
             return false;
         } else {
             if (fallDistance >= 2.0) {
@@ -1454,13 +_,13 @@
     protected void playStepSound(final BlockPos onPos, final BlockState onState) {
         if (this.isInWater()) {
             this.waterSwimSound();
-            this.playMuffledStepSound(onState);
+            this.playMuffledStepSound(onState, onPos);
         } else {
             BlockPos blockpos = this.getPrimaryStepSoundBlockPos(onPos);
             if (!onPos.equals(blockpos)) {
                 BlockState blockstate = this.level().getBlockState(blockpos);
                 if (blockstate.is(BlockTags.COMBINATION_STEP_SOUND_BLOCKS)) {
-                    this.playCombinationStepSounds(blockstate, onState);
+                    this.playCombinationStepSounds(blockstate, onState, blockpos, onPos);
                 } else {
                     super.playStepSound(blockpos, blockstate);
                 }
@@ -1490,7 +_,13 @@
         this.tryResetCurrentImpulseContext();
     }
 
-    public void giveExperiencePoints(final int i) {
+    public void giveExperiencePoints(final int points) {
+        var event = net.minecraftforge.event.ForgeEventFactory.onPlayerXpChange(this, points);
+        if (event == null) {
+            return;
+        }
+        int i = event.getAmount();
+
         this.increaseScore(i);
         this.experienceProgress = this.experienceProgress + (float)i / this.getXpNeededForNextLevel();
         this.totalExperience = Mth.clamp(this.totalExperience + i, 0, Integer.MAX_VALUE);
@@ -1518,7 +_,7 @@
     }
 
     public void onEnchantmentPerformed(final ItemStack itemStack, final int enchantmentCost) {
-        this.experienceLevel -= enchantmentCost;
+        giveExperienceLevels(-enchantmentCost);
         if (this.experienceLevel < 0) {
             this.experienceLevel = 0;
             this.experienceProgress = 0.0F;
@@ -1528,7 +_,13 @@
         this.enchantmentSeed = this.random.nextInt();
     }
 
-    public void giveExperienceLevels(final int amount) {
+    public void giveExperienceLevels(final int points) {
+        var event = net.minecraftforge.event.ForgeEventFactory.onPlayerLevelChange(this, points);
+        if (event == null) {
+            return;
+        }
+        int amount = event.getLevels();
+
         this.experienceLevel = IntMath.saturatedAdd(this.experienceLevel, amount);
         if (this.experienceLevel < 0) {
             this.experienceLevel = 0;
@@ -1667,7 +_,13 @@
 
     @Override
     public Component getDisplayName() {
-        MutableComponent mutablecomponent = PlayerTeam.formatNameForTeam(this.getTeam(), this.getName());
+        if (this.displayname == null) {
+            this.displayname = net.minecraftforge.event.ForgeEventFactory.getPlayerDisplayName(this, this.getName());
+        }
+        MutableComponent mutablecomponent = Component.literal("");
+        mutablecomponent = prefixes.stream().reduce(mutablecomponent, MutableComponent::append);
+        mutablecomponent = mutablecomponent.append(PlayerTeam.formatNameForTeam(this.getTeam(), this.displayname));
+        mutablecomponent = suffixes.stream().reduce(mutablecomponent, MutableComponent::append);
         return this.decorateDisplayNameComponent(mutablecomponent);
     }
 
@@ -1857,18 +_,19 @@
             Predicate<ItemStack> predicate = ((ProjectileWeaponItem)heldWeapon.getItem()).getSupportedHeldProjectiles();
             ItemStack itemstack = ProjectileWeaponItem.getHeldProjectile(this, predicate);
             if (!itemstack.isEmpty()) {
-                return itemstack;
+                return net.minecraftforge.common.ForgeHooks.getProjectile(this, heldWeapon, itemstack);
             } else {
                 predicate = ((ProjectileWeaponItem)heldWeapon.getItem()).getAllSupportedProjectiles();
 
                 for (int i = 0; i < this.inventory.getContainerSize(); i++) {
                     ItemStack itemstack1 = this.inventory.getItem(i);
                     if (predicate.test(itemstack1)) {
-                        return itemstack1;
+                        return net.minecraftforge.common.ForgeHooks.getProjectile(this, heldWeapon, itemstack1);
                     }
                 }
 
-                return this.hasInfiniteMaterials() ? new ItemStack(Items.ARROW) : ItemStack.EMPTY;
+                var vanilla = this.abilities.instabuild ? new ItemStack(Items.ARROW) : ItemStack.EMPTY;
+                return net.minecraftforge.common.ForgeHooks.getProjectile(this, heldWeapon, vanilla);
             }
         }
     }
@@ -1989,6 +_,54 @@
         double d0 = this.blockInteractionRange() + buffer;
         return new AABB(pos).distanceToSqr(this.getEyePosition()) < d0 * d0;
     }
+
+    public Collection<MutableComponent> getPrefixes() {
+        return this.prefixes;
+    }
+
+    public Collection<MutableComponent> getSuffixes() {
+        return this.suffixes;
+    }
+
+    private Component displayname = null;
+
+    /**
+     * Force the displayed name to refresh, by firing {@link net.minecraftforge.event.entity.player.PlayerEvent.NameFormat}, using the real player name as event parameter.
+     */
+    public void refreshDisplayName() {
+        this.displayname = net.minecraftforge.event.ForgeEventFactory.getPlayerDisplayName(this, this.getName());
+    }
+
+    private final net.minecraftforge.common.util.LazyOptional<net.minecraftforge.items.IItemHandler>
+        playerMainHandler, playerEquipmentHandler, playerJoinedHandler;
+
+    @Override
+    public <T> net.minecraftforge.common.util.LazyOptional<T> getCapability(net.minecraftforge.common.capabilities.Capability<T> capability, @Nullable Direction facing) {
+        if (capability == net.minecraftforge.common.capabilities.ForgeCapabilities.ITEM_HANDLER && this.isAlive()) {
+            if (facing == null) return playerJoinedHandler.cast();
+            else if (facing.getAxis().isVertical()) return playerMainHandler.cast();
+            else if (facing.getAxis().isHorizontal()) return playerEquipmentHandler.cast();
+        }
+        return super.getCapability(capability, facing);
+    }
+
+    /**
+     * Force a pose for the player. If set, the vanilla pose determination and clearance check is skipped. Make sure the pose is clear yourself (e.g. in PlayerTick).
+     * This has to be set just once, do not set it every tick.
+     * Make sure to clear (null) the pose if not required anymore and only use if necessary.
+     */
+    public void setForcedPose(@Nullable Pose pose) {
+        this.forcedPose = pose;
+    }
+
+    /**
+     * @return The forced pose if set, null otherwise
+     */
+    @Nullable
+    public Pose getForcedPose() {
+        return this.forcedPose;
+    }
+
 
     public boolean shouldRotateWithMinecart() {
         return false;
