--- a/net/minecraft/world/entity/npc/villager/VillagerTrades.java
+++ b/net/minecraft/world/entity/npc/villager/VillagerTrades.java
@@ -854,7 +_,7 @@
                 5
             )
         )
-        .build();
+        .build().stream().collect(net.minecraft.util.Util.toMutableList()); // Forge: Turned into an ArrayList so it can be modified in VillagerTradingManager
     public static final Map<ResourceKey<VillagerProfession>, Int2ObjectMap<VillagerTrades.ItemListing[]>> EXPERIMENTAL_TRADES = Map.of(
         VillagerProfession.LIBRARIAN,
         toIntMap(
@@ -1279,6 +_,7 @@
         private final int villagerXp;
 
         public EmeraldsForVillagerTypeItem(int p_453689_, int p_451560_, int p_457573_, Map<ResourceKey<VillagerType>, Item> p_450639_) {
+            if (false) // FORGE: Modders can add custom villager types, so remove this validation
             BuiltInRegistries.VILLAGER_TYPE.registryKeySet().stream().filter(p_455509_ -> !p_450639_.containsKey(p_455509_)).findAny().ifPresent(p_460167_ -> {
                 throw new IllegalStateException("Missing trade for villager type: " + p_460167_);
             });
@@ -1295,7 +_,9 @@
                 if (resourcekey == null) {
                     return null;
                 } else {
-                    ItemCost itemcost = new ItemCost(this.trades.get(resourcekey), this.cost);
+                    Item item = this.trades.get(resourcekey);
+                    if (item == null) return null; // FORGE: Account for modded villager types by returning null if there is no trade
+                    ItemCost itemcost = new ItemCost(item, this.cost);
                     return new MerchantOffer(itemcost, new ItemStack(Items.EMERALD), this.maxUses, this.villagerXp, 0.05F);
                 }
             } else {
