--- a/net/minecraft/world/entity/vehicle/minecart/MinecartSpawner.java
+++ b/net/minecraft/world/entity/vehicle/minecart/MinecartSpawner.java
@@ -19,6 +_,11 @@
         public void broadcastEvent(Level p_451201_, BlockPos p_454356_, int p_454162_) {
             p_451201_.broadcastEntityEvent(MinecartSpawner.this, (byte)p_454162_);
         }
+
+        @Override
+        public net.minecraft.world.entity.@org.jspecify.annotations.Nullable Entity getSpawnerEntity() {
+            return MinecartSpawner.this;
+        }
     };
     private final Runnable ticker;
 
