--- a/net/minecraft/world/entity/vehicle/minecart/MinecartCommandBlock.java
+++ b/net/minecraft/world/entity/vehicle/minecart/MinecartCommandBlock.java
@@ -84,6 +_,8 @@
 
     @Override
     public InteractionResult interact(Player p_459569_, InteractionHand p_455676_) {
+        InteractionResult ret = super.interact(p_459569_, p_455676_);
+        if (ret.consumesAction()) return ret;
         if (!p_459569_.canUseGameMasterBlocks()) {
             return InteractionResult.PASS;
         } else {
