--- a/net/minecraft/world/entity/vehicle/minecart/NewMinecartBehavior.java
+++ b/net/minecraft/world/entity/vehicle/minecart/NewMinecartBehavior.java
@@ -169,7 +_,7 @@
 
     public void adjustToRails(final BlockPos targetBlockPos, final BlockState currentState, final boolean instant) {
         if (BaseRailBlock.isRail(currentState)) {
-            RailShape railshape = currentState.getValue(((BaseRailBlock)currentState.getBlock()).getShapeProperty());
+            RailShape railshape = ((BaseRailBlock)currentState.getBlock()).getRailDirection(currentState, this.level(), targetBlockPos, this.minecart);
             Pair<Vec3i, Vec3i> pair = AbstractMinecart.exits(railshape);
             Vec3 vec3 = new Vec3(pair.getFirst()).scale(0.5);
             Vec3 vec31 = new Vec3(pair.getSecond()).scale(0.5);
@@ -257,11 +_,11 @@
             if (flag) {
                 this.minecart.resetFallDistance();
                 this.minecart.setOldPosAndRot();
-                if (blockstate.is(Blocks.ACTIVATOR_RAIL)) {
+                if (blockstate.getBlock() instanceof PoweredRailBlock power && power.isActivatorRail()) {
                     this.minecart.activateMinecart(level, blockpos.getX(), blockpos.getY(), blockpos.getZ(), blockstate.getValue(PoweredRailBlock.POWERED));
                 }
 
-                RailShape railshape = blockstate.getValue(((BaseRailBlock)blockstate.getBlock()).getShapeProperty());
+                RailShape railshape = ((BaseRailBlock)blockstate.getBlock()).getRailDirection(blockstate, this.level(), blockpos, this.minecart);
                 Vec3 vec31 = this.calculateTrackSpeed(level, vec3.horizontal(), newminecartbehavior$trackiteration, blockpos, blockstate, railshape);
                 if (newminecartbehavior$trackiteration.firstIteration) {
                     newminecartbehavior$trackiteration.movementLeft = vec31.horizontalDistance();
@@ -397,7 +_,7 @@
     }
 
     private Vec3 calculateHaltTrackSpeed(final Vec3 deltaMovement, final BlockState state) {
-        if (state.is(Blocks.POWERED_RAIL) && !state.getValue(PoweredRailBlock.POWERED)) {
+        if (state.getBlock() instanceof PoweredRailBlock powered && !powered.isActivatorRail() && !state.getValue(PoweredRailBlock.POWERED)) {
             return deltaMovement.length() < 0.03 ? Vec3.ZERO : deltaMovement.scale(0.5);
         } else {
             return deltaMovement;
@@ -405,7 +_,7 @@
     }
 
     private Vec3 calculateBoostTrackSpeed(final Vec3 deltaMovement, final BlockPos pos, final BlockState state) {
-        if (state.is(Blocks.POWERED_RAIL) && state.getValue(PoweredRailBlock.POWERED)) {
+        if (state.getBlock() instanceof PoweredRailBlock powered && !powered.isActivatorRail() && state.getValue(PoweredRailBlock.POWERED)) {
             if (deltaMovement.length() > 0.01) {
                 return deltaMovement.normalize().scale(deltaMovement.length() + 0.06);
             } else {
@@ -457,7 +_,7 @@
                 BlockState blockstate = this.level().getBlockState(BlockPos.containing(vec36));
                 if (flag) {
                     if (BaseRailBlock.isRail(blockstate)) {
-                        RailShape railshape = blockstate.getValue(((BaseRailBlock)blockstate.getBlock()).getShapeProperty());
+                        RailShape railshape = ((BaseRailBlock)blockstate.getBlock()).getRailDirection(blockstate, this.level(), BlockPos.containing(vec36), this.minecart);
                         if (this.restAtVShape(shape, railshape)) {
                             return 0.0;
                         }
