--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -127,6 +_,7 @@
     private int simulationDistance;
     private boolean allowCommandsForAllPlayers;
     private int sendAllPlayerInfoIn;
+    private final List<ServerPlayer> playersView = java.util.Collections.unmodifiableList(players);
 
     public PlayerList(
         final MinecraftServer server,
@@ -184,6 +_,7 @@
         servergamepacketlistenerimpl.send(new ClientboundPlayerAbilitiesPacket(player.getAbilities()));
         servergamepacketlistenerimpl.send(new ClientboundSetHeldSlotPacket(player.getInventory().getSelectedSlot()));
         RecipeManager recipemanager = this.server.getRecipeManager();
+        net.minecraftforge.event.OnDatapackSyncEvent.BUS.post(new net.minecraftforge.event.OnDatapackSyncEvent(this, player));
         servergamepacketlistenerimpl.send(
             new ClientboundUpdateRecipesPacket(recipemanager.getSynchronizedItemProperties(), recipemanager.getSynchronizedStonecutterRecipes())
         );
@@ -217,6 +_,7 @@
         player.initInventoryMenu();
         this.server.notificationManager().playerJoined(player);
         servergamepacketlistenerimpl.resumeFlushing();
+        net.minecraftforge.event.ForgeEventFactory.firePlayerLoggedIn(player);
     }
 
     protected void updateEntireScoreboard(final ServerScoreboard scoreboard, final ServerPlayer player) {
@@ -290,6 +_,7 @@
     }
 
     protected void save(final ServerPlayer player) {
+        if (player.connection == null) return; // Not sure if still needed -Paint_Ninja
         this.playerIo.save(player);
         ServerStatsCounter serverstatscounter = this.stats.get(player.getUUID());
         if (serverstatscounter != null) {
@@ -303,6 +_,7 @@
     }
 
     public void remove(final ServerPlayer player) {
+        net.minecraftforge.event.ForgeEventFactory.firePlayerLoggedOut(player);
         ServerLevel serverlevel = player.level();
         player.awardStat(Stats.LEAVE_GAME);
         this.save(player);
@@ -429,6 +_,7 @@
         this.playersByUUID.put(serverplayer.getUUID(), serverplayer);
         serverplayer.initInventoryMenu();
         serverplayer.setHealth(serverplayer.getHealth());
+        net.minecraftforge.event.ForgeEventFactory.firePlayerRespawnEvent(serverplayer, keepAllPlayerData);
         ServerPlayer.RespawnConfig serverplayer$respawnconfig = serverplayer.getRespawnConfig();
         if (!keepAllPlayerData && serverplayer$respawnconfig != null) {
             LevelData.RespawnData leveldata$respawndata = serverplayer$respawnconfig.respawnData();
@@ -542,6 +_,7 @@
     }
 
     public void op(final NameAndId nameAndId, final Optional<LevelBasedPermissionSet> permissions, final Optional<Boolean> canBypassPlayerLimit) {
+        if (net.minecraftforge.event.ForgeEventFactory.onPermissionChanged(nameAndId, permissions.orElse(this.server.operatorUserPermissions()), this)) return;
         this.ops
             .add(
                 new ServerOpListEntry(
@@ -555,6 +_,7 @@
     }
 
     public void deop(final NameAndId nameAndId) {
+        if (net.minecraftforge.event.ForgeEventFactory.onPermissionChanged(nameAndId, null, this)) return;
         if (this.ops.remove(nameAndId)) {
             ServerPlayer serverplayer = this.getPlayer(nameAndId.id());
             if (serverplayer != null) {
@@ -824,7 +_,7 @@
     }
 
     public List<ServerPlayer> getPlayers() {
-        return this.players;
+        return this.playersView; //Unmodifiable view, we don't want people removing things without us knowing.
     }
 
     public @Nullable ServerPlayer getPlayer(final UUID uuid) {
@@ -850,6 +_,7 @@
             playeradvancements.reload(this.server.getAdvancements());
         }
 
+        net.minecraftforge.event.OnDatapackSyncEvent.BUS.post(new net.minecraftforge.event.OnDatapackSyncEvent(this, null));
         this.broadcastAll(new ClientboundUpdateTagsPacket(TagNetworkSerialization.serializeTagsToNetwork(this.registries)));
         RecipeManager recipemanager = this.server.getRecipeManager();
         ClientboundUpdateRecipesPacket clientboundupdaterecipespacket = new ClientboundUpdateRecipesPacket(
@@ -864,5 +_,9 @@
 
     public boolean isAllowCommandsForAllPlayers() {
         return this.allowCommandsForAllPlayers;
+    }
+
+    public PlayerDataStorage getPlayerIo() {
+        return this.playerIo;
     }
 }
