--- a/net/minecraft/world/level/storage/loot/LootPool.java
+++ b/net/minecraft/world/level/storage/loot/LootPool.java
@@ -33,15 +_,20 @@
                 LootItemCondition.CODEC.optionalFieldOf("condition").forGetter(p -> p.condition),
                 LootItemFunctions.CODEC.optionalFieldOf("modifier").forGetter(p -> p.modifier),
                 ContextIntProviders.CODEC.fieldOf("rolls").forGetter(p -> p.rolls),
-                ContextFloatProviders.CODEC.optionalFieldOf("bonus_rolls", ContextFloatProviders.exactly(0.0F)).forGetter(p -> p.bonusRolls)
+                ContextFloatProviders.CODEC.optionalFieldOf("bonus_rolls", ContextFloatProviders.exactly(0.0F)).forGetter(p -> p.bonusRolls),
+                Codec.STRING.optionalFieldOf("name").forGetter(p -> p.name.filter(n -> !n.startsWith("custom#"))),
+                net.minecraftforge.common.crafting.conditions.ICondition.OPTIONAL_FEILD_CODEC.forGetter(p -> p.forge_condition)
             )
             .apply(i, LootPool::new)
     );
+    public static final Codec<LootPool> CONDITIONAL_CODEC = net.minecraftforge.common.crafting.conditions.ConditionCodec.checkingDecode(CODEC, () -> lootPool().build());
     private final List<LootPoolEntryContainer> entries;
     private final Optional<Holder<LootItemCondition>> condition;
     private final Optional<Holder<LootItemFunction>> modifier;
     private Holder<ContextIntProvider> rolls;
     private Holder<ContextFloatProvider> bonusRolls;
+    private java.util.Optional<String> name;
+    private java.util.Optional<net.minecraftforge.common.crafting.conditions.ICondition> forge_condition;
 
     private LootPool(
         final List<LootPoolEntryContainer> entries,
@@ -50,11 +_,25 @@
         final Holder<ContextIntProvider> rolls,
         final Holder<ContextFloatProvider> bonusRolls
     ) {
+        this(entries, condition, modifier, rolls, bonusRolls, java.util.Optional.empty(), java.util.Optional.empty());
+    }
+
+    private LootPool(
+        final List<LootPoolEntryContainer> entries,
+        final Optional<Holder<LootItemCondition>> condition,
+        final Optional<Holder<LootItemFunction>> modifier,
+        final Holder<ContextIntProvider> rolls,
+        final Holder<ContextFloatProvider> bonusRolls,
+        final java.util.Optional<String> name,
+        final java.util.Optional<net.minecraftforge.common.crafting.conditions.ICondition> forge_condition
+    ) {
         this.entries = entries;
         this.condition = condition;
         this.modifier = modifier;
         this.rolls = rolls;
         this.bonusRolls = bonusRolls;
+        this.name = name;
+        this.forge_condition = forge_condition;
     }
 
     private void addRandomItem(final Consumer<ItemStack> result, final LootContext context) {
@@ -114,18 +_,51 @@
         return new LootPool.Builder();
     }
 
+    private boolean isFrozen = false;
+    public void freeze() { this.isFrozen = true; }
+    public boolean isFrozen(){ return this.isFrozen; }
+    private void checkFrozen() {
+       if (this.isFrozen())
+          throw new RuntimeException("Attempted to modify LootPool after being frozen!");
+    }
+    @org.jetbrains.annotations.Nullable
+    public String getName() { return this.name.orElse(null); }
+    void setName(final String name) {
+       if (this.name.isPresent())
+          throw new UnsupportedOperationException("Cannot change the name of a pool when it has a name set!");
+       this.name = java.util.Optional.of(name);
+    }
+    public Holder<ContextIntProvider> getRolls()      { return this.rolls; }
+    public Holder<ContextFloatProvider> getBonusRolls() { return this.bonusRolls; }
+    public void setRolls     (Holder<ContextIntProvider> v){ checkFrozen(); this.rolls = v; }
+    public void setBonusRolls(Holder<ContextFloatProvider> v){ checkFrozen(); this.bonusRolls = v; }
+
     public static class Builder implements FunctionUserBuilder<LootPool.Builder>, ConditionUserBuilder<LootPool.Builder> {
         private final ImmutableList.Builder<LootPoolEntryContainer> entries = ImmutableList.builder();
         private final ImmutableList.Builder<Holder<LootItemCondition>> conditions = ImmutableList.builder();
         private final ImmutableList.Builder<Holder<LootItemFunction>> functions = ImmutableList.builder();
         private Holder<ContextIntProvider> rolls = ContextIntProviders.exactly(1);
         private Holder<ContextFloatProvider> bonusRolls = ContextFloatProviders.exactly(0.0F);
+        @org.jetbrains.annotations.Nullable
+        private String name;
+        @org.jetbrains.annotations.Nullable
+        private net.minecraftforge.common.crafting.conditions.ICondition forge_condition;
 
         public LootPool.Builder setRolls(final Holder<ContextIntProvider> rolls) {
             this.rolls = rolls;
             return this;
         }
 
+        public LootPool.Builder name(String name) {
+            this.name = name;
+            return this;
+        }
+
+        public LootPool.Builder when(net.minecraftforge.common.crafting.conditions.ICondition value) {
+            this.forge_condition = value;
+            return this;
+        }
+
         public LootPool.Builder unwrap() {
             return this;
         }
@@ -164,7 +_,9 @@
                 ConditionUserBuilder.buildCondition(this.conditions.build()),
                 FunctionUserBuilder.buildFunction(this.functions.build()),
                 this.rolls,
-                this.bonusRolls
+                this.bonusRolls,
+                java.util.Optional.ofNullable(this.name),
+                java.util.Optional.ofNullable(forge_condition)
             );
         }
     }
