--- a/net/minecraft/world/level/storage/loot/LootContext.java
+++ b/net/minecraft/world/level/storage/loot/LootContext.java
@@ -26,11 +_,17 @@
     private final RandomSource random;
     private final HolderGetter.Provider lootDataResolver;
     private final Set<LootContext.VisitedEntry<?>> visitedElements = Sets.newLinkedHashSet();
+    private Identifier queriedLootTableId;
 
     private LootContext(final LootParams params, final RandomSource random, final HolderGetter.Provider lootDataResolver) {
+        this(params, random, lootDataResolver, null);
+    }
+
+    LootContext(final LootParams params, final RandomSource random, final HolderGetter.Provider lootDataResolver, final Identifier queriedLootTableId) {
         this.params = params;
         this.random = random;
         this.lootDataResolver = lootDataResolver;
+        this.queriedLootTableId = queriedLootTableId;
     }
 
     public boolean hasParameter(final ContextKey<?> key) {
@@ -73,6 +_,10 @@
         return this.params.getLevel();
     }
 
+    public LootParams getParams() {
+        return this.params;
+    }
+
     public static LootContext.VisitedEntry<LootTable> createVisitedEntry(final LootTable table) {
         return new LootContext.VisitedEntry<>(LootDataType.TABLE, table);
     }
@@ -89,6 +_,18 @@
         return new LootContext.VisitedEntry<>(LootDataType.SLOT_SOURCE, slotSource);
     }
 
+    public int getLootingModifier() {
+        return net.minecraftforge.common.ForgeHooks.getLootingLevel(getOptional(LootContextParams.THIS_ENTITY), getOptional(LootContextParams.ATTACKING_ENTITY), getOptional(LootContextParams.DAMAGE_SOURCE));
+    }
+
+    public void setQueriedLootTableId(Identifier queriedLootTableId) {
+        if (this.queriedLootTableId == null && queriedLootTableId != null) this.queriedLootTableId = queriedLootTableId;
+    }
+
+    public Identifier getQueriedLootTableId() {
+        return this.queriedLootTableId == null ? net.minecraftforge.common.loot.LootTableIdCondition.UNKNOWN_LOOT_TABLE : this.queriedLootTableId;
+    }
+
     public enum BlockEntityTarget implements StringRepresentable, LootContextArg.SimpleGetter<BlockEntity> {
         BLOCK_ENTITY("block_entity", LootContextParams.BLOCK_ENTITY);
 
@@ -114,11 +_,18 @@
     public static class Builder {
         private final LootParams params;
         private @Nullable RandomSource random;
+        private Identifier queriedLootTableId; // Forge: correctly pass around loot table ID with copy constructor
 
         public Builder(final LootParams params) {
             this.params = params;
         }
 
+        public Builder(LootContext context) {
+            this.params = context.params;
+            this.random = context.random;
+            this.queriedLootTableId = context.queriedLootTableId;
+        }
+
         public LootContext.Builder withOptionalRandomSeed(final long seed) {
             if (seed != 0L) {
                 this.random = RandomSource.create(seed);
@@ -132,6 +_,11 @@
             return this;
         }
 
+        public LootContext.Builder withQueriedLootTableId(Identifier queriedLootTableId) {
+            this.queriedLootTableId = queriedLootTableId;
+            return this;
+        }
+
         public ServerLevel getLevel() {
             return this.params.getLevel();
         }
@@ -140,7 +_,7 @@
             ServerLevel level = this.getLevel();
             MinecraftServer server = level.getServer();
             RandomSource random = Optional.ofNullable(this.random).or(() -> randomSequenceKey.map(server::getRandomSequence)).orElseGet(level::getRandom);
-            return new LootContext(this.params, random, server.reloadableRegistries().lookup());
+            return new LootContext(this.params, random, server.reloadableRegistries().lookup(), queriedLootTableId);
         }
     }
 
