--- a/net/minecraft/client/renderer/rendertype/RenderTypes.java
+++ b/net/minecraft/client/renderer/rendertype/RenderTypes.java
@@ -574,7 +_,7 @@
     }
 
     public static RenderType text(final Identifier texture) {
-        return TEXT.apply(texture);
+        return net.minecraftforge.client.ForgeRenderTypes.getText(texture);
     }
 
     public static RenderType textBackground() {
@@ -582,19 +_,19 @@
     }
 
     public static RenderType textIntensity(final Identifier texture) {
-        return TEXT_INTENSITY.apply(texture);
+        return net.minecraftforge.client.ForgeRenderTypes.getTextIntensity(texture);
     }
 
     public static RenderType textPolygonOffset(final Identifier texture) {
-        return TEXT_POLYGON_OFFSET.apply(texture);
+        return net.minecraftforge.client.ForgeRenderTypes.getTextPolygonOffset(texture);
     }
 
     public static RenderType textIntensityPolygonOffset(final Identifier texture) {
-        return TEXT_INTENSITY_POLYGON_OFFSET.apply(texture);
+        return net.minecraftforge.client.ForgeRenderTypes.getTextIntensityPolygonOffset(texture);
     }
 
     public static RenderType textSeeThrough(final Identifier texture) {
-        return TEXT_SEE_THROUGH.apply(texture);
+        return net.minecraftforge.client.ForgeRenderTypes.getTextSeeThrough(texture);
     }
 
     public static RenderType textBackgroundSeeThrough() {
@@ -602,7 +_,7 @@
     }
 
     public static RenderType textIntensitySeeThrough(final Identifier texture) {
-        return TEXT_INTENSITY_SEE_THROUGH.apply(texture);
+        return net.minecraftforge.client.ForgeRenderTypes.getTextIntensitySeeThrough(texture);
     }
 
     public static RenderType lightning() {
