--- a/com/mojang/blaze3d/vertex/VertexFormat.java
+++ b/com/mojang/blaze3d/vertex/VertexFormat.java
@@ -29,6 +_,7 @@
     private final int[] offsetsByElement = new int[32];
     private @Nullable GpuBuffer immediateDrawVertexBuffer;
     private @Nullable GpuBuffer immediateDrawIndexBuffer;
+    private final com.google.common.collect.ImmutableMap<String, VertexFormatElement> elementMapping;
 
     private VertexFormat(final List<VertexFormatElement> elements, final List<String> names, final IntList offsets, final int vertexSize) {
         this.elements = elements;
@@ -41,6 +_,11 @@
             int j = vertexformatelement != null ? elements.indexOf(vertexformatelement) : -1;
             this.offsetsByElement[i] = j != -1 ? offsets.getInt(j) : -1;
         }
+
+        ImmutableMap.Builder<String, VertexFormatElement> elementMapping = ImmutableMap.builder();
+        for (int i = 0; i < elements.size(); i++)
+            elementMapping.put(names.get(i), elements.get(i));
+        this.elementMapping = elementMapping.buildOrThrow();
     }
 
     public static VertexFormat.Builder builder() {
@@ -139,6 +_,9 @@
         this.immediateDrawIndexBuffer = uploadToBuffer(this.immediateDrawIndexBuffer, buffer, 72, () -> "Immediate index buffer for " + this);
         return this.immediateDrawIndexBuffer;
     }
+
+    public ImmutableMap<String, VertexFormatElement> getElementMapping() { return elementMapping; }
+    public int getOffset(int index) { return offsetsByElement[index]; }
 
     @OnlyIn(Dist.CLIENT)
     public static class Builder {
