--- a/net/minecraft/core/RegistrySetBuilder.java
+++ b/net/minecraft/core/RegistrySetBuilder.java
@@ -23,7 +_,8 @@
 import org.apache.commons.lang3.mutable.MutableObject;
 import org.jspecify.annotations.Nullable;
 
-public class RegistrySetBuilder {
+public class RegistrySetBuilder implements net.minecraftforge.common.extensions.IForgeRegistrySetBuilder {
+    private final it.unimi.dsi.fastutil.objects.Object2IntMap<ResourceKey<? extends Registry<?>>> indexes = new it.unimi.dsi.fastutil.objects.Object2IntLinkedOpenHashMap<>();
     private final List<RegistrySetBuilder.RegistryStub<?>> entries = new ArrayList<>();
 
     private static <T> HolderGetter<T> wrapContextLookup(final HolderLookup.RegistryLookup<T> original) {
@@ -67,14 +_,43 @@
     public <T> RegistrySetBuilder add(
         final ResourceKey<? extends Registry<T>> key, final Lifecycle lifecycle, final RegistrySetBuilder.RegistryBootstrap<T> bootstrap
     ) {
-        this.entries.add(new RegistrySetBuilder.RegistryStub<>(key, lifecycle, bootstrap));
+        add(new RegistrySetBuilder.RegistryStub<>(key, lifecycle, bootstrap));
         return this;
     }
 
+    // Forge: This merges two entries in this builder together, so that we can reuse vanilla boostrap methods, or potentially other modder's methods.
+    private <T> RegistrySetBuilder.RegistryStub<T> add(RegistrySetBuilder.RegistryStub<T> entry) {
+        int idx = indexes.getOrDefault(entry.key(), -1);
+        if (idx != -1) {
+            var _oldB = ((RegistrySetBuilder.RegistryStub<T>)this.entries.get(idx)).bootstrap();
+            var _newB = entry.bootstrap();
+            var _new = new RegistrySetBuilder.RegistryStub<>(entry.key(), entry.lifecycle(), ctx -> {
+                _oldB.run(ctx);
+                _newB.run(ctx);
+            });
+            this.entries.set(idx, _new);
+            return _new;
+        } else {
+            this.indexes.put(entry.key(), this.entries.size());
+            this.entries.add(entry);
+            return entry;
+        }
+    }
+
+    public RegistrySetBuilder copy() {
+        var ret = new RegistrySetBuilder();
+        this.entries.forEach(ret::add);
+        return ret;
+    }
+
     public <T> RegistrySetBuilder add(final ResourceKey<? extends Registry<T>> key, final RegistrySetBuilder.RegistryBootstrap<T> bootstrap) {
         return this.add(key, Lifecycle.stable(), bootstrap);
     }
 
+    public List<? extends ResourceKey<? extends Registry<?>>> getEntryKeys() {
+        return this.entries.stream().map(RegistrySetBuilder.RegistryStub::key).toList();
+    }
+
     private RegistrySetBuilder.BuildState createState(final RegistryAccess context) {
         RegistrySetBuilder.BuildState registrysetbuilder$buildstate = RegistrySetBuilder.BuildState.create(
             context, this.entries.stream().map(RegistrySetBuilder.RegistryStub::key)
@@ -183,14 +_,16 @@
             throw new NullPointerException("No cloner for " + registryKey.identifier());
         } else {
             Map<ResourceKey<T>, Holder.Reference<T>> map = new HashMap<>();
-            HolderLookup.RegistryLookup<T> registrylookup = patchProvider.lookupOrThrow(registryKey);
+            HolderLookup.RegistryLookup<T> registrylookup = patchProvider.lookup(registryKey).orElse(null);
             registrylookup.listElements().forEach(elementHolder -> {
                 ResourceKey<T> resourcekey = elementHolder.key();
                 RegistrySetBuilder.LazyHolder<T> lazyholder = new RegistrySetBuilder.LazyHolder<>(owner, resourcekey);
                 lazyholder.supplier = () -> cloner.clone((T)elementHolder.value(), patchProvider, targetProvider.get());
                 map.put(resourcekey, lazyholder);
             });
-            HolderLookup.RegistryLookup<T> registrylookup1 = fallbackProvider.lookupOrThrow(registryKey);
+            HolderLookup.RegistryLookup<T> registrylookup1 = fallbackProvider.lookup(registryKey).orElse(null);
+            Lifecycle lifecycle = registrylookup.registryLifecycle();
+            if (registrylookup1 != null) {
             registrylookup1.listElements().forEach(elementHolder -> {
                 ResourceKey<T> resourcekey = elementHolder.key();
                 map.computeIfAbsent(resourcekey, key -> {
@@ -199,7 +_,8 @@
                     return lazyholder;
                 });
             });
-            Lifecycle lifecycle = registrylookup.registryLifecycle().add(registrylookup1.registryLifecycle());
+            lifecycle = registrylookup.registryLifecycle().add(registrylookup1.registryLifecycle());
+            }
             return lookupFromMap(registryKey, lifecycle, owner, map);
         }
     }
@@ -275,6 +_,11 @@
                 @Override
                 public <S> HolderGetter<S> lookup(final ResourceKey<? extends Registry<? extends S>> key) {
                     return (HolderGetter<S>)BuildState.this.registries.getOrDefault(key.identifier(), BuildState.this.lookup);
+                }
+
+                @Override
+                public <S> Optional<HolderLookup.RegistryLookup<S>> registryLookup(ResourceKey<? extends Registry<? extends S>> registry) {
+                   return Optional.ofNullable((HolderLookup.RegistryLookup<S>) BuildState.this.registries.get(registry.identifier()));
                 }
             };
         }
