--- a/net/minecraft/world/entity/monster/piglin/Piglin.java
+++ b/net/minecraft/world/entity/monster/piglin/Piglin.java
@@ -322,7 +_,7 @@
         } else if (this.isChargingCrossbow()) {
             return PiglinArmPose.CROSSBOW_CHARGE;
         } else {
-            return this.isHolding(Items.CROSSBOW) && CrossbowItem.isCharged(this.getWeaponItem()) ? PiglinArmPose.CROSSBOW_HOLD : PiglinArmPose.DEFAULT;
+            return this.isHolding(is -> is.getItem() instanceof net.minecraft.world.item.CrossbowItem) && CrossbowItem.isCharged(this.getWeaponItem()) ? PiglinArmPose.CROSSBOW_HOLD : PiglinArmPose.DEFAULT;
         }
     }
 
@@ -359,7 +_,7 @@
     }
 
     protected void holdInOffHand(final ItemStack itemStack) {
-        if (itemStack.is(PiglinAi.BARTERING_ITEM)) {
+        if (itemStack.isPiglinCurrency()) {
             this.setItemSlot(EquipmentSlot.OFFHAND, itemStack);
             this.setGuaranteedDrop(EquipmentSlot.OFFHAND);
         } else {
@@ -369,7 +_,7 @@
 
     @Override
     public boolean wantsToPickUp(final ServerLevel level, final ItemStack itemStack) {
-        return level.getGameRules().get(GameRules.MOB_GRIEFING) && this.canPickUpLoot() && PiglinAi.wantsToPickup(this, itemStack);
+        return net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(level, this) && this.canPickUpLoot() && PiglinAi.wantsToPickup(this, itemStack);
     }
 
     protected boolean canReplaceCurrentItem(final ItemStack newItemStack) {
