--- a/net/minecraft/core/registries/BuiltInRegistries.java
+++ b/net/minecraft/core/registries/BuiltInRegistries.java
@@ -294,7 +_,9 @@
     ) {
         Bootstrap.checkBootstrapCalled(() -> "registry " + p_259230_.identifier());
         Identifier identifier = p_259230_.identifier();
-        LOADERS.put(identifier, () -> p_259210_.run(p_260327_));
+        var maybeWrapped = net.minecraftforge.registries.GameData.getWrapper(p_259230_, p_260327_);
+        p_260327_ = maybeWrapped;
+        LOADERS.put(identifier, () -> p_259210_.run(maybeWrapped));
         WRITABLE_REGISTRY.register((ResourceKey)p_259230_, p_260327_, RegistrationInfo.BUILT_IN);
         return p_260327_;
     }
@@ -330,7 +_,7 @@
 
             if (p_448623_ instanceof DefaultedRegistry) {
                 Identifier identifier = ((DefaultedRegistry)p_448623_).getDefaultKey();
-                Objects.requireNonNull(p_448623_.getValue(identifier), "Missing default of DefaultedMappedRegistry: " + identifier);
+                Objects.requireNonNull(p_448623_.getValue(identifier), "Missing default of DefaultedMappedRegistry: " + p_448623_.key() + " - " + identifier);
             }
         });
     }
