--- a/net/minecraft/world/entity/monster/zombie/Zombie.java
+++ b/net/minecraft/world/entity/monster/zombie/Zombie.java
@@ -245,17 +_,21 @@
     }
 
     protected void convertToZombieType(final ServerLevel level, final EntityType<? extends Zombie> zombieType) {
+        if (!net.minecraftforge.event.ForgeEventFactory.canLivingConvert(this, zombieType, timer -> this.conversionTime = timer)) return;
         this.convertTo(
             zombieType,
             ConversionParams.single(this, true, true),
-            newZombie -> newZombie.handleAttributes(
-                level.getCurrentDifficultyAt(newZombie.blockPosition()).getSpecialMultiplier(), EntitySpawnReason.CONVERSION
-            )
+            newZombie -> {
+                newZombie.handleAttributes(level.getCurrentDifficultyAt(newZombie.blockPosition()).getSpecialMultiplier(), EntitySpawnReason.CONVERSION);
+                net.minecraftforge.event.ForgeEventFactory.onLivingConvert(this, newZombie);
+            }
         );
     }
 
     @VisibleForTesting
     public boolean convertVillagerToZombieVillager(final ServerLevel level, final Villager villager) {
+        if (!net.minecraftforge.event.ForgeEventFactory.canLivingConvert(villager, EntityType.ZOMBIE_VILLAGER, (timer) -> {}))
+            return false;
         ZombieVillager zombievillager = villager.convertTo(
             EntityType.ZOMBIE_VILLAGER,
             ConversionParams.single(villager, true, true),
@@ -267,6 +_,7 @@
                 zombie.setGossips(villager.getGossips().copy());
                 zombie.setTradeOffers(villager.getOffers().copy());
                 zombie.setVillagerXp(villager.getVillagerXp());
+                net.minecraftforge.event.ForgeEventFactory.onLivingConvert(villager, zombie);
                 if (!this.isSilent()) {
                     level.levelEvent(null, 1026, this.blockPosition(), 0);
                 }
@@ -289,19 +_,26 @@
                 livingentity = (LivingEntity)source.getEntity();
             }
 
-            if (livingentity != null
+            var vanilla = (livingentity != null
                 && level.getDifficulty() == Difficulty.HARD
                 && this.random.nextFloat() < this.getAttributeValue(Attributes.SPAWN_REINFORCEMENTS_CHANCE)
-                && level.isSpawningMonsters()) {
+                && level.isSpawningMonsters());
                 int i = Mth.floor(this.getX());
                 int j = Mth.floor(this.getY());
                 int k = Mth.floor(this.getZ());
                 EntityType<? extends Zombie> entitytype = this.getType();
-                Zombie zombie = entitytype.create(level, EntitySpawnReason.REINFORCEMENT);
-                if (zombie == null) {
-                    return true;
-                }
-
+
+            var event = net.minecraftforge.event.ForgeEventFactory.fireZombieSummonAid(this, level(), i, j, k, livingentity, this.getAttributeValue(Attributes.SPAWN_REINFORCEMENTS_CHANCE));
+
+            Zombie zombie = null;
+            if (event.getResult().isAllowed() || (vanilla && event.getResult().isDefault())) {
+                if (event.getCustomSummonedAid() != null)
+                    zombie = event.getCustomSummonedAid();
+                else
+                    zombie = entitytype.create(this.level(), EntitySpawnReason.REINFORCEMENT);
+            }
+
+            if (zombie != null) {
                 for (int l = 0; l < 50; l++) {
                     int i1 = i + Mth.nextInt(this.random, 7, 40) * Mth.nextInt(this.random, -1, 1);
                     int j1 = j + Mth.nextInt(this.random, 7, 40) * Mth.nextInt(this.random, -1, 1);
