--- a/net/minecraft/network/protocol/common/custom/CustomPacketPayload.java
+++ b/net/minecraft/network/protocol/common/custom/CustomPacketPayload.java
@@ -18,7 +_,15 @@
     }
 
     static <T extends CustomPacketPayload> CustomPacketPayload.Type<T> createType(final String id) {
-        return new CustomPacketPayload.Type<>(Identifier.withDefaultNamespace(id));
+        return new CustomPacketPayload.Type<>(Identifier.parse(id));
+    }
+
+    static <T extends CustomPacketPayload> CustomPacketPayload.Type<T> createType(String namespace, String path) {
+        return new CustomPacketPayload.Type<>(Identifier.fromNamespaceAndPath(namespace, path));
+    }
+
+    static <T extends CustomPacketPayload> CustomPacketPayload.Type<T> createType(Identifier name) {
+        return new CustomPacketPayload.Type<>(name);
     }
 
     static <B extends FriendlyByteBuf> StreamCodec<B, CustomPacketPayload> codec(
