--- a/net/minecraft/network/syncher/EntityDataSerializers.java
+++ b/net/minecraft/network/syncher/EntityDataSerializers.java
@@ -128,15 +_,17 @@
     public static final EntityDataSerializer<HumanoidArm> HUMANOID_ARM = EntityDataSerializer.forValueType(HumanoidArm.STREAM_CODEC);
 
     public static void registerSerializer(EntityDataSerializer<?> p_135051_) {
+        int id =
         SERIALIZERS.add(p_135051_);
+        if (id >= 256) throw new RuntimeException("Vanilla DataSerializer ID limit exceeded");
     }
 
     public static @Nullable EntityDataSerializer<?> getSerializer(int p_135049_) {
-        return SERIALIZERS.byId(p_135049_);
+        return net.minecraftforge.common.ForgeHooks.getSerializer(p_135049_, SERIALIZERS);
     }
 
     public static int getSerializedId(EntityDataSerializer<?> p_135053_) {
-        return SERIALIZERS.getId(p_135053_);
+        return net.minecraftforge.common.ForgeHooks.getSerializerId(p_135053_, SERIALIZERS);
     }
 
     private EntityDataSerializers() {
