--- a/net/minecraft/server/packs/repository/PackDetector.java
+++ b/net/minecraft/server/packs/repository/PackDetector.java
@@ -19,6 +_,10 @@
     }
 
     public @Nullable T detectPackResources(final Path content, final List<ForbiddenSymlinkInfo> issues) throws IOException {
+        return detectPackResources(content, issues, true);
+    }
+    @Nullable
+    public T detectPackResources(Path content, List<ForbiddenSymlinkInfo> issues, boolean requireMeta) throws IOException {
         Path path = content;
 
         BasicFileAttributes basicfileattributes;
@@ -43,10 +_,11 @@
             if (!issues.isEmpty()) {
                 return null;
             } else {
-                return !Files.isRegularFile(path.resolve("pack.mcmeta")) ? null : this.createDirectoryPack(path);
+                return !Files.isRegularFile(path.resolve("pack.mcmeta")) && requireMeta ? null : this.createDirectoryPack(path);
             }
         } else {
-            return basicfileattributes.isRegularFile() && path.getFileName().toString().endsWith(".zip") ? this.createZipPack(path) : null;
+            var name = path.getFileName().toString();
+            return basicfileattributes.isRegularFile() && (name.endsWith(".zip") || name.endsWith(".jar")) ? this.createZipPack(path) : null;
         }
     }
 
