--- a/net/minecraft/server/packs/repository/Pack.java
+++ b/net/minecraft/server/packs/repository/Pack.java
@@ -23,6 +_,7 @@
     private final Pack.ResourcesSupplier resources;
     private final Pack.Metadata metadata;
     private final PackSelectionConfig selectionConfig;
+    private final boolean hidden; // Forge: Allow packs to be hidden from the UI entirely
 
     public static @Nullable Pack readMetaAndCreate(
         final PackLocationInfo location, final Pack.ResourcesSupplier resources, final PackType packType, final PackSelectionConfig selectionConfig
@@ -39,6 +_,7 @@
         this.resources = resources;
         this.metadata = metadata;
         this.selectionConfig = selectionConfig;
+        this.hidden = metadata.isHidden();
     }
 
     public static Pack.@Nullable Metadata readPackMetadata(
@@ -62,7 +_,7 @@
                 PackCompatibility packcompatibility = PackCompatibility.forVersion(packmetadatasection.supportedFormats(), currentPackVersion);
                 OverlayMetadataSection overlaymetadatasection = packresources.getMetadataSection(OverlayMetadataSection.forPackType(type));
                 List<String> list = overlaymetadatasection != null ? overlaymetadatasection.overlaysForVersion(currentPackVersion) : List.of();
-                pack$metadata = new Pack.Metadata(packmetadatasection.description(), packcompatibility, featureflagset, list);
+                pack$metadata = new Pack.Metadata(packmetadatasection.description(), packcompatibility, featureflagset, list, packresources.isHidden());
             }
 
             return pack$metadata;
@@ -124,6 +_,10 @@
         return this.location.source();
     }
 
+    public boolean isHidden() {
+        return this.hidden;
+    }
+
     @Override
     public boolean equals(final Object o) {
         if (this == o) {
@@ -138,7 +_,10 @@
         return this.location.hashCode();
     }
 
-    public record Metadata(Component description, PackCompatibility compatibility, FeatureFlagSet requestedFeatures, List<String> overlays) {
+    public record Metadata(Component description, PackCompatibility compatibility, FeatureFlagSet requestedFeatures, List<String> overlays, boolean isHidden) {
+        public Metadata(Component description, PackCompatibility compatibility, FeatureFlagSet requestedFeatures, List<String> overlays) {
+            this(description, compatibility, requestedFeatures, overlays, false);
+        }
     }
 
     public static enum Position {
