--- a/net/minecraft/world/item/BoneMealItem.java
+++ b/net/minecraft/world/item/BoneMealItem.java
@@ -38,7 +_,7 @@
         BlockPos blockpos = context.getClickedPos();
         BlockPos blockpos1 = blockpos.relative(context.getClickedFace());
         ItemStack itemstack = context.getItemInHand();
-        if (growCrop(itemstack, level, blockpos)) {
+        if (applyBonemeal(itemstack, level, blockpos, context.getPlayer())) {
             if (!level.isClientSide()) {
                 itemstack.causeUseVibration(context.getPlayer(), GameEvent.ITEM_INTERACT_FINISH);
                 level.levelEvent(1505, blockpos, 15);
@@ -62,8 +_,18 @@
         }
     }
 
+    @Deprecated //Forge: Use Player/Hand version
     public static boolean growCrop(final ItemStack itemStack, final Level level, final BlockPos pos) {
+        if (level instanceof net.minecraft.server.level.ServerLevel) {
+            return applyBonemeal(itemStack, level, pos, null);
+        }
+        return false;
+    }
+
+    public static boolean applyBonemeal(final ItemStack itemStack, final Level level, final BlockPos pos, net.minecraft.world.entity.player.Player player) {
         BlockState blockstate = level.getBlockState(pos);
+        int hook = net.minecraftforge.event.ForgeEventFactory.onApplyBonemeal(player, level, pos, blockstate, itemStack);
+        if (hook != 0) return hook > 0;
         if (blockstate.getBlock() instanceof BonemealableBlock bonemealableblock && bonemealableblock.isValidBonemealTarget(level, pos, blockstate)) {
             if (level instanceof ServerLevel) {
                 if (bonemealableblock.isBonemealSuccess(level, level.getRandom(), pos, blockstate)) {
