--- a/net/minecraft/world/item/BowItem.java
+++ b/net/minecraft/world/item/BowItem.java
@@ -32,6 +_,9 @@
                 return false;
             } else {
                 int i = this.getUseDuration(itemStack, entity) - remainingTime;
+                i = net.minecraftforge.event.ForgeEventFactory.onArrowLoose(itemStack, level, player, i, true);
+                if (i < 0) return false;
+
                 float f = getPowerForTime(i);
                 if (f < 0.1) {
                     return false;
@@ -95,6 +_,8 @@
     public InteractionResult use(final Level level, final Player player, final InteractionHand hand) {
         ItemStack itemstack = player.getItemInHand(hand);
         boolean flag = !player.getProjectile(itemstack).isEmpty();
+        var ret = net.minecraftforge.event.ForgeEventFactory.onArrowNock(itemstack, level, player, hand, flag);
+        if (ret != null) return ret;
         if (!player.hasInfiniteMaterials() && !flag) {
             return InteractionResult.FAIL;
         } else {
