--- a/net/minecraft/world/item/BucketItem.java
+++ b/net/minecraft/world/item/BucketItem.java
@@ -34,9 +_,21 @@
 public class BucketItem extends Item implements DispensibleContainerItem {
     private final Fluid content;
 
+    // Forge: Use the other constructor that takes a Supplier
+    @Deprecated
     public BucketItem(final Fluid content, final Item.Properties properties) {
         super(properties);
         this.content = content;
+        this.fluidSupplier = net.minecraftforge.registries.ForgeRegistries.FLUIDS.getDelegateOrThrow(content);
+    }
+
+    /**
+     * @param supplier A fluid supplier such as {@link net.minecraftforge.registries.RegistryObject<Fluid>}
+     */
+    public BucketItem(java.util.function.Supplier<? extends Fluid> supplier, Item.Properties builder) {
+       super(builder);
+       this.content = null;
+       this.fluidSupplier = supplier;
     }
 
     @Override
@@ -45,6 +_,8 @@
         BlockHitResult blockhitresult = getPlayerPOVHitResult(
             level, player, this.content == Fluids.EMPTY ? ClipContext.Fluid.SOURCE_ONLY : ClipContext.Fluid.NONE
         );
+        var ret = net.minecraftforge.event.ForgeEventFactory.onBucketUse(player, level, itemstack, blockhitresult);
+        if (ret != null) return ret;
         if (blockhitresult.getType() == HitResult.Type.MISS) {
             return InteractionResult.PASS;
         } else if (blockhitresult.getType() != HitResult.Type.BLOCK) {
@@ -61,7 +_,7 @@
                     ItemStack itemstack3 = bucketpickup.pickupBlock(player, level, blockpos, blockstate1);
                     if (!itemstack3.isEmpty()) {
                         player.awardStat(Stats.ITEM_USED.get(this));
-                        bucketpickup.getPickupSound().ifPresent(soundEvent -> player.playSound(soundEvent, 1.0F, 1.0F));
+                        bucketpickup.getPickupSound(blockstate1).ifPresent(soundEvent -> player.playSound(soundEvent, 1.0F, 1.0F));
                         level.gameEvent(player, GameEvent.FLUID_PICKUP, blockpos);
                         ItemStack itemstack2 = ItemUtils.createFilledResult(itemstack, player, itemstack3);
                         if (!level.isClientSide()) {
@@ -75,8 +_,8 @@
                 return InteractionResult.FAIL;
             } else {
                 BlockState blockstate = level.getBlockState(blockpos);
-                BlockPos blockpos2 = blockstate.getBlock() instanceof LiquidBlockContainer && this.content == Fluids.WATER ? blockpos : blockpos1;
-                if (this.emptyContents(player, level, blockpos2, blockhitresult)) {
+                BlockPos blockpos2 = canBlockContainFluid(level, blockpos, blockstate) ? blockpos : blockpos1;
+                if (this.emptyContents(player, level, blockpos2, blockhitresult, itemstack)) {
                     this.checkExtraContent(player, level, itemstack, blockpos2);
                     if (player instanceof ServerPlayer) {
                         CriteriaTriggers.PLACED_BLOCK.trigger((ServerPlayer)player, blockpos2, itemstack);
@@ -100,8 +_,13 @@
     public void checkExtraContent(final @Nullable LivingEntity user, final Level level, final ItemStack itemStack, final BlockPos pos) {
     }
 
+    @Deprecated //Forge: use the ItemStack sensitive version
     @Override
     public boolean emptyContents(final @Nullable LivingEntity user, final Level level, final BlockPos pos, final @Nullable BlockHitResult hitResult) {
+        return this.emptyContents(user, level, pos, hitResult, null);
+    }
+
+    public boolean emptyContents(final @Nullable LivingEntity user, final Level level, final BlockPos pos, final @Nullable BlockHitResult hitResult, final @Nullable ItemStack container) {
         if (!(this.content instanceof FlowingFluid flowingfluid)) {
             return false;
         } else {
@@ -113,8 +_,12 @@
                 || block instanceof LiquidBlockContainer liquidblockcontainer
                     && liquidblockcontainer.canPlaceLiquid(user, level, pos, blockstate, this.content);
             boolean flag3 = blockstate.isAir() || flag2 && (!flag1 || hitResult == null);
+            java.util.Optional<net.minecraftforge.fluids.FluidStack> containedFluidStack = java.util.Optional.ofNullable(container).flatMap(net.minecraftforge.fluids.FluidUtil::getFluidContained);
             if (!flag3) {
-                return hitResult != null && this.emptyContents(user, level, hitResult.getBlockPos().relative(hitResult.getDirection()), null);
+                return hitResult != null && this.emptyContents(user, level, hitResult.getBlockPos().relative(hitResult.getDirection()), null, container);
+            } else if (containedFluidStack.isPresent() && this.content.getFluidType().isVaporizedOnPlacement(level, pos, containedFluidStack.get())) {
+                this.content.getFluidType().onVaporize(user, level, pos, containedFluidStack.get());
+                return true;
             } else if (level.environmentAttributes().getValue(EnvironmentAttributes.WATER_EVAPORATES, pos) && this.content.is(FluidTags.WATER)) {
                 int l = pos.getX();
                 int i = pos.getY();
@@ -131,7 +_,7 @@
                 }
 
                 return true;
-            } else if (block instanceof LiquidBlockContainer liquidblockcontainer1 && this.content == Fluids.WATER) {
+            } else if (block instanceof LiquidBlockContainer liquidblockcontainer1 && liquidblockcontainer1.canPlaceLiquid(user, level, pos, blockstate, content) && this.content == Fluids.WATER) {
                 liquidblockcontainer1.placeLiquid(level, pos, blockstate, flowingfluid.getSource(false));
                 this.playEmptySound(user, level, pos);
                 return true;
@@ -152,8 +_,33 @@
 
     protected void playEmptySound(final @Nullable LivingEntity user, final LevelAccessor level, final BlockPos pos) {
         SoundEvent soundevent = this.content.is(FluidTags.LAVA) ? SoundEvents.BUCKET_EMPTY_LAVA : SoundEvents.BUCKET_EMPTY;
+        var custom = this.content.getFluidType().getSound(user, level, pos, net.minecraftforge.common.SoundActions.BUCKET_EMPTY);
+        if (custom != null) {
+            soundevent = custom;
+        }
         level.playSound(user, pos, soundevent, SoundSource.BLOCKS, 1.0F, 1.0F);
         level.gameEvent(user, GameEvent.FLUID_PLACE, pos);
+    }
+
+    /** Forge: TODO: Forge ItemStack capabilities - Lex 042724
+    @Override
+    public net.minecraftforge.common.capabilities.ICapabilityProvider initCapabilities(ItemStack stack, @Nullable net.minecraft.nbt.CompoundTag nbt) {
+        if (this.getClass() == BucketItem.class) {
+            return new net.minecraftforge.fluids.capability.wrappers.FluidBucketWrapper(stack);
+        } else {
+            return super.initCapabilities(stack, nbt);
+        }
+    }
+    */
+
+    private final java.util.function.Supplier<? extends Fluid> fluidSupplier;
+
+    public Fluid getFluid() {
+        return fluidSupplier.get();
+    }
+
+    protected boolean canBlockContainFluid(Level worldIn, BlockPos posIn, BlockState blockstate) {
+        return blockstate.getBlock() instanceof LiquidBlockContainer liquid && liquid.canPlaceLiquid(null, worldIn, posIn, blockstate, this.content);
     }
 
     public Fluid getContent() {
