--- a/net/minecraft/world/item/ItemStack.java
+++ b/net/minecraft/world/item/ItemStack.java
@@ -102,7 +_,7 @@
 import org.jspecify.annotations.Nullable;
 import org.slf4j.Logger;
 
-public final class ItemStack implements DataComponentHolder, ItemInstance {
+public final class ItemStack extends net.minecraftforge.common.capabilities.CapabilityProvider.ItemStacks implements DataComponentHolder, ItemInstance, net.minecraftforge.common.extensions.IForgeItemStack {
     private static final List<Component> OP_NBT_WARNING = List.of(
         Component.translatable("item.op_warning.line1").withStyle(ChatFormatting.RED, ChatFormatting.BOLD),
         Component.translatable("item.op_warning.line2").withStyle(ChatFormatting.RED),
@@ -260,12 +_,15 @@
     }
 
     private ItemStack(final Holder<Item> item, final int count, final PatchedDataComponentMap components) {
+        super(true);
         this.item = item;
         this.count = count;
         this.components = components;
+        gatherCapabilities(() -> this.item.get().getCapabilityProvider(this));
     }
 
     private ItemStack(final @Nullable Void nullMarker) {
+        super(false);
         this.item = null;
         this.components = new PatchedDataComponentMap(DataComponentMap.EMPTY);
     }
@@ -358,13 +_,22 @@
     }
 
     public InteractionResult useOn(final UseOnContext context) {
+        if (!context.getLevel().isClientSide()) return net.minecraftforge.common.ForgeHooks.onPlaceItemIntoWorld(context);
+        return onItemUse(context, (c) -> getItem().useOn(context));
+    }
+
+    public InteractionResult onItemUseFirst(UseOnContext context) {
+        return onItemUse(context, (c) -> getItem().onItemUseFirst(this, context));
+    }
+
+    private InteractionResult onItemUse(UseOnContext context, java.util.function.Function<UseOnContext, InteractionResult> callback) {
         Player player = context.getPlayer();
         BlockPos blockpos = context.getClickedPos();
         if (player != null && !player.getAbilities().mayBuild && !this.canPlaceOnBlockInAdventureMode(new BlockInWorld(context.getLevel(), blockpos, false))) {
             return InteractionResult.PASS;
         } else {
             Item item = this.getItem();
-            InteractionResult interactionresult = item.useOn(context);
+            InteractionResult interactionresult = callback.apply(context);
             if (player != null
                 && interactionresult instanceof InteractionResult.Success interactionresult$success
                 && interactionresult$success.wasItemInteraction()) {
@@ -447,18 +_,26 @@
     }
 
     public void hurtAndBreak(final int amount, final ServerLevel level, final @Nullable ServerPlayer player, final Consumer<Item> onBreak) {
-        int i = this.processDurabilityChange(amount, level, player);
+        // FORGE: use context-sensitive sister of processDurabilityChange that calls IForgeItem.damageItem
+        int i = this.processDurabilityChange(amount, level, player, true, onBreak);
         if (i != 0) {
             this.applyDamage(this.getDamageValue() + i, player, onBreak);
         }
     }
 
     private int processDurabilityChange(final int amount, final ServerLevel level, final @Nullable ServerPlayer player) {
+        return this.processDurabilityChange(amount, level, player, false, item -> { });
+    }
+
+    /** FORGE: context-sensitive sister of processDurabilityChange that calls IForgeItem.damageItem */
+    private int processDurabilityChange(int amount, final ServerLevel level, final @Nullable ServerPlayer player, final boolean canBreak, final Consumer<Item> onBreak) {
         if (!this.isDamageableItem()) {
             return 0;
         } else if (player != null && player.hasInfiniteMaterials()) {
             return 0;
         } else {
+            // FORGE: modify the base damage based on the item's impl of IForgeItem.damageItem
+            amount = this.damageItem(amount, level, player, canBreak, onBreak);
             return amount > 0 ? EnchantmentHelper.processDurabilityChange(level, this, amount) : amount;
         }
     }
@@ -498,7 +_,13 @@
                 amount,
                 serverlevel,
                 owner instanceof ServerPlayer serverplayer ? serverplayer : null,
-                brokenItem -> owner.onEquippedItemBroken(brokenItem, slot)
+                brokenItem -> {
+                    if (owner instanceof Player player) {
+                        net.minecraftforge.event.ForgeEventFactory.onPlayerDestroyItem(player, this, slot);
+                        if (player.getUseItem() == this) player.stopUsingItem(); // Forge: fix MC-168573
+                    }
+                    owner.onEquippedItemBroken(brokenItem, slot);
+                }
             );
         }
     }
@@ -857,6 +_,7 @@
             List<Component> list = Lists.newArrayList();
             list.add(this.getStyledHoverName());
             this.addDetailsToTooltip(context, tooltipdisplay, player, tooltipFlag, list::add);
+            net.minecraftforge.event.ForgeEventFactory.onItemTooltip(this, player, list, tooltipFlag);
             return list;
         }
     }
@@ -1105,6 +_,7 @@
         }
     }
 
+    /** @deprecated Forge: Use {@linkplain net.minecraftforge.common.extensions.IForgeItemStack#onDestroyed(ItemEntity, net.minecraft.world.damagesource.DamageSource) damage source sensitive version} */
     public void onDestroyed(final ItemEntity itemEntity) {
         this.getItem().onDestroyed(itemEntity);
     }
