--- a/net/minecraft/world/item/SpawnEggItem.java
+++ b/net/minecraft/world/item/SpawnEggItem.java
@@ -148,6 +_,10 @@
         return typedentitydata != null ? typedentitydata.type() : null;
     }
 
+    public EntityType<?> getDefaultType() {
+        return this.defaultType;
+    }
+
     @Override
     public FeatureFlagSet requiredFeatures() {
         return Optional.ofNullable(this.components().get(DataComponents.ENTITY_DATA))
