--- a/net/minecraft/world/entity/boss/wither/WitherBoss.java
+++ b/net/minecraft/world/entity/boss/wither/WitherBoss.java
@@ -323,7 +_,7 @@
 
             if (this.destroyBlocksTick > 0) {
                 this.destroyBlocksTick--;
-                if (this.destroyBlocksTick == 0 && level.getGameRules().get(GameRules.MOB_GRIEFING)) {
+                if (this.destroyBlocksTick == 0 && net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(level, this)) {
                     boolean flag = false;
                     int l = Mth.floor(this.getBbWidth() / 2.0F + 1.0F);
                     int i1 = Mth.floor(this.getBbHeight());
@@ -332,7 +_,7 @@
                         this.getBlockX() - l, this.getBlockY(), this.getBlockZ() - l, this.getBlockX() + l, this.getBlockY() + i1, this.getBlockZ() + l
                     )) {
                         BlockState blockstate = level.getBlockState(blockpos);
-                        if (canDestroy(blockstate)) {
+                        if (blockstate.canEntityDestroy(level, blockpos, this) && net.minecraftforge.event.ForgeEventFactory.onEntityDestroyBlock(this, blockpos, blockstate)) {
                             flag = level.destroyBlock(blockpos, true, this) || flag;
                         }
                     }
@@ -351,6 +_,10 @@
         }
     }
 
+    /**
+     * @deprecated Forge: Use {@link BlockState#canEntityDestroy(net.minecraft.world.level.BlockGetter, BlockPos, Entity)} instead.
+     */
+    @Deprecated
     public static boolean canDestroy(final BlockState state) {
         return !state.isAir() && !state.is(BlockTags.WITHER_IMMUNE);
     }
