--- a/net/minecraft/server/advancements/AdvancementVisibilityEvaluator.java
+++ b/net/minecraft/server/advancements/AdvancementVisibilityEvaluator.java
@@ -71,6 +_,16 @@
         evaluateVisibility(advancementnode, stack, isDone, output);
     }
 
+    public static boolean isVisible(AdvancementNode advancement, Predicate<AdvancementNode> test) {
+        Stack<AdvancementVisibilityEvaluator.VisibilityRule> stack = new ObjectArrayList<>();
+
+        for (int i = 0; i <= 2; i++) {
+            stack.push(AdvancementVisibilityEvaluator.VisibilityRule.NO_CHANGE);
+        }
+
+        return evaluateVisibility(advancement.root(), stack, test, (node, currentVis) -> {});
+    }
+
     @FunctionalInterface
     public interface Output {
         void accept(AdvancementNode advancement, boolean visible);
