--- a/net/minecraft/world/level/block/BaseFireBlock.java
+++ b/net/minecraft/world/level/block/BaseFireBlock.java
@@ -158,6 +_,7 @@
         if (!oldState.is(state.getBlock())) {
             if (inPortalDimension(level)) {
                 Optional<PortalShape> optionalShape = PortalShape.findEmptyPortalShape(level, pos, Direction.Axis.X);
+                optionalShape = net.minecraftforge.event.ForgeEventFactory.onTrySpawnPortal(level, pos, optionalShape);
                 if (optionalShape.isPresent()) {
                     optionalShape.get().createPortalBlocks(level);
                     return;
@@ -201,7 +_,7 @@
 
         for (Direction face : Direction.values()) {
             BlockState state = level.getBlockState(testPos.set(pos).move(face));
-            if (PortalShape.FRAME.test(state)) {
+            if (PortalShape.FRAME_TEST.test(state, level, testPos)) {
                 Direction.Axis preferredAxis = forwardDirection.getAxis().isHorizontal()
                     ? forwardDirection.getCounterClockWise().getAxis()
                     : Direction.Plane.HORIZONTAL.getRandomAxis(level.getRandom());
