--- a/net/minecraft/world/level/block/BucketPickup.java
+++ b/net/minecraft/world/level/block/BucketPickup.java
@@ -9,8 +_,10 @@
 import net.minecraft.world.level.block.state.BlockState;
 import org.jspecify.annotations.Nullable;
 
-public interface BucketPickup {
+public interface BucketPickup extends net.minecraftforge.common.extensions.IForgeBucketPickup {
     ItemStack pickupBlock(@Nullable LivingEntity user, LevelAccessor level, BlockPos pos, BlockState state);
 
+    /** @deprecated Forge: Use state-sensitive variant instead */
+    @Deprecated
     Optional<SoundEvent> getPickupSound();
 }
