--- a/net/minecraft/world/level/block/CactusFlowerBlock.java
+++ b/net/minecraft/world/level/block/CactusFlowerBlock.java
@@ -31,6 +_,6 @@
     @Override
     protected boolean mayPlaceOn(final BlockState state, final BlockGetter level, final BlockPos pos) {
         BlockState blockstate = level.getBlockState(pos);
-        return blockstate.is(BlockTags.SUPPORT_OVERRIDE_CACTUS_FLOWER) || blockstate.isFaceSturdy(level, pos, Direction.UP, SupportType.CENTER);
+        return blockstate.is(BlockTags.SUPPORT_OVERRIDE_CACTUS_FLOWER) || blockstate.isFaceSturdy(level, pos, Direction.UP, SupportType.CENTER) || blockstate.getBlock() instanceof FarmlandBlock;
     }
 }
