--- a/net/minecraft/world/level/block/ChorusFlowerBlock.java
+++ b/net/minecraft/world/level/block/ChorusFlowerBlock.java
@@ -65,7 +_,7 @@
         BlockPos above = pos.above();
         if (level.isEmptyBlock(above) && above.getY() <= level.getMaxY()) {
             int currentAge = state.getValue(AGE);
-            if (currentAge < 5) {
+            if (currentAge < 5 && net.minecraftforge.common.ForgeHooks.onCropsGrowPre(level, above, state, true)) {
                 boolean growUpwards = false;
                 boolean pillarOnSupportBlock = false;
                 BlockState belowState = level.getBlockState(pos.below());
@@ -121,6 +_,7 @@
                 } else {
                     this.placeDeadFlower(level, pos);
                 }
+                net.minecraftforge.common.ForgeHooks.onCropsGrowPost(level, pos, state);
             }
         }
     }
