--- a/net/minecraft/world/level/block/ConcretePowderBlock.java
+++ b/net/minecraft/world/level/block/ConcretePowderBlock.java
@@ -36,7 +_,7 @@
 
     @Override
     public void onLand(final Level level, final BlockPos pos, final BlockState state, final BlockState replacedBlock, final FallingBlockEntity entity) {
-        if (shouldSolidify(level, pos, replacedBlock)) {
+        if (shouldSolidify(level, pos, state, replacedBlock.getFluidState())) { // Forge: Use block of falling entity instead of block at replaced position, and check if shouldSolidify with the FluidState of the replaced block
             level.setBlock(pos, this.concrete.defaultBlockState(), 3);
         }
     }
@@ -49,20 +_,24 @@
         return shouldSolidify(blockgetter, blockpos, blockstate) ? this.concrete.defaultBlockState() : super.getStateForPlacement(context);
     }
 
+    private static boolean shouldSolidify(final BlockGetter level, final BlockPos pos, final BlockState state, final net.minecraft.world.level.material.FluidState fluidState) {
+        return state.canBeHydrated(level, pos, fluidState, pos) || touchesLiquid(level, pos, state);
+    }
+
     private static boolean shouldSolidify(final BlockGetter level, final BlockPos pos, final BlockState replacedBlock) {
-        return canSolidify(replacedBlock) || touchesLiquid(level, pos);
+        return shouldSolidify(level, pos, replacedBlock, level.getFluidState(pos));
     }
 
-    private static boolean touchesLiquid(final BlockGetter level, final BlockPos pos) {
+    private static boolean touchesLiquid(final BlockGetter level, BlockPos pos, BlockState state) {
         boolean flag = false;
         BlockPos.MutableBlockPos blockpos$mutableblockpos = pos.mutable();
 
         for (Direction direction : Direction.values()) {
             BlockState blockstate = level.getBlockState(blockpos$mutableblockpos);
-            if (direction != Direction.DOWN || canSolidify(blockstate)) {
+            if (direction != Direction.DOWN || state.canBeHydrated(level, pos, blockstate.getFluidState(), blockpos$mutableblockpos)) {
                 blockpos$mutableblockpos.setWithOffset(pos, direction);
                 blockstate = level.getBlockState(blockpos$mutableblockpos);
-                if (canSolidify(blockstate) && !blockstate.isFaceSturdy(level, pos, direction.getOpposite())) {
+                if (state.canBeHydrated(level, pos, blockstate.getFluidState(), blockpos$mutableblockpos) && !blockstate.isFaceSturdy(level, pos, direction.getOpposite())) {
                     flag = true;
                     break;
                 }
@@ -87,7 +_,7 @@
         final BlockState neighbourState,
         final RandomSource random
     ) {
-        return touchesLiquid(level, pos)
+        return touchesLiquid(level, pos, state)
             ? this.concrete.defaultBlockState()
             : super.updateShape(state, level, ticks, pos, directionToNeighbour, neighbourPos, neighbourState, random);
     }
