--- a/net/minecraft/world/level/block/CoralBlock.java
+++ b/net/minecraft/world/level/block/CoralBlock.java
@@ -61,9 +_,10 @@
     }
 
     protected boolean scanForWater(final BlockGetter level, final BlockPos blockPos) {
+        BlockState state = level.getBlockState(blockPos);
         for (Direction direction : Direction.values()) {
             FluidState fluidstate = level.getFluidState(blockPos.relative(direction));
-            if (fluidstate.is(FluidTags.WATER)) {
+            if (state.canBeHydrated(level, blockPos, fluidstate, blockPos.relative(direction))) {
                 return true;
             }
         }
