--- a/net/minecraft/world/level/block/CropBlock.java
+++ b/net/minecraft/world/level/block/CropBlock.java
@@ -47,7 +_,7 @@
 
     @Override
     protected boolean mayPlaceOn(final BlockState state, final BlockGetter level, final BlockPos pos) {
-        return state.is(BlockTags.SUPPORTS_CROPS);
+        return state.getBlock() instanceof FarmlandBlock;
     }
 
     protected IntegerProperty getAgeProperty() {
@@ -81,8 +_,9 @@
             int i = this.getAge(state);
             if (i < this.getMaxAge()) {
                 float f = getGrowthSpeed(this, level, pos);
-                if (random.nextInt((int)(25.0F / f) + 1) == 0) {
+                if (net.minecraftforge.common.ForgeHooks.onCropsGrowPre(level, pos, state, random.nextInt((int)(25.0F / f) + 1) == 0)) {
                     level.setBlock(pos, this.getStateForAge(i + 1), 2);
+                    net.minecraftforge.common.ForgeHooks.onCropsGrowPost(level, pos, state);
                 }
             }
         }
@@ -105,9 +_,9 @@
             for (int j = -1; j <= 1; j++) {
                 float f1 = 0.0F;
                 BlockState blockstate = level.getBlockState(blockpos.offset(i, 0, j));
-                if (blockstate.is(BlockTags.GROWS_CROPS)) {
+                if (blockstate.canSustainPlant(level, blockpos.offset(i, 0, j), net.minecraft.core.Direction.UP, (net.minecraftforge.common.IPlantable)type)) {
                     f1 = 1.0F;
-                    if (blockstate.getValueOrElse(FarmlandBlock.MOISTURE, 0) > 0) {
+                    if (blockstate.isFertile(level, blockpos.offset(i, 0, j))) {
                         f1 = 3.0F;
                     }
                 }
@@ -159,7 +_,7 @@
         final InsideBlockEffectApplier effectApplier,
         final boolean isPrecise
     ) {
-        if (level instanceof ServerLevel serverlevel && entity instanceof Ravager && serverlevel.getGameRules().get(GameRules.MOB_GRIEFING)) {
+        if (level instanceof ServerLevel serverlevel && entity instanceof Ravager && net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(serverlevel, entity)) {
             serverlevel.destroyBlock(pos, true, entity);
         }
 
