--- a/net/minecraft/world/level/block/DropperBlock.java
+++ b/net/minecraft/world/level/block/DropperBlock.java
@@ -56,7 +_,7 @@
                 level.levelEvent(1001, pos, 0);
             } else {
                 ItemStack itemstack = dispenserblockentity.getItem(i);
-                if (!itemstack.isEmpty()) {
+                if (!itemstack.isEmpty() && net.minecraftforge.items.VanillaInventoryCodeHooks.dropperInsertHook(level, pos, dispenserblockentity, i, itemstack)) {
                     Direction direction = level.getBlockState(pos).getValue(FACING);
                     Container container = HopperBlockEntity.getContainerAt(level, pos.relative(direction));
                     ItemStack itemstack1;
