--- a/net/minecraft/world/level/block/GrowingPlantHeadBlock.java
+++ b/net/minecraft/world/level/block/GrowingPlantHeadBlock.java
@@ -48,10 +_,13 @@
 
     @Override
     protected void randomTick(final BlockState state, final ServerLevel level, final BlockPos pos, final RandomSource random) {
-        if (state.getValue(AGE) < 25 && random.nextDouble() < this.growPerTickProbability) {
+        var vanilla = random.nextDouble() < this.growPerTickProbability;
+        var target = pos.relative(this.growthDirection);
+        if (state.getValue(AGE) < 25 && net.minecraftforge.common.ForgeHooks.onCropsGrowPre(level, target, level.getBlockState(target), vanilla)) {
             BlockPos growthPos = pos.relative(this.growthDirection);
             if (this.canGrowInto(level.getBlockState(growthPos))) {
                 level.setBlockAndUpdate(growthPos, this.getGrowIntoState(state, level.getRandom()));
+                net.minecraftforge.common.ForgeHooks.onCropsGrowPost(level, growthPos, level.getBlockState(growthPos));
             }
         }
     }
