--- a/net/minecraft/world/level/block/PoweredRailBlock.java
+++ b/net/minecraft/world/level/block/PoweredRailBlock.java
@@ -16,6 +_,7 @@
     public static final MapCodec<PoweredRailBlock> CODEC = simpleCodec(PoweredRailBlock::new);
     public static final EnumProperty<RailShape> SHAPE = BlockStateProperties.RAIL_SHAPE_STRAIGHT;
     public static final BooleanProperty POWERED = BlockStateProperties.POWERED;
+    private final boolean isActivator;  // TRUE for an Activator Rail, FALSE for Powered Rail
 
     @Override
     public MapCodec<PoweredRailBlock> codec() {
@@ -23,7 +_,16 @@
     }
 
     public PoweredRailBlock(final BlockBehaviour.Properties properties) {
+        this(properties, false);
+    }
+
+    protected PoweredRailBlock(final BlockBehaviour.Properties properties, boolean isPoweredRail) {
         super(true, properties);
+        this.isActivator = !isPoweredRail;
+        this.registerDefaultState();
+    }
+
+    protected void registerDefaultState() {
         this.registerDefaultState(this.stateDefinition.any().setValue(SHAPE, RailShape.NORTH_SOUTH).setValue(POWERED, false).setValue(WATERLOGGED, false));
     }
 
@@ -35,7 +_,7 @@
             int j = pos.getY();
             int k = pos.getZ();
             boolean flag = true;
-            RailShape railshape = state.getValue(SHAPE);
+            RailShape railshape = state.getValue(getShapeProperty());
             switch (railshape) {
                 case NORTH_SOUTH:
                     if (forward) {
@@ -104,10 +_,10 @@
 
     protected boolean isSameRailWithPower(final Level level, final BlockPos pos, final boolean forward, final int searchDepth, final RailShape dir) {
         BlockState blockstate = level.getBlockState(pos);
-        if (!blockstate.is(this)) {
+        if (!(blockstate.getBlock() instanceof PoweredRailBlock other) || this.isActivatorRail() != other.isActivatorRail()) {
             return false;
         } else {
-            RailShape railshape = blockstate.getValue(SHAPE);
+            RailShape railshape = other.getRailDirection(blockstate, level, pos, null);
             if (dir != RailShape.EAST_WEST
                 || railshape != RailShape.NORTH_SOUTH && railshape != RailShape.ASCENDING_NORTH && railshape != RailShape.ASCENDING_SOUTH) {
                 if (dir != RailShape.NORTH_SOUTH
@@ -115,7 +_,7 @@
                     if (!blockstate.getValue(POWERED)) {
                         return false;
                     } else {
-                        return level.hasNeighborSignal(pos) ? true : this.findPoweredRailSignal(level, pos, blockstate, forward, searchDepth + 1);
+                        return level.hasNeighborSignal(pos) ? true : other.findPoweredRailSignal(level, pos, blockstate, forward, searchDepth + 1);
                     }
                 } else {
                     return false;
@@ -135,7 +_,7 @@
         if (flag1 != flag) {
             level.setBlock(pos, state.setValue(POWERED, flag1), 3);
             level.updateNeighborsAt(pos.below(), this);
-            if (state.getValue(SHAPE).isSlope()) {
+            if (state.getValue(getShapeProperty()).isSlope()) {
                 level.updateNeighborsAt(pos.above(), this);
             }
         }
@@ -162,6 +_,10 @@
 
     @Override
     protected void createBlockStateDefinition(final StateDefinition.Builder<Block, BlockState> builder) {
-        builder.add(SHAPE, POWERED, WATERLOGGED);
+        builder.add(getShapeProperty(), POWERED, WATERLOGGED);
+    }
+
+    public boolean isActivatorRail() {
+        return isActivator;
     }
 }
