--- a/net/minecraft/world/level/block/SpongeBlock.java
+++ b/net/minecraft/world/level/block/SpongeBlock.java
@@ -53,6 +_,7 @@
     }
 
     private boolean removeWaterBreadthFirstSearch(final Level level, final BlockPos startPos) {
+        BlockState spongeState = level.getBlockState(startPos);
         return BlockPos.breadthFirstTraversal(
                 startPos,
                 6,
@@ -68,7 +_,7 @@
                     } else {
                         BlockState blockstate = level.getBlockState(pos);
                         FluidState fluidstate = level.getFluidState(pos);
-                        if (!fluidstate.is(FluidTags.WATER)) {
+                        if (!spongeState.canBeHydrated(level, startPos, fluidstate, pos)) {
                             return BlockPos.TraversalNodeStatus.SKIP;
                         } else if (blockstate.getBlock() instanceof BucketPickup bucketpickup
                             && !bucketpickup.pickupBlock(null, level, pos, blockstate).isEmpty()) {
