--- a/net/minecraft/world/level/block/StemBlock.java
+++ b/net/minecraft/world/level/block/StemBlock.java
@@ -76,14 +_,16 @@
 
     @Override
     protected boolean mayPlaceOn(final BlockState state, final BlockGetter level, final BlockPos pos) {
-        return state.is(this.stemSupportBlocks);
+        return state.is(this.stemSupportBlocks) || state.getBlock() instanceof FarmlandBlock;
     }
 
     @Override
     protected void randomTick(BlockState state, final ServerLevel level, final BlockPos pos, final RandomSource random) {
+        if (!level.isAreaLoaded(pos, 1)) return; // Forge: prevent loading unloaded chunks when checking neighbor's light
         if (level.getRawBrightness(pos, 0) >= 9) {
             float growthSpeed = CropBlock.getGrowthSpeed(this, level, pos);
-            if (random.nextInt((int)(25.0F / growthSpeed) + 1) == 0) {
+            var vanilla = random.nextInt((int)(25.0F / growthSpeed) + 1) == 0;
+            if (net.minecraftforge.common.ForgeHooks.onCropsGrowPre(level, pos, state, vanilla)) {
                 int age = state.getValue(AGE);
                 if (age < 7) {
                     state = state.setValue(AGE, age + 1);
@@ -92,7 +_,7 @@
                     Direction direction = Direction.Plane.HORIZONTAL.getRandomDirection(random);
                     BlockPos relative = pos.relative(direction);
                     BlockState stateBelow = level.getBlockState(relative.below());
-                    if (level.getBlockState(relative).isAir() && stateBelow.is(this.fruitSupportBlocks)) {
+                    if (level.getBlockState(relative).isAir() && (stateBelow.is(this.fruitSupportBlocks) || stateBelow.getBlock() instanceof FarmlandBlock)) {
                         Registry<Block> blocks = level.registryAccess().lookupOrThrow(Registries.BLOCK);
                         Optional<Block> fruit = blocks.getOptional(this.fruit);
                         Optional<Block> stem = blocks.getOptional(this.attachedStem);
@@ -102,6 +_,7 @@
                         }
                     }
                 }
+                net.minecraftforge.common.ForgeHooks.onCropsGrowPost(level, pos, state);
             }
         }
     }
@@ -134,5 +_,10 @@
     @Override
     protected void createBlockStateDefinition(final StateDefinition.Builder<Block, BlockState> builder) {
         builder.add(AGE);
+    }
+
+    @Override
+    public net.minecraftforge.common.PlantType getPlantType(BlockGetter world, BlockPos pos) {
+        return net.minecraftforge.common.PlantType.CROP;
     }
 }
