--- a/net/minecraft/world/level/block/SugarCaneBlock.java
+++ b/net/minecraft/world/level/block/SugarCaneBlock.java
@@ -19,7 +_,7 @@
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
-public class SugarCaneBlock extends Block {
+public class SugarCaneBlock extends Block implements net.minecraftforge.common.IPlantable {
     public static final MapCodec<SugarCaneBlock> CODEC = simpleCodec(SugarCaneBlock::new);
     public static final IntegerProperty AGE = BlockStateProperties.AGE_15;
     private static final VoxelShape SHAPE = Block.column(12.0, 0.0, 16.0);
@@ -57,12 +_,15 @@
 
             if (i < 3) {
                 int j = state.getValue(AGE);
+                if (net.minecraftforge.common.ForgeHooks.onCropsGrowPre(level, pos, state, true)) {
                 if (j == 15) {
                     level.setBlockAndUpdate(pos.above(), this.defaultBlockState());
+                    net.minecraftforge.common.ForgeHooks.onCropsGrowPost(level, pos.above(), this.defaultBlockState());
                     level.setBlock(pos, state.setValue(AGE, 0), 260);
                 } else {
                     level.setBlock(pos, state.setValue(AGE, j + 1), 260);
                 }
+                }
             }
         }
     }
@@ -87,6 +_,8 @@
 
     @Override
     protected boolean canSurvive(final BlockState state, final LevelReader level, final BlockPos pos) {
+        BlockState soil = level.getBlockState(pos.below());
+        if (soil.canSustainPlant(level, pos.below(), Direction.UP, this)) return true;
         BlockState blockstate = level.getBlockState(pos.below());
         if (blockstate.is(this)) {
             return true;
@@ -97,7 +_,7 @@
                 for (Direction direction : Direction.Plane.HORIZONTAL) {
                     BlockState blockstate1 = level.getBlockState(blockpos.relative(direction));
                     FluidState fluidstate = level.getFluidState(blockpos.relative(direction));
-                    if (fluidstate.is(FluidTags.SUPPORTS_SUGAR_CANE_ADJACENTLY) || blockstate1.is(BlockTags.SUPPORTS_SUGAR_CANE_ADJACENTLY)) {
+                    if (fluidstate.is(FluidTags.SUPPORTS_SUGAR_CANE_ADJACENTLY) || blockstate1.is(BlockTags.SUPPORTS_SUGAR_CANE_ADJACENTLY) || state.canBeHydrated(level, pos, fluidstate, blockpos.relative(direction))) {
                         return true;
                     }
                 }
@@ -110,5 +_,15 @@
     @Override
     protected void createBlockStateDefinition(final StateDefinition.Builder<Block, BlockState> builder) {
         builder.add(AGE);
+    }
+
+    @Override
+    public net.minecraftforge.common.PlantType getPlantType(BlockGetter world, BlockPos pos) {
+        return net.minecraftforge.common.PlantType.BEACH;
+    }
+
+    @Override
+    public BlockState getPlant(BlockGetter world, BlockPos pos) {
+       return defaultBlockState();
     }
 }
