--- a/net/minecraft/world/level/levelgen/feature/trunkplacers/TrunkPlacer.java
+++ b/net/minecraft/world/level/levelgen/feature/trunkplacers/TrunkPlacer.java
@@ -63,6 +_,8 @@
     ) {
         BlockState blockBelowTrunk = tree.belowTrunkProvider().value().getOptionalState(level, random, pos);
         if (blockBelowTrunk != null) {
+            var levelReader = (net.minecraft.world.level.LevelReader)level;
+            if (!levelReader.getBlockState(pos).onTreeGrow(levelReader, trunkSetter, random, pos, tree))
             trunkSetter.accept(pos, blockBelowTrunk);
         }
     }
