--- a/net/minecraft/client/renderer/entity/EntityRenderer.java
+++ b/net/minecraft/client/renderer/entity/EntityRenderer.java
@@ -126,16 +_,17 @@
         final S state, final PoseStack poseStack, final SubmitNodeCollector submitNodeCollector, final CameraRenderState camera, final int offset
     ) {
         poseStack.pushPose();
-        if (state.scoreText != null) {
+        var event = net.minecraftforge.client.event.ForgeEventFactoryClient.fireRenderNameTagEvent(state, this, poseStack, submitNodeCollector, camera);
+        if (event.getScoreContent() != null) {
             submitNodeCollector.submitNameTag(
-                poseStack, state.nameTagAttachment, offset, state.scoreText, !state.isDiscrete, state.lightCoords, state.distanceToCameraSq, camera
+                poseStack, state.nameTagAttachment, offset, event.getScoreContent(), !state.isDiscrete, state.lightCoords, state.distanceToCameraSq, camera
             );
             poseStack.translate(0.0F, 9.0F * 1.15F * 0.025F, 0.0F);
         }
 
-        if (state.nameTag != null) {
+        if (event.getContent() != null) {
             submitNodeCollector.submitNameTag(
-                poseStack, state.nameTagAttachment, offset, state.nameTag, !state.isDiscrete, state.lightCoords, state.distanceToCameraSq, camera
+                poseStack, state.nameTagAttachment, offset, event.getContent(), !state.isDiscrete, state.lightCoords, state.distanceToCameraSq, camera
             );
         }
 
@@ -187,7 +_,7 @@
 
         if (this.entityRenderDispatcher.camera != null) {
             state.distanceToCameraSq = this.entityRenderDispatcher.distanceToSqr(entity);
-            boolean flag1 = state.distanceToCameraSq < 4096.0 && this.shouldShowName(entity, state.distanceToCameraSq);
+            boolean flag1 = net.minecraftforge.client.ForgeHooksClient.isNameplateInRenderDistance(entity, state.distanceToCameraSq) && this.shouldShowName(entity, state.distanceToCameraSq);
             if (flag1) {
                 state.nameTag = this.getNameTag(entity);
                 state.nameTagAttachment = entity.getAttachments().getNullable(EntityAttachment.NAME_TAG, 0, entity.getYRot(partialTicks));
