--- a/net/minecraft/world/entity/projectile/Projectile.java
+++ b/net/minecraft/world/entity/projectile/Projectile.java
@@ -363,7 +_,7 @@
     @Override
     public boolean mayInteract(final ServerLevel level, final BlockPos pos) {
         Entity entity = this.getOwner();
-        return entity instanceof Player ? entity.mayInteract(level, pos) : entity == null || level.getGameRules().get(GameRules.MOB_GRIEFING);
+        return entity instanceof Player ? entity.mayInteract(level, pos) : entity == null || net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(level, entity);
     }
 
     public boolean mayBreak(final ServerLevel level) {
