--- a/net/minecraft/world/level/portal/PortalShape.java
+++ b/net/minecraft/world/level/portal/PortalShape.java
@@ -27,7 +_,9 @@
     public static final int MAX_WIDTH = 21;
     private static final int MIN_HEIGHT = 3;
     public static final int MAX_HEIGHT = 21;
+    /** @deprecated Use location sensitive version with FRAME_TEST */
     public static final Predicate<BlockState> FRAME = state -> state.is(BlockTags.NETHER_PORTAL_FRAME);
+    public static final net.minecraft.world.level.block.state.BlockBehaviour.StatePredicate FRAME_TEST = (state, level, pos) -> state.isPortalFrame(level, pos);
     private static final float SAFE_TRAVEL_MAX_ENTITY_XY = 4.0F;
     private static final double SAFE_TRAVEL_MAX_VERTICAL_DELTA = 1.0;
     private final Direction.Axis axis;
@@ -105,14 +_,14 @@
             blockPos.set(pos).move(direction, width);
             BlockState blockState = level.getBlockState(blockPos);
             if (!isEmpty(blockState)) {
-                if (FRAME.test(blockState)) {
+                if (FRAME_TEST.test(blockState, level, blockPos)) {
                     return width;
                 }
                 break;
             }
 
             BlockState belowState = level.getBlockState(blockPos.move(Direction.DOWN));
-            if (!FRAME.test(belowState)) {
+            if (!FRAME_TEST.test(belowState, level, blockPos)) {
                 break;
             }
         }
@@ -133,7 +_,7 @@
     ) {
         for (int i = 0; i < width; i++) {
             BlockPos.MutableBlockPos framePos = pos.set(bottomLeft).move(Direction.UP, height).move(rightDir, i);
-            if (!FRAME.test(level.getBlockState(framePos))) {
+            if (!FRAME_TEST.test(level.getBlockState(framePos), level, framePos)) {
                 return false;
             }
         }
@@ -151,12 +_,12 @@
     ) {
         for (int height = 0; height < 21; height++) {
             pos.set(bottomLeft).move(Direction.UP, height).move(rightDir, -1);
-            if (!FRAME.test(level.getBlockState(pos))) {
+            if (!FRAME_TEST.test(level.getBlockState(pos), level, pos)) {
                 return height;
             }
 
             pos.set(bottomLeft).move(Direction.UP, height).move(rightDir, width);
-            if (!FRAME.test(level.getBlockState(pos))) {
+            if (!FRAME_TEST.test(level.getBlockState(pos), level, pos)) {
                 return height;
             }
 
