--- a/net/minecraft/client/renderer/block/model/SimpleModelWrapper.java
+++ b/net/minecraft/client/renderer/block/model/SimpleModelWrapper.java
@@ -18,7 +_,7 @@
 import org.slf4j.Logger;
 
 @OnlyIn(Dist.CLIENT)
-public record SimpleModelWrapper(QuadCollection quads, boolean useAmbientOcclusion, TextureAtlasSprite particleIcon) implements BlockModelPart {
+public record SimpleModelWrapper(QuadCollection quads, boolean useAmbientOcclusion, TextureAtlasSprite particleIcon, net.minecraft.client.renderer.chunk.ChunkSectionLayer layer, net.minecraft.client.renderer.chunk.ChunkSectionLayer layerFast) implements BlockModelPart {
     private static final Logger LOGGER = LogUtils.getLogger();
 
     public static BlockModelPart bake(ModelBaker p_395631_, Identifier p_457297_, ModelState p_396899_) {
@@ -44,8 +_,13 @@
             LOGGER.warn("Rejecting block model {}, since it contains sprites from outside of supported atlas: {}", p_457297_, multimap);
             return p_395631_.missingBlockModelPart();
         } else {
-            return new SimpleModelWrapper(quadcollection, flag, textureatlassprite);
+            var ctx = resolvedmodel.getContext();
+            return new SimpleModelWrapper(quadcollection, flag, textureatlassprite, ctx.getRenderType().block(), ctx.getRenderTypeFast().block());
         }
+    }
+
+    public SimpleModelWrapper(QuadCollection quadcollection, boolean useAmbientOcclusion, TextureAtlasSprite particleIcon) {
+        this(quadcollection, useAmbientOcclusion, particleIcon, null, null);
     }
 
     @Override
