--- a/net/minecraft/util/context/ContextMap.java
+++ b/net/minecraft/util/context/ContextMap.java
@@ -42,6 +_,10 @@
         return (T)this.params.getOrDefault(param, _default);
     }
 
+    public Set<ContextKey<?>> keySet() {
+        return java.util.Collections.unmodifiableSet(this.params.keySet());
+    }
+
     public static class Builder {
         private final Map<ContextKey<?>, Object> params = new Reference2ObjectOpenHashMap<>();
 
@@ -78,6 +_,10 @@
                     return this.build();
                 }
             }
+        }
+
+        public ContextMap create() {
+            return new ContextMap(this.params);
         }
     }
 }
