--- a/net/minecraft/server/level/ChunkMap.java
+++ b/net/minecraft/server/level/ChunkMap.java
@@ -402,6 +_,7 @@
                 this.modified = true;
             }
 
+            net.minecraftforge.event.ForgeEventFactory.fireChunkTicketLevelUpdated(this.level, node, oldLevel, level, chunk);
             return chunk;
         }
     }
@@ -533,6 +_,7 @@
                 if (this.pendingUnloads.remove(pos, chunkHolder) && chunkaccess != null) {
                     if (chunkaccess instanceof LevelChunk levelchunk) {
                         levelchunk.setLoaded(false);
+                        net.minecraftforge.event.ForgeEventFactory.onChunkUnload(chunkaccess);
                     }
 
                     this.save(chunkaccess);
@@ -775,6 +_,7 @@
                 this.activeChunkWrites.incrementAndGet();
                 SerializableChunkData serializablechunkdata = SerializableChunkData.copyOf(this.level, chunk);
                 CompletableFuture<CompoundTag> completablefuture = CompletableFuture.supplyAsync(serializablechunkdata::write, Util.backgroundExecutor());
+                net.minecraftforge.event.ForgeEventFactory.onChunkDataSave(chunk, chunk.getWorldForge() != null ? chunk.getWorldForge() : this.level, serializablechunkdata);
                 this.write(chunkpos, completablefuture::join).handle((ignored, throwable) -> {
                     if (throwable != null) {
                         this.level.getServer().reportChunkSaveFailure(throwable, this.storageInfo(), chunkpos);
@@ -1153,7 +_,7 @@
     }
 
     protected void addEntity(final Entity entity) {
-        if (!(entity instanceof EnderDragonPart)) {
+        if (!(entity instanceof net.minecraftforge.entity.PartEntity)) {
             EntityType<?> entitytype = entity.getType();
             int i = entitytype.clientTrackingRange() * 16;
             if (i != 0) {
