--- a/net/minecraft/server/level/ServerPlayerGameMode.java
+++ b/net/minecraft/server/level/ServerPlayerGameMode.java
@@ -151,6 +_,10 @@
     public void handleBlockBreakAction(
         final BlockPos pos, final ServerboundPlayerActionPacket.Action action, final Direction direction, final int maxY, final int sequence
     ) {
+        var event = net.minecraftforge.event.ForgeEventFactory.onLeftClickBlock(player, pos, direction, action);
+        if (event == null) {
+           return;
+        }
         if (!this.player.isWithinBlockInteractionRange(pos, 1.0)) {
             this.debugLogging(pos, false, sequence, "too far");
         } else if (pos.getY() > maxY) {
@@ -185,6 +_,7 @@
                 float f = 1.0F;
                 BlockState blockstate = this.level.getBlockState(pos);
                 if (!blockstate.isAir()) {
+                    if (!event.getUseBlock().isDenied()) {
                     EnchantmentHelper.onHitBlock(
                         this.level,
                         this.player.getMainHandItem(),
@@ -196,6 +_,7 @@
                         item -> this.player.onEquippedItemBroken(item, EquipmentSlot.MAINHAND)
                     );
                     blockstate.attack(this.level, pos, this.player);
+                    }
                     f = blockstate.getDestroyProgress(this.player, this.player.level(), pos);
                 }
 
@@ -262,7 +_,8 @@
 
     public boolean destroyBlock(final BlockPos pos) {
         BlockState blockstate1 = this.level.getBlockState(pos);
-        if (!this.player.getMainHandItem().canDestroyBlock(blockstate1, this.level, pos, this.player)) {
+        int exp = net.minecraftforge.common.ForgeHooks.onBlockBreakEvent(level, gameModeForPlayer, player, pos);
+        if (exp == -1) {
             return false;
         } else {
             BlockEntity blockentity = this.level.getBlockEntity(pos);
@@ -270,34 +_,50 @@
             if (block instanceof GameMasterBlock && !this.player.canUseGameMasterBlocks()) {
                 this.level.sendBlockUpdated(pos, blockstate1, blockstate1, 3);
                 return false;
+            } else if (player.getMainHandItem().onBlockStartBreak(pos, player)) {
+               return false;
             } else if (this.player.blockActionRestricted(this.level, pos, this.gameModeForPlayer)) {
                 return false;
             } else {
-                BlockState blockstate = block.playerWillDestroy(this.level, pos, blockstate1, this.player);
-                boolean flag1 = this.level.removeBlock(pos, false);
-                if (SharedConstants.DEBUG_BLOCK_BREAK) {
-                    LOGGER.info("server broke {} {} -> {}", pos, blockstate, this.level.getBlockState(pos));
-                }
-
-                if (flag1) {
-                    block.destroy(this.level, pos, blockstate);
-                }
-
+                BlockState blockstate = blockstate1;
                 if (this.player.preventsBlockDrops()) {
+                    removeBlock(pos, false);
                     return true;
                 } else {
                     ItemStack itemstack = this.player.getMainHandItem();
                     ItemStack itemstack1 = itemstack.copy();
-                    boolean flag = this.player.hasCorrectToolForDrops(blockstate);
+                    boolean flag = blockstate.canHarvestBlock(this.level, pos, this.player); // previously player.hasCorrectToolForDrops(blockstate)
                     itemstack.mineBlock(this.level, blockstate, pos, this.player);
+
+                    if (itemstack.isEmpty() && !itemstack1.isEmpty()) {
+                        net.minecraftforge.event.ForgeEventFactory.onPlayerDestroyItem(this.player, itemstack1, InteractionHand.MAIN_HAND);
+                    }
+                    boolean flag1 = removeBlock(pos, flag);
+
                     if (flag1 && flag) {
                         block.playerDestroy(this.level, this.player, pos, blockstate, blockentity, itemstack1);
                     }
 
+                    if (flag && exp > 0) {
+                        blockstate1.getBlock().popExperience(level, pos, exp);
+                    }
+
                     return true;
                 }
             }
         }
+    }
+
+    private boolean removeBlock(BlockPos pos, boolean canHarvest) {
+        BlockState state = this.level.getBlockState(pos);
+        boolean removed = state.onDestroyedByPlayer(this.level, pos, this.player, canHarvest, this.level.getFluidState(pos));
+        if (removed) {
+            if (SharedConstants.DEBUG_BLOCK_BREAK) {
+                LOGGER.info("server broke {} {} -> {}", pos, state, this.level.getBlockState(pos));
+            }
+            state.getBlock().destroy(this.level, pos, state);
+        }
+        return removed;
     }
 
     public InteractionResult useItem(final ServerPlayer player, final Level level, final ItemStack itemStack, final InteractionHand hand) {
@@ -306,6 +_,8 @@
         } else if (player.getCooldowns().isOnCooldown(itemStack)) {
             return InteractionResult.PASS;
         } else {
+            InteractionResult cancelResult = net.minecraftforge.common.ForgeHooks.onItemRightClick(player, hand);
+            if (cancelResult != null) return cancelResult;
             int i = itemStack.getCount();
             int j = itemStack.getDamageValue();
             InteractionResult interactionresult = itemStack.use(level, player, hand);
@@ -346,6 +_,11 @@
         if (!blockstate.getBlock().isEnabled(level.enabledFeatures())) {
             return InteractionResult.FAIL;
         } else if (this.gameModeForPlayer == GameType.SPECTATOR) {
+        }
+
+        var event = new net.minecraftforge.event.entity.player.PlayerInteractEvent.RightClickBlock(player, hand, blockpos, hitResult);
+        if (net.minecraftforge.event.entity.player.PlayerInteractEvent.RightClickBlock.BUS.post(event)) return event.getCancellationResult();
+        if (this.gameModeForPlayer == GameType.SPECTATOR) {
             MenuProvider menuprovider = blockstate.getMenuProvider(level, blockpos);
             if (menuprovider != null) {
                 player.openMenu(menuprovider);
@@ -354,10 +_,16 @@
                 return InteractionResult.PASS;
             }
         } else {
+            UseOnContext useoncontext = new UseOnContext(player, hand, hitResult);
+            if (!event.getUseItem().isDenied()) {
+                InteractionResult result = itemStack.onItemUseFirst(useoncontext);
+                if (result != InteractionResult.PASS) return result;
+            }
             boolean flag = !player.getMainHandItem().isEmpty() || !player.getOffhandItem().isEmpty();
             boolean flag1 = player.isSecondaryUseActive() && flag;
+            flag1 &= !(player.getMainHandItem().doesSneakBypassUse(level, blockpos, player) && player.getOffhandItem().doesSneakBypassUse(level, blockpos, player));
             ItemStack itemstack = itemStack.copy();
-            if (!flag1) {
+            if (event.getUseBlock().isAllowed() || (!event.getUseBlock().isDenied() && !flag1)) {
                 InteractionResult interactionresult = blockstate.useItemOn(player.getItemInHand(hand), level, player, hand, hitResult);
                 if (interactionresult.consumesAction()) {
                     CriteriaTriggers.ITEM_USED_ON_BLOCK.trigger(player, blockpos, itemstack);
@@ -373,8 +_,8 @@
                 }
             }
 
-            if (!itemStack.isEmpty() && !player.getCooldowns().isOnCooldown(itemStack)) {
-                UseOnContext useoncontext = new UseOnContext(player, hand, hitResult);
+            if (event.getUseItem().isAllowed() || (!itemStack.isEmpty() && !player.getCooldowns().isOnCooldown(itemStack))) {
+                if (event.getUseItem().isDenied()) return InteractionResult.PASS;
                 InteractionResult interactionresult2;
                 if (player.hasInfiniteMaterials()) {
                     int i = itemStack.getCount();
