--- a/net/minecraft/world/item/enchantment/EnchantmentHelper.java
+++ b/net/minecraft/world/item/enchantment/EnchantmentHelper.java
@@ -612,7 +_,7 @@
     public static List<EnchantmentInstance> getAvailableEnchantmentResults(final int value, final ItemStack itemStack, final Stream<Holder<Enchantment>> source) {
         List<EnchantmentInstance> list = Lists.newArrayList();
         boolean flag = itemStack.is(Items.BOOK);
-        source.filter(enchantment -> enchantment.value().isPrimaryItem(itemStack) || flag).forEach(holder -> {
+        source.filter(enchantment -> itemStack.canApplyAtEnchantingTable(enchantment) || flag).forEach(holder -> {
             Enchantment enchantment = holder.value();
 
             for (int i = enchantment.getMaxLevel(); i >= enchantment.getMinLevel(); i--) {
