--- a/net/minecraft/world/item/crafting/BannerDuplicateRecipe.java
+++ b/net/minecraft/world/item/crafting/BannerDuplicateRecipe.java
@@ -98,7 +_,7 @@
         for (int i = 0; i < nonnulllist.size(); i++) {
             ItemStack itemstack = input.getItem(i);
             if (!itemstack.isEmpty()) {
-                ItemStackTemplate itemstacktemplate = itemstack.getItem().getCraftingRemainder();
+                ItemStackTemplate itemstacktemplate = itemstack.getCraftingRemainder();
                 if (itemstacktemplate != null) {
                     nonnulllist.set(i, itemstacktemplate.create());
                 } else if (!itemstack.getOrDefault(DataComponents.BANNER_PATTERNS, BannerPatternLayers.EMPTY).layers().isEmpty()) {
