--- a/net/minecraft/world/item/crafting/BookCloningRecipe.java
+++ b/net/minecraft/world/item/crafting/BookCloningRecipe.java
@@ -131,7 +_,7 @@
 
         for (int i = 0; i < nonnulllist.size(); i++) {
             ItemStack itemstack = input.getItem(i);
-            ItemStackTemplate itemstacktemplate = itemstack.getItem().getCraftingRemainder();
+            ItemStackTemplate itemstacktemplate = itemstack.getCraftingRemainder();
             if (itemstacktemplate != null) {
                 nonnulllist.set(i, itemstacktemplate.create());
             } else if (itemstack.has(DataComponents.WRITTEN_BOOK_CONTENT)) {
