--- a/net/minecraft/world/item/crafting/RecipeManager.java
+++ b/net/minecraft/world/item/crafting/RecipeManager.java
@@ -63,15 +_,22 @@
     private SelectableRecipe.SingleInputSet<StonecutterRecipe> stonecutterRecipes = SelectableRecipe.SingleInputSet.empty();
     private List<RecipeManager.ServerDisplayInfo> allDisplays = List.of();
     private Map<ResourceKey<Recipe<?>>, List<RecipeManager.ServerDisplayInfo>> recipeToDisplay = Map.of();
+    private final net.minecraftforge.common.crafting.conditions.ICondition.IContext context; //Forge: add context
 
+    /** @deprecated Forge: use {@linkplain RecipeManager#RecipeManager(net.minecraftforge.common.crafting.conditions.ICondition.IContext) constructor with context}. */
     public RecipeManager(final HolderLookup.Provider registries) {
+        this(registries, net.minecraftforge.common.crafting.conditions.ICondition.IContext.EMPTY);
+    }
+
+    public RecipeManager(HolderLookup.Provider registries, net.minecraftforge.common.crafting.conditions.ICondition.IContext context) {
         this.registries = registries;
+        this.context = context;
     }
 
     protected RecipeMap prepare(final ResourceManager manager, final ProfilerFiller profiler) {
         SortedMap<Identifier, Recipe<?>> sortedmap = new TreeMap<>();
         SimpleJsonResourceReloadListener.scanDirectory(
-            manager, RECIPE_LISTER, this.registries.createSerializationContext(JsonOps.INSTANCE), Recipe.CODEC, sortedmap
+            manager, RECIPE_LISTER, this.context.wrap(this.registries.createSerializationContext(JsonOps.INSTANCE)), Recipe.CODEC, sortedmap
         );
         List<RecipeHolder<?>> list = new ArrayList<>(sortedmap.size());
         sortedmap.forEach((id, recipe) -> {
@@ -88,6 +_,7 @@
     }
 
     public void finalizeRecipeLoading(final FeatureFlagSet enabledFlags) {
+        //net.minecraftforge.event.ForgeEventFactory.onTagsUpdated(this.registries, false, false);
         List<SelectableRecipe.SingleInputEntry<StonecutterRecipe>> list = new ArrayList<>();
         List<RecipeManager.IngredientCollector> list1 = RECIPE_PROPERTY_SETS.entrySet()
             .stream()
