--- a/net/minecraft/server/dedicated/ServerWatchdog.java
+++ b/net/minecraft/server/dedicated/ServerWatchdog.java
@@ -78,7 +_,7 @@
         ThreadMXBean threadMXBean = ManagementFactory.getThreadMXBean();
         ThreadInfo[] threadInfos = threadMXBean.dumpAllThreads(true, true);
         StringBuilder builder = new StringBuilder();
-        Error exception = new Error("Watchdog");
+        Error exception = new Error(String.format(java.util.Locale.ENGLISH, "ServerHangWatchdog detected that a single server tick took too long")); // Forge: don't just make a crash report with a seemingly-inexplicable Error
 
         for (ThreadInfo threadInfo : threadInfos) {
             if (threadInfo.getThreadId() == mainThreadId) {
