--- a/net/minecraft/world/entity/ai/behavior/StartAttacking.java
+++ b/net/minecraft/world/entity/ai/behavior/StartAttacking.java
@@ -29,7 +_,11 @@
                             if (!body.canAttack(targetEntity)) {
                                 return false;
                             } else {
-                                attackTarget.set(targetEntity);
+                                var changeTargetEvent = net.minecraftforge.event.ForgeEventFactory.onLivingChangeTargetBehavior(body, targetEntity);
+                                if (changeTargetEvent == null)
+                                    return false;
+
+                                attackTarget.set(changeTargetEvent.getNewTarget());
                                 cantReachSince.erase();
                                 return true;
                             }
