--- a/net/minecraft/world/entity/ai/behavior/Swim.java
+++ b/net/minecraft/world/entity/ai/behavior/Swim.java
@@ -22,7 +_,7 @@
     }
 
     protected boolean checkExtraStartConditions(final ServerLevel level, final Mob body) {
-        return body.isInFluidDeeperThan(body.getFluidJumpThreshold(), this.fluid) || body.isInLava();
+        return body.isInFluidDeeperThan(body.getFluidJumpThreshold(), this.fluid) || body.isInLava() || body.isInFluidType((fluidType, height) -> body.canSwimInFluidType(fluidType) && height > body.getFluidJumpThreshold());
     }
 
     protected boolean canStillUse(final ServerLevel level, final Mob body, final long timestamp) {
