--- a/net/minecraft/client/gui/screens/inventory/EffectsInInventory.java
+++ b/net/minecraft/client/gui/screens/inventory/EffectsInInventory.java
@@ -47,12 +_,16 @@
         int j = this.screen.width - i;
         Collection<MobEffectInstance> collection = this.minecraft.player.getActiveEffects();
         if (!collection.isEmpty() && j >= 32) {
-            int k = j >= 120 ? j - 7 : 32;
+            var event = net.minecraftforge.client.event.ForgeEventFactoryClient.onScreenEffectSize(this.screen, j, j < 120, i);
+            if (event == null) return;
+            int k = !event.isCompact() ? j - 7 : 32;
+            i = event.getHorizontalOffset();
             int l = 33;
             if (collection.size() > 5) {
                 l = 132 / (collection.size() - 1);
             }
 
+            Iterable<MobEffectInstance> iterable = collection.stream().filter(net.minecraftforge.client.ForgeHooksClient::shouldRenderEffect).sorted().toList();
             this.renderEffects(p_456221_, collection, i, l, p_453611_, p_457243_, k);
         }
     }
@@ -66,6 +_,11 @@
 
         for (MobEffectInstance mobeffectinstance : iterable) {
             boolean flag = mobeffectinstance.isAmbient();
+            var renderer = net.minecraftforge.client.extensions.common.IClientMobEffectExtensions.of(mobeffectinstance);
+            if (renderer.renderInventory(mobeffectinstance, this, p_362146_, p_370153_, i, 0)) {
+                i += p_365612_;
+                continue;
+            }
             Component component = this.getEffectName(mobeffectinstance);
             Component component1 = MobEffectUtil.formatDuration(mobeffectinstance, 1.0F, this.minecraft.level.tickRateManager().tickrate());
             int j = this.renderBackground(p_362146_, font, component, component1, p_370153_, i, flag, p_457235_);
@@ -123,5 +_,9 @@
         }
 
         return mutablecomponent;
+    }
+
+    public AbstractContainerScreen<?> getScreen() {
+        return this.screen;
     }
 }
