--- a/net/minecraft/client/gui/screens/inventory/EffectsInInventory.java
+++ b/net/minecraft/client/gui/screens/inventory/EffectsInInventory.java
@@ -47,12 +_,16 @@
         int availableWidth = this.screen.width - xo;
         Collection<MobEffectInstance> activeEffects = this.minecraft.player.getActiveEffects();
         if (!activeEffects.isEmpty() && availableWidth >= 32) {
-            int maxWidth = availableWidth >= 120 ? availableWidth - 7 : 32;
+            var event = net.minecraftforge.client.event.ForgeEventFactoryClient.onScreenEffectSize(this.screen, availableWidth, availableWidth < 120, xo);
+            if (event == null) return;
+            int maxWidth = !event.isCompact() ? availableWidth - 7 : 32;
+            xo = event.getHorizontalOffset();
             int yStep = 33;
             if (activeEffects.size() > 5) {
                 yStep = 132 / (activeEffects.size() - 1);
             }
 
+            Iterable<MobEffectInstance> iterable = activeEffects.stream().filter(net.minecraftforge.client.ForgeHooksClient::shouldRenderEffect).sorted().toList();
             this.extractEffects(graphics, activeEffects, xo, yStep, mouseX, mouseY, maxWidth);
         }
     }
@@ -72,6 +_,11 @@
 
         for (MobEffectInstance effect : sortedEffects) {
             boolean isAmbient = effect.isAmbient();
+            var renderer = net.minecraftforge.client.extensions.common.IClientMobEffectExtensions.of(effect);
+            if (renderer.extractInventory(effect, this, graphics, x0, y0, 0)) {
+                y0 += yStep;
+                continue;
+            }
             Component effectText = this.getEffectName(effect);
             Component duration = MobEffectUtil.formatDuration(effect, 1.0F, this.minecraft.level.tickRateManager().tickrate());
             int textureWidth = this.extractBackground(graphics, font, effectText, duration, x0, y0, isAmbient, maxWidth);
@@ -136,5 +_,9 @@
         }
 
         return name;
+    }
+
+    public AbstractContainerScreen<?> getScreen() {
+        return this.screen;
     }
 }
