--- a/net/minecraft/world/level/chunk/status/ChunkStatusTasks.java
+++ b/net/minecraft/world/level/chunk/status/ChunkStatusTasks.java
@@ -201,9 +_,15 @@
             levelChunk.setFullStatus(holder::getFullStatus);
             levelChunk.runPostLoad();
             levelChunk.setLoaded(true);
+            try {
+                holder.currentlyLoading = levelChunk; // Forge - bypass the future chain when getChunk is called, this prevents deadlocks.
             levelChunk.registerAllBlockEntitiesAfterLevelLoad();
             levelChunk.registerTickContainerInLevel(level);
             levelChunk.setUnsavedListener(context.unsavedListener());
+                net.minecraftforge.event.ForgeEventFactory.onChunkLoad(levelChunk, !(protoChunk instanceof ImposterProtoChunk));
+            } finally {
+                holder.currentlyLoading = null; // Forge - Stop bypassing the future chain.
+            }
             return levelChunk;
         }, context.mainThreadExecutor());
     }
