--- a/net/minecraft/client/multiplayer/ClientLevel.java
+++ b/net/minecraft/client/multiplayer/ClientLevel.java
@@ -150,6 +_,7 @@
         cache.put(BiomeColors.FOLIAGE_COLOR_RESOLVER, new BlockTintCache(pos -> this.calculateBlockTint(pos, BiomeColors.FOLIAGE_COLOR_RESOLVER)));
         cache.put(BiomeColors.DRY_FOLIAGE_COLOR_RESOLVER, new BlockTintCache(pos -> this.calculateBlockTint(pos, BiomeColors.DRY_FOLIAGE_COLOR_RESOLVER)));
         cache.put(BiomeColors.WATER_COLOR_RESOLVER, new BlockTintCache(pos -> this.calculateBlockTint(pos, BiomeColors.WATER_COLOR_RESOLVER)));
+        net.minecraftforge.client.ColorResolverManager.registerBlockTintCaches(ClientLevel.this, cache);
     });
     private final ClientChunkCache chunkSource;
     private final Deque<Runnable> lightUpdateQueue = Queues.newArrayDeque();
@@ -161,6 +_,8 @@
     private final EnvironmentAttributeSystem environmentAttributes;
     private final int seaLevel;
     private static final Set<Item> MARKER_PARTICLE_ITEMS = Set.of(Items.BARRIER, Items.LIGHT);
+    private final it.unimi.dsi.fastutil.ints.Int2ObjectMap<net.minecraftforge.entity.PartEntity<?>> partEntities = new it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap<>();
+    private final net.minecraftforge.client.model.data.ModelDataManager modelDataManager = new net.minecraftforge.client.model.data.ModelDataManager(this);
 
     public void handleBlockChangedAck(final int sequence) {
         if (SharedConstants.DEBUG_BLOCK_BREAK) {
@@ -273,6 +_,8 @@
 
             runnable.run();
         }
+        this.gatherCapabilities();
+        net.minecraftforge.event.ForgeEventFactory.onLevelLoad(this);
     }
 
     public @Nullable EndFlashState endFlashState() {
@@ -352,6 +_,7 @@
         entity.setOldPosAndRot();
         entity.tickCount++;
         Profiler.get().push(entity.typeHolder()::getRegisteredName);
+        if (entity.canUpdate())
         entity.tick();
         Profiler.get().pop();
 
@@ -412,8 +_,10 @@
     }
 
     public void addEntity(final Entity entity) {
+        if (net.minecraftforge.event.ForgeEventFactory.onEntityJoinLevel(entity, this)) return;
         this.removeEntity(entity.getId(), Entity.RemovalReason.DISCARDED);
         this.entityStorage.addEntity(entity);
+        entity.onAddedToLevel();
     }
 
     public void removeEntity(final int id, final Entity.RemovalReason reason) {
@@ -579,8 +_,10 @@
         final float pitch,
         final long seed
     ) {
+        var event = net.minecraftforge.event.ForgeEventFactory.onPlaySoundAtPosition(this, x, y, z, sound, source, volume, pitch);
+        if (event == null || event.getSound() == null) return;
         if (except == this.minecraft.player) {
-            this.playSound(x, y, z, sound.value(), source, volume, pitch, false, seed);
+            this.playSound(x, y, z, event.getSound().value(), event.getSource(), event.getNewVolume(), event.getNewPitch(), false, seed);
         }
     }
 
@@ -594,8 +_,10 @@
         final float pitch,
         final long seed
     ) {
+        var event = net.minecraftforge.event.ForgeEventFactory.onPlaySoundAtEntity(this, except, sound, source, volume, pitch);
+        if (event == null || event.getSound() == null) return;
         if (except == this.minecraft.player) {
-            this.minecraft.getSoundManager().play(new EntityBoundSoundInstance(sound.value(), source, volume, pitch, sourceEntity, seed));
+            this.minecraft.getSoundManager().play(new EntityBoundSoundInstance(event.getSound().value(), event.getSource(), event.getNewVolume(), event.getNewPitch(), sourceEntity, seed));
         }
     }
 
@@ -951,7 +_,7 @@
 
     @Override
     public void addDestroyBlockEffect(final BlockPos pos, final BlockState blockState) {
-        if (!blockState.isAir() && blockState.shouldSpawnTerrainParticles()) {
+        if (!blockState.isAir() && !net.minecraftforge.client.extensions.common.IClientBlockExtensions.of(blockState).addDestroyEffects(blockState, this, pos, this.minecraft.particleEngine)) {
             VoxelShape voxelshape = blockState.getShape(this, pos);
             double d0 = 0.25;
             voxelshape.forAllBoxes(
@@ -978,6 +_,7 @@
                                         new TerrainParticle(
                                             this, pos.getX() + d7, pos.getY() + d8, pos.getZ() + d9, d4 - 0.5, d5 - 0.5, d6 - 0.5, blockState, pos
                                         )
+                                        .updateSprite(blockState, pos)
                                     );
                             }
                         }
@@ -987,6 +_,13 @@
         }
     }
 
+    public void addBreakingBlockEffect(BlockPos pos, net.minecraft.world.phys.BlockHitResult target) {
+        var state = this.getBlockState(pos);
+        if (!net.minecraftforge.client.extensions.common.IClientBlockExtensions.of(state).addHitEffects(state, this, target, this.minecraft.particleEngine))
+            addBreakingBlockEffect(pos, target.getDirection());
+    }
+
+    /** @deprecated Forge - Use crack(BlockPos, BlockHitResult) as it has more context and gives more control to modders */
     public void addBreakingBlockEffect(final BlockPos pos, final Direction direction) {
         BlockState blockstate = this.getBlockState(pos);
         if (blockstate.getRenderShape() != RenderShape.INVISIBLE && blockstate.shouldSpawnTerrainParticles()) {
@@ -1022,7 +_,7 @@
                 d0 = i + aabb.maxX + 0.1F;
             }
 
-            this.minecraft.particleEngine.add(new TerrainParticle(this, d0, d1, d2, 0.0, 0.0, 0.0, blockstate, pos).setPower(0.2F).scale(0.6F));
+            this.minecraft.particleEngine.add(new TerrainParticle(this, d0, d1, d2, 0.0, 0.0, 0.0, blockstate, pos).updateSprite(blockstate, pos).setPower(0.2F).scale(0.6F));
         }
     }
 
@@ -1088,6 +_,16 @@
         this.explosionTracker.track(center, radius, blockCount, blockParticles);
     }
 
+    @Override
+    public java.util.Collection<net.minecraftforge.entity.PartEntity<?>> getPartEntities() {
+       return this.partEntities.values();
+    }
+
+    @Override
+    public net.minecraftforge.client.model.data.ModelDataManager getModelDataManager() {
+       return modelDataManager;
+    }
+
     @OnlyIn(Dist.CLIENT)
     public static class ClientLevelData implements WritableLevelData {
         private final boolean hardcore;
@@ -1143,6 +_,7 @@
         }
 
         public void setDifficulty(final Difficulty difficulty) {
+            net.minecraftforge.event.ForgeEventFactory.onDifficultyChange(difficulty, this.difficulty);
             this.difficulty = difficulty;
         }
 
@@ -1190,6 +_,12 @@
                     break;
                 default:
             }
+
+            if (entity.isMultipartEntity()) {
+               for (var part : entity.getParts()) {
+                  ClientLevel.this.partEntities.put(part.getId(), part);
+               }
+            }
         }
 
         public void onTrackingEnd(final Entity entity) {
@@ -1202,6 +_,15 @@
                     ClientLevel.this.dragonParts.removeAll(Arrays.asList(enderdragon.getSubEntities()));
                     break;
                 default:
+            }
+
+            entity.onRemovedFromLevel();
+            net.minecraftforge.event.ForgeEventFactory.onEntityLeaveLevel(entity, ClientLevel.this);
+
+            if (entity.isMultipartEntity()) {
+               for (var part : entity.getParts()) {
+                  ClientLevel.this.partEntities.remove(part.getId());
+               }
             }
         }
 
