--- a/net/minecraft/client/multiplayer/SessionSearchTrees.java
+++ b/net/minecraft/client/multiplayer/SessionSearchTrees.java
@@ -33,8 +_,8 @@
     private static final SessionSearchTrees.Key RECIPE_COLLECTIONS = new SessionSearchTrees.Key();
     public static final SessionSearchTrees.Key CREATIVE_NAMES = new SessionSearchTrees.Key();
     public static final SessionSearchTrees.Key CREATIVE_TAGS = new SessionSearchTrees.Key();
-    private CompletableFuture<SearchTree<ItemStack>> creativeByNameSearch = CompletableFuture.completedFuture(SearchTree.empty());
-    private CompletableFuture<SearchTree<ItemStack>> creativeByTagSearch = CompletableFuture.completedFuture(SearchTree.empty());
+    private CompletableFuture<SearchTree<ItemStack>> EMPTY = CompletableFuture.completedFuture(SearchTree.empty());
+    private Map<SessionSearchTrees.Key, CompletableFuture<SearchTree<ItemStack>>> creativeSearch = new IdentityHashMap<>();
     private CompletableFuture<SearchTree<RecipeCollection>> recipeSearch = CompletableFuture.completedFuture(SearchTree.empty());
     private final Map<SessionSearchTrees.Key, Runnable> reloaders = new IdentityHashMap<>();
 
@@ -89,44 +_,52 @@
     }
 
     public void updateCreativeTags(final List<ItemStack> items) {
+        for (var entry : net.minecraftforge.client.CreativeModeTabSearchRegistry.getTagSearchKeys().entrySet())
         this.register(
-            CREATIVE_TAGS,
+            entry.getValue(),
             () -> {
-                CompletableFuture<?> previous = this.creativeByTagSearch;
-                this.creativeByTagSearch = CompletableFuture.supplyAsync(
-                    () -> new IdSearchTree<>(itemStack -> itemStack.tags().map(TagKey::location), items), Util.backgroundExecutor()
-                );
+                var tabItems = entry.getKey() == net.minecraft.world.item.CreativeModeTabs.searchTab() ? items : List.copyOf(entry.getKey().getDisplayItems());
+                CompletableFuture<?> previous = this.creativeSearch.getOrDefault(entry.getValue(), EMPTY);
+                this.creativeSearch.put(entry.getValue(), CompletableFuture.supplyAsync(
+                    () -> new IdSearchTree<>(itemStack -> itemStack.tags().map(TagKey::location), tabItems), Util.backgroundExecutor()
+                ));
                 previous.cancel(true);
             }
         );
     }
 
     public SearchTree<ItemStack> creativeTagSearch() {
-        return this.creativeByTagSearch.join();
+        return this.getSearchTree(CREATIVE_TAGS);
     }
 
     public void updateCreativeTooltips(final HolderLookup.Provider registries, final List<ItemStack> itemStacks) {
+        for (var entry : net.minecraftforge.client.CreativeModeTabSearchRegistry.getNameSearchKeys().entrySet())
         this.register(
-            CREATIVE_NAMES,
+            entry.getValue(),
             () -> {
+                var items = entry.getKey() == net.minecraft.world.item.CreativeModeTabs.searchTab() ? itemStacks : List.copyOf(entry.getKey().getDisplayItems());
                 Item.TooltipContext tooltipContext = Item.TooltipContext.of(registries);
                 TooltipFlag tooltipFlag = TooltipFlag.Default.NORMAL.asCreative();
-                CompletableFuture<?> previous = this.creativeByNameSearch;
-                this.creativeByNameSearch = CompletableFuture.supplyAsync(
+                CompletableFuture<?> previous = this.creativeSearch.getOrDefault(entry.getValue(), EMPTY);
+                this.creativeSearch.put(entry.getValue(), CompletableFuture.supplyAsync(
                     () -> new FullTextSearchTree<>(
                         itemStack -> getTooltipLines(Stream.of(itemStack), tooltipContext, tooltipFlag),
                         itemStack -> itemStack.typeHolder().unwrapKey().map(ResourceKey::identifier).stream(),
-                        itemStacks
+                        items
                     ),
                     Util.backgroundExecutor()
-                );
+                ));
                 previous.cancel(true);
             }
         );
     }
 
     public SearchTree<ItemStack> creativeNameSearch() {
-        return this.creativeByNameSearch.join();
+        return this.getSearchTree(CREATIVE_NAMES);
+    }
+
+    public SearchTree<ItemStack> getSearchTree(SessionSearchTrees.Key key) {
+        return this.creativeSearch.getOrDefault(key, EMPTY).join();
     }
 
     @OnlyIn(Dist.CLIENT)
