--- a/net/minecraft/world/effect/MobEffect.java
+++ b/net/minecraft/world/effect/MobEffect.java
@@ -36,7 +_,8 @@
 import net.minecraft.world.flag.FeatureFlags;
 import org.jspecify.annotations.Nullable;
 
-public class MobEffect implements FeatureElement {
+
+public class MobEffect implements FeatureElement, net.minecraftforge.common.extensions.IForgeMobEffect {
     public static final Codec<Holder<MobEffect>> CODEC = BuiltInRegistries.MOB_EFFECT.holderByNameCodec();
     public static final StreamCodec<RegistryFriendlyByteBuf, Holder<MobEffect>> STREAM_CODEC = ByteBufCodecs.holderRegistry(Registries.MOB_EFFECT);
     private static final int AMBIENT_ALPHA = Mth.floor(38.25F);
@@ -58,12 +_,14 @@
             int i = effectInstance.isAmbient() ? AMBIENT_ALPHA : 255;
             return ColorParticleOption.create(ParticleTypes.ENTITY_EFFECT, ARGB.color(i, color));
         };
+        initClient();
     }
 
     protected MobEffect(final MobEffectCategory category, final int color, final ParticleOptions particleOptions) {
         this.category = category;
         this.color = color;
         this.particleFactory = ignored -> particleOptions;
+        initClient();
     }
 
     public int getBlendInDurationTicks() {
@@ -207,5 +_,23 @@
         public AttributeModifier create(final int amplifier) {
             return new AttributeModifier(this.id, this.amount * (amplifier + 1), this.operation);
         }
+    }
+
+    private Object effectRenderer;
+    @org.jetbrains.annotations.ApiStatus.Internal
+    public Object getEffectRendererInternal() {
+        return effectRenderer;
+    }
+
+    private void initClient() {
+        // Minecraft instance isn't available in datagen, so don't call initializeClient if in datagen
+        if (net.minecraftforge.fml.loading.FMLEnvironment.dist == net.minecraftforge.api.distmarker.Dist.CLIENT && !net.minecraftforge.fml.loading.FMLLoader.getLaunchHandler().isData()) {
+            initializeClient(properties -> {
+                this.effectRenderer = properties;
+            });
+        }
+    }
+
+    public void initializeClient(java.util.function.Consumer<net.minecraftforge.client.extensions.common.IClientMobEffectExtensions> consumer) {
     }
 }
