--- a/net/minecraft/world/level/block/grower/TreeGrower.java
+++ b/net/minecraft/world/level/block/grower/TreeGrower.java
@@ -114,6 +_,9 @@
         ResourceKey<Feature> megaFeatureKey = this.getConfiguredMegaFeature(random);
         if (megaFeatureKey != null) {
             Holder<Feature> featureHolder = level.registryAccess().lookupOrThrow(Registries.FEATURE).get(megaFeatureKey).orElse(null);
+            var event = net.minecraftforge.event.ForgeEventFactory.blockGrowFeature(level, random, pos, featureHolder);
+            featureHolder = event.getFeature();
+            if (event.getResult().isDenied()) return false;
             if (featureHolder != null) {
                 Optional<TreeGrower.TwoByTwoSaplingPos> twoByTwoSaplingPos = findTwoByTwoSaplingPos(level, state, pos);
                 if (twoByTwoSaplingPos.isPresent()) {
