--- a/net/minecraft/world/level/pathfinder/WalkNodeEvaluator.java
+++ b/net/minecraft/world/level/pathfinder/WalkNodeEvaluator.java
@@ -484,6 +_,16 @@
                 for (int k = -1; k <= 1; k++) {
                     if (i != 0 || k != 0) {
                         PathType pathtype = context.getPathTypeFromState(x + i, y + j, z + k);
+
+                        var pos = new BlockPos(x + i, y + j, z + k);
+                        var blockstate = context.level().getBlockState(pos);
+
+                        var adjacentBlockPathType = blockstate.getAdjacentBlockPathType(context.level(), pos, null, blockPathType);
+                        if (adjacentBlockPathType != null) return adjacentBlockPathType;
+
+                        var adjacentFluidPathType = blockstate.getFluidState().getAdjacentBlockPathType(context.level(), pos, null, blockPathType);
+                        if (adjacentFluidPathType != null) return adjacentFluidPathType;
+
                         if (pathtype == PathType.DAMAGING) {
                             return PathType.DAMAGING_IN_NEIGHBOR;
                         }
@@ -510,6 +_,10 @@
     protected static PathType getPathTypeFromState(final BlockGetter level, final BlockPos pos) {
         BlockState blockstate = level.getBlockState(pos);
         Block block = blockstate.getBlock();
+
+        var type = blockstate.getBlockPathType(level, pos, null);
+        if (type != null) return type;
+
         if (blockstate.isAir()) {
             return PathType.OPEN;
         } else if (blockstate.is(BlockTags.TRAPDOORS) || blockstate.is(Blocks.LILY_PAD) || blockstate.is(Blocks.BIG_DRIPLEAF)) {
@@ -524,6 +_,8 @@
             return PathType.COCOA;
         } else if (!blockstate.is(Blocks.WITHER_ROSE) && !blockstate.is(Blocks.POINTED_DRIPSTONE)) {
             FluidState fluidstate = blockstate.getFluidState();
+            var nonLoggableFluidPathType = fluidstate.getBlockPathType(level, pos, null, false);
+            if (nonLoggableFluidPathType != null) return nonLoggableFluidPathType;
             if (fluidstate.is(FluidTags.LAVA)) {
                 return PathType.LAVA;
             } else if (isBurningBlock(blockstate)) {
@@ -544,6 +_,8 @@
                 if (!blockstate.isPathfindable(PathComputationType.LAND)) {
                     return PathType.BLOCKED;
                 } else {
+                    var loggableFluidPathType = fluidstate.getBlockPathType(level, pos, null, true);
+                    if (loggableFluidPathType != null) return loggableFluidPathType;
                     return fluidstate.is(FluidTags.WATER) ? PathType.WATER : PathType.OPEN;
                 }
             } else {
