--- a/net/minecraft/client/gui/Gui.java
+++ b/net/minecraft/client/gui/Gui.java
@@ -224,12 +_,6 @@
             LOGGER.error("setScreen called from non-game thread");
         }
 
-        if (this.screen != null) {
-            this.screen.removed();
-        } else {
-            this.minecraft.setLastInputType(InputType.NONE);
-        }
-
         if (screen == null) {
             if (this.clientLevelTeardownInProgress) {
                 throw new IllegalStateException("Trying to return to in-game GUI during disconnection");
@@ -248,6 +_,22 @@
             }
         }
 
+        net.minecraftforge.client.ForgeHooksClient.clearGuiLayers(this);
+        Screen old = this.screen;
+        if (screen != null) {
+           var event = net.minecraftforge.client.event.ForgeEventFactoryClient.onScreenOpening(old, screen);
+           if (event == null) return;
+           screen = event;
+        }
+
+        if (screen != old) {
+            if (old != null) {
+                net.minecraftforge.client.event.ForgeEventFactoryClient.onScreenClose(old);
+                old.removed();
+            } else {
+                this.minecraft.setLastInputType(InputType.NONE);
+            }
+        }
         this.screen = screen;
         if (this.screen != null) {
             this.screen.added();
