--- a/net/minecraft/client/particle/BlockMarker.java
+++ b/net/minecraft/client/particle/BlockMarker.java
@@ -13,7 +_,7 @@
     private final SingleQuadParticle.Layer layer;
 
     private BlockMarker(final ClientLevel level, final double x, final double y, final double z, final BlockState state) {
-        super(level, x, y, z, Minecraft.getInstance().getModelManager().getBlockStateModelSet().getParticleMaterial(state).sprite());
+        super(level, x, y, z, Minecraft.getInstance().getModelManager().getBlockStateModelSet().get(state).particleMaterial(level.getModelDataManager().getAtOrEmpty(x, y, z)).sprite());
         this.gravity = 0.0F;
         this.lifetime = 80;
         this.hasPhysics = false;
