--- a/net/minecraft/client/particle/TerrainParticle.java
+++ b/net/minecraft/client/particle/TerrainParticle.java
@@ -43,7 +_,7 @@
         this.gCol = 0.6F;
         this.bCol = 0.6F;
         BlockTintSource blocktintsource = Minecraft.getInstance().getBlockColors().getTintSource(blockState, 0);
-        if (blocktintsource != null) {
+        if (blocktintsource != null && net.minecraftforge.client.extensions.common.IClientBlockExtensions.of(blockState).areBreakingParticlesTinted(blockState, level, pos)) {
             int i = blocktintsource.colorAsTerrainParticle(blockState, level, pos);
             this.rCol *= (i >> 16 & 0xFF) / 255.0F;
             this.gCol *= (i >> 8 & 0xFF) / 255.0F;
@@ -93,8 +_,14 @@
     ) {
         BlockState blockstate = options.getState();
         return !blockstate.isAir() && !blockstate.is(Blocks.MOVING_PISTON) && blockstate.shouldSpawnTerrainParticles()
-            ? new TerrainParticle(level, x, y, z, xAux, yAux, zAux, blockstate)
+            ? (TerrainParticle)new TerrainParticle(level, x, y, z, xAux, yAux, zAux, blockstate).updateSprite(blockstate, options.getPos())
             : null;
+    }
+
+    public Particle updateSprite(BlockState state, BlockPos pos) { //FORGE: we cannot assume that the x y z of the particles match the block pos of the block.
+        if (pos != null) // There are cases where we are not able to obtain the correct source pos, and need to fallback to the non-model data version
+           this.setSprite(Minecraft.getInstance().getModelManager().getBlockStateModelSet().get(state).particleMaterial(level.getModelDataManager().getAt(pos)).sprite());
+        return this;
     }
 
     @OnlyIn(Dist.CLIENT)
