--- a/net/minecraft/world/entity/decoration/HangingEntity.java
+++ b/net/minecraft/world/entity/decoration/HangingEntity.java
@@ -84,7 +_,9 @@
             return false;
         }
 
-        boolean isSupported = this.level().findBlocksIn(this.calculateSupportBox()).filterState(HangingEntity::isSupportingBlock).allMatched();
+        boolean isSupported = !this.level().findBlocksIn(this.calculateSupportBox()).forEachUntil((pos, state) ->
+            net.minecraft.util.Continuation.continueIf(isSupportingBlock(state) || net.minecraft.world.level.block.Block.canSupportCenter(this.level(), pos, this.getDirection()))
+        );
         return isSupported && this.canCoexist(false);
     }
 
