--- a/net/minecraft/client/renderer/texture/atlas/SpriteResourceLoader.java
+++ b/net/minecraft/client/renderer/texture/atlas/SpriteResourceLoader.java
@@ -68,6 +_,9 @@
                 frameSize = new FrameSize(image.getWidth(), image.getHeight());
             }
 
+            SpriteContents contents = net.minecraftforge.client.ForgeHooksClient.loadSpriteContents(spriteLocation, resource, frameSize, image, additionalMetadata);
+            if (contents != null) return contents;
+
             return new SpriteContents(spriteLocation, frameSize, image, animationInfo, additionalMetadata, textureInfo);
         };
     }
