--- a/net/minecraft/client/renderer/texture/atlas/SpriteResourceLoader.java
+++ b/net/minecraft/client/renderer/texture/atlas/SpriteResourceLoader.java
@@ -68,6 +_,9 @@
                 framesize = new FrameSize(nativeimage.getWidth(), nativeimage.getHeight());
             }
 
+            SpriteContents contents = net.minecraftforge.client.ForgeHooksClient.loadSpriteContents(spriteLocation, resource, framesize, nativeimage, list);
+            if (contents != null) return contents;
+
             return new SpriteContents(spriteLocation, framesize, nativeimage, optional, list, optional1);
         };
     }
