--- a/net/minecraft/tags/EntityTypeTags.java
+++ b/net/minecraft/tags/EntityTypeTags.java
@@ -55,4 +_,12 @@
     private static TagKey<EntityType<?>> create(String p_203849_) {
         return TagKey.create(Registries.ENTITY_TYPE, Identifier.withDefaultNamespace(p_203849_));
     }
+
+    public static TagKey<EntityType<?>> create(String namepsace, String path) {
+        return create(Identifier.fromNamespaceAndPath(namepsace, path));
+    }
+
+    public static TagKey<EntityType<?>> create(Identifier name) {
+        return TagKey.create(Registries.ENTITY_TYPE, name);
+    }
 }
