--- a/net/minecraft/tags/PoiTypeTags.java
+++ b/net/minecraft/tags/PoiTypeTags.java
@@ -15,4 +_,12 @@
     private static TagKey<PoiType> create(final String name) {
         return TagKey.create(Registries.POINT_OF_INTEREST_TYPE, Identifier.withDefaultNamespace(name));
     }
+
+    public static TagKey<PoiType> create(String namepsace, String path) {
+        return create(Identifier.fromNamespaceAndPath(namepsace, path));
+    }
+
+    public static TagKey<PoiType> create(Identifier name) {
+        return TagKey.create(Registries.POINT_OF_INTEREST_TYPE, name);
+    }
 }
