--- a/net/minecraft/tags/StructureTags.java
+++ b/net/minecraft/tags/StructureTags.java
@@ -29,4 +_,12 @@
     private static TagKey<Structure> create(final String name) {
         return TagKey.create(Registries.STRUCTURE, Identifier.withDefaultNamespace(name));
     }
+
+    public static TagKey<Structure> create(String namepsace, String path) {
+        return create(Identifier.fromNamespaceAndPath(namepsace, path));
+    }
+
+    public static TagKey<Structure> create(Identifier name) {
+        return TagKey.create(Registries.STRUCTURE, name);
+    }
 }
