--- a/net/minecraft/tags/TagEntry.java
+++ b/net/minecraft/tags/TagEntry.java
@@ -111,6 +_,18 @@
         return stringbuilder.toString();
     }
 
+    public Identifier getId() {
+        return id;
+    }
+
+    public boolean isRequired() {
+        return required;
+    }
+
+    public boolean isTag() {
+        return tag;
+    }
+
     public interface Lookup<T> {
         @Nullable T element(Identifier p_459727_, boolean p_362986_);
 
