--- a/net/minecraft/tags/WorldPresetTags.java
+++ b/net/minecraft/tags/WorldPresetTags.java
@@ -14,4 +_,12 @@
     private static TagKey<WorldPreset> create(final String name) {
         return TagKey.create(Registries.WORLD_PRESET, Identifier.withDefaultNamespace(name));
     }
+
+    public static TagKey<WorldPreset> create(String namepsace, String path) {
+        return create(Identifier.fromNamespaceAndPath(namepsace, path));
+    }
+
+    public static TagKey<WorldPreset> create(Identifier name) {
+        return TagKey.create(Registries.WORLD_PRESET, name);
+    }
 }
