--- a/net/minecraft/world/level/lighting/BlockLightEngine.java
+++ b/net/minecraft/world/level/lighting/BlockLightEngine.java
@@ -109,7 +_,7 @@
     }
 
     private int getEmission(final long blockNode, final BlockState state) {
-        int i = state.getLightEmission();
+        int i = state.getLightEmission(chunkSource.getLevel(), mutablePos);
         return i > 0 && this.storage.lightOnInSection(SectionPos.blockToSection(blockNode)) ? i : 0;
     }
 
@@ -119,7 +_,7 @@
         LightChunk lightchunk = this.chunkSource.getChunkForLighting(pos.x(), pos.z());
         if (lightchunk != null) {
             lightchunk.findBlockLightSources((lightPos, state) -> {
-                int i = state.getLightEmission();
+                int i = state.getLightEmission(chunkSource.getLevel(), lightPos);
                 this.enqueueIncrease(lightPos.asLong(), LightEngine.QueueEntry.increaseLightFromEmission(i, isEmptyShape(state)));
             });
         }
