--- a/net/minecraft/world/inventory/AbstractContainerMenu.java
+++ b/net/minecraft/world/inventory/AbstractContainerMenu.java
@@ -438,6 +_,7 @@
                 ItemStack itemstack10 = this.getCarried();
                 player.updateTutorialInventoryAction(itemstack10, slot7.getItem(), clickaction);
                 if (!this.tryItemClickBehaviourOverride(player, clickaction, slot7, itemstack9, itemstack10)) {
+                    if (!net.minecraftforge.event.ForgeEventFactory.onItemStackedOn(itemstack10, itemstack9, slot7, clickaction, player, createCarriedSlotAccess()))
                     if (itemstack9.isEmpty()) {
                         if (!itemstack10.isEmpty()) {
                             int i3 = clickaction == ClickAction.PRIMARY ? itemstack10.getCount() : 1;
@@ -658,7 +_,7 @@
                 ItemStack itemstack = slot.getItem();
                 if (!itemstack.isEmpty() && ItemStack.isSameItemSameComponents(itemStack, itemstack)) {
                     int j = itemstack.getCount() + itemStack.getCount();
-                    int k = slot.getMaxStackSize(itemstack);
+                    int k = Math.min(slot.getMaxStackSize(itemstack), itemstack.getMaxStackSize());
                     if (j <= k) {
                         itemStack.setCount(0);
                         itemstack.setCount(j);
