--- a/net/minecraft/world/inventory/EnchantmentMenu.java
+++ b/net/minecraft/world/inventory/EnchantmentMenu.java
@@ -75,7 +_,7 @@
 
             @Override
             public boolean mayPlace(final ItemStack itemStack) {
-                return itemStack.is(Items.LAPIS_LAZULI);
+                return itemStack.is(net.minecraftforge.common.Tags.Items.ENCHANTING_FUELS);
             }
 
             @Override
@@ -103,23 +_,24 @@
             if (!itemstack.isEmpty() && itemstack.isEnchantable()) {
                 this.access.execute((level, pos) -> {
                     IdMap<Holder<Enchantment>> idmap = level.registryAccess().lookupOrThrow(Registries.ENCHANTMENT).asHolderIdMap();
-                    int j = 0;
+                    float j = 0;
 
                     for (BlockPos blockpos : EnchantingTableBlock.BOOKSHELF_OFFSETS) {
                         if (EnchantingTableBlock.isValidBookShelf(level, pos, blockpos)) {
-                            j++;
+                            j += level.getBlockState(pos.offset(blockpos)).getEnchantPowerBonus(level, pos.offset(blockpos));
                         }
                     }
 
                     this.random.setSeed(this.enchantmentSeed.get());
 
                     for (int k = 0; k < 3; k++) {
-                        this.costs[k] = EnchantmentHelper.getEnchantmentCost(this.random, k, j, itemstack);
+                        this.costs[k] = EnchantmentHelper.getEnchantmentCost(this.random, k, (int)j, itemstack);
                         this.enchantClue[k] = -1;
                         this.levelClue[k] = -1;
                         if (this.costs[k] < k + 1) {
                             this.costs[k] = 0;
                         }
+                        this.costs[k] = net.minecraftforge.event.ForgeEventFactory.onEnchantmentLevelSet(level, pos, k, (int)j, itemstack, costs[k]);
                     }
 
                     for (int l = 0; l < 3; l++) {
@@ -246,7 +_,7 @@
                 if (!this.moveItemStackTo(itemstack1, 2, 38, true)) {
                     return ItemStack.EMPTY;
                 }
-            } else if (itemstack1.is(Items.LAPIS_LAZULI)) {
+            } else if (itemstack1.is(net.minecraftforge.common.Tags.Items.ENCHANTING_FUELS)) {
                 if (!this.moveItemStackTo(itemstack1, 1, 2, true)) {
                     return ItemStack.EMPTY;
                 }
