--- a/net/minecraft/world/inventory/ResultSlot.java
+++ b/net/minecraft/world/inventory/ResultSlot.java
@@ -1,4 +_,4 @@
     protected void checkTakeAchievements(final ItemStack carried) {
         if (this.removeCount > 0) {
             carried.onCraftedBy(this.player, this.removeCount);
+            net.minecraftforge.event.ForgeEventFactory.firePlayerCraftingEvent(this.player, carried, this.craftSlots);
         }
 
         if (this.container instanceof RecipeCraftingHolder recipecraftingholder) {
@@ -86,7 +_,9 @@
         CraftingInput craftinginput = craftinginput$positioned.input();
         int i = craftinginput$positioned.left();
         int j = craftinginput$positioned.top();
+        net.minecraftforge.common.ForgeHooks.setCraftingPlayer(player);
         NonNullList<ItemStack> nonnulllist = this.getRemainingItems(craftinginput, player.level());
+        net.minecraftforge.common.ForgeHooks.setCraftingPlayer(null);
 
         for (int k = 0; k < craftinginput.height(); k++) {
             for (int l = 0; l < craftinginput.width(); l++) {
