--- a/net/minecraft/world/level/levelgen/feature/treedecorators/AlterGroundDecorator.java
+++ b/net/minecraft/world/level/levelgen/feature/treedecorators/AlterGroundDecorator.java
@@ -58,6 +_,7 @@
             BlockPos cursor = pos.above(dy);
             BlockState replaceWith = this.provider.getOptionalState(context.level(), context.random(), cursor);
             if (replaceWith != null) {
+                replaceWith = net.minecraftforge.event.ForgeEventFactory.alterGround(context.level(), context.random(), cursor, replaceWith);
                 context.setBlock(cursor, replaceWith);
                 break;
             }
