--- a/net/minecraft/world/level/levelgen/feature/treedecorators/AlterGroundDecorator.java
+++ b/net/minecraft/world/level/levelgen/feature/treedecorators/AlterGroundDecorator.java
@@ -58,6 +_,7 @@
             BlockPos blockpos = pos.above(i);
             BlockState blockstate = this.provider.getOptionalState(context.level(), context.random(), blockpos);
             if (blockstate != null) {
+                blockstate = net.minecraftforge.event.ForgeEventFactory.alterGround(context.level(), context.random(), blockpos, blockstate);
                 context.setBlock(blockpos, blockstate);
                 break;
             }
