--- a/net/minecraft/gametest/framework/TestCommand.java
+++ b/net/minecraft/gametest/framework/TestCommand.java
@@ -481,7 +_,12 @@
                 return Optional.empty();
             } else {
                 Holder.Reference<GameTestInstance> test = maybeTest.get();
-                GameTestInfo testInfo = new GameTestInfo(test, blockEntity.getRotation(), level, retryOptions);
+                // Forge: The rotation is stored in the structure block, and added in the GameTestInfo constructor.
+                // So reverse it to find the manually specified rotation so the test runs the same every time.
+                var rotation = blockEntity.test().flatMap(blockEntity.getLevel().registryAccess()::get).map(Holder::value).map(GameTestInstance::rotation).orElse(Rotation.NONE);
+                var steps = StructureUtils.getRotationStepsForRotation(blockEntity.getRotation()) - StructureUtils.getRotationStepsForRotation(rotation);
+                if (steps < 0) steps += 4;
+                GameTestInfo testInfo = new GameTestInfo(test, StructureUtils.getRotationForRotationSteps(steps), level, retryOptions);
                 testInfo.setTestBlockPos(testBlockPos);
                 return !verifyStructureExists(source, testInfo.getStructure()) ? Optional.empty() : Optional.of(testInfo);
             }
