--- a/net/minecraft/advancements/AdvancementRewards.java
+++ b/net/minecraft/advancements/AdvancementRewards.java
@@ -44,6 +_,7 @@
         LootParams lootparams = new LootParams.Builder(serverlevel)
             .withParameter(LootContextParams.THIS_ENTITY, player)
             .withParameter(LootContextParams.ORIGIN, player.position())
+            .withLuck(player.getLuck())
             .create(LootContextParamSets.ADVANCEMENT_REWARD);
         boolean flag = false;
 
