--- a/net/minecraft/world/level/DataPackConfig.java
+++ b/net/minecraft/world/level/DataPackConfig.java
@@ -16,7 +_,7 @@
     private final List<String> disabled;
 
     public DataPackConfig(final List<String> enabled, final List<String> disabled) {
-        this.enabled = ImmutableList.copyOf(enabled);
+        this.enabled = new java.util.ArrayList<>(enabled);
         this.disabled = ImmutableList.copyOf(disabled);
     }
 
@@ -26,5 +_,9 @@
 
     public List<String> getDisabled() {
         return this.disabled;
+    }
+
+    public void addModPacks(List<String> modPacks) {
+        enabled.addAll(modPacks.stream().filter(p -> !enabled.contains(p)).toList());
     }
 }
