--- a/net/minecraft/world/level/LevelReader.java
+++ b/net/minecraft/world/level/LevelReader.java
@@ -175,6 +_,11 @@
     }
 
     @Deprecated
+    default boolean isAreaLoaded(BlockPos center, int range) {
+       return this.hasChunksAt(center.offset(-range, -range, -range), center.offset(range, range, range));
+    }
+
+    @Deprecated
     default boolean hasChunkAt(final int blockX, final int blockZ) {
         return this.hasChunk(SectionPos.blockToSectionCoord(blockX), SectionPos.blockToSectionCoord(blockZ));
     }
