--- a/net/minecraft/client/sounds/SoundEngine.java
+++ b/net/minecraft/client/sounds/SoundEngine.java
@@ -83,6 +_,7 @@
         this.options = options;
         this.soundBuffers = new SoundBufferLibrary(resourceProvider);
         this.lastSeenDevices = this.deviceTracker.currentDevices();
+        net.minecraftforge.client.event.sound.SoundEngineLoadEvent.BUS.post(new net.minecraftforge.client.event.sound.SoundEngineLoadEvent(this));
     }
 
     public void reload() {
@@ -100,6 +_,7 @@
 
         this.destroy();
         this.loadLibrary();
+        net.minecraftforge.client.event.sound.SoundEngineLoadEvent.BUS.post(new net.minecraftforge.client.event.sound.SoundEngineLoadEvent(this));
     }
 
     private synchronized void loadLibrary() {
@@ -363,15 +_,18 @@
         }
     }
 
-    public SoundEngine.PlayResult play(final SoundInstance instance) {
+    public SoundEngine.PlayResult play(final SoundInstance instanceArg) {
         if (!this.loaded) {
             return SoundEngine.PlayResult.NOT_STARTED;
         }
 
-        if (!instance.canPlaySound()) {
+        if (!instanceArg.canPlaySound()) {
             return SoundEngine.PlayResult.NOT_STARTED;
         }
 
+        final var instance = net.minecraftforge.client.ForgeHooksClient.playSound(this, instanceArg);
+        if (instance == null || !instance.canPlaySound())
+            return SoundEngine.PlayResult.NOT_STARTED;
         WeighedSoundEvents soundEvent = instance.getOrResolve(this.soundManager);
         Identifier eventLocation = instance.getIdentifier();
         if (soundEvent == null) {
@@ -447,15 +_,18 @@
                     channel.setSelfPosition(position);
                     channel.setRelative(instance.isRelative());
                 });
+                SoundInstance soundinstance = instance;
                 if (!isStreaming) {
                     this.soundBuffers.getCompleteBuffer(sound.getPath()).thenAccept(soundBuffer -> handle.execute(channel -> {
                         channel.attachStaticBuffer(soundBuffer);
                         channel.play();
+                        net.minecraftforge.client.event.ForgeEventFactoryClient.onPlaySoundSource(this, soundinstance, channel);
                     }));
                 } else {
                     this.soundBuffers.getStream(sound.getPath(), isLooping).thenAccept(stream -> handle.execute(channel -> {
                         channel.attachBufferStream(stream);
                         channel.play();
+                        net.minecraftforge.client.event.ForgeEventFactoryClient.onPlayStreamingSource(this, soundinstance, channel);
                     }));
                 }
 
