--- a/net/minecraft/client/sounds/SoundEngine.java
+++ b/net/minecraft/client/sounds/SoundEngine.java
@@ -84,6 +_,7 @@
         this.options = options;
         this.soundBuffers = new SoundBufferLibrary(resourceProvider);
         this.lastSeenDevices = this.deviceTracker.currentDevices();
+        net.minecraftforge.client.event.sound.SoundEngineLoadEvent.BUS.post(new net.minecraftforge.client.event.sound.SoundEngineLoadEvent(this));
     }
 
     public void reload() {
@@ -101,6 +_,7 @@
 
         this.destroy();
         this.loadLibrary();
+        net.minecraftforge.client.event.sound.SoundEngineLoadEvent.BUS.post(new net.minecraftforge.client.event.sound.SoundEngineLoadEvent(this));
     }
 
     private synchronized void loadLibrary() {
@@ -342,12 +_,15 @@
         }
     }
 
-    public SoundEngine.PlayResult play(final SoundInstance instance) {
+    public SoundEngine.PlayResult play(SoundInstance instance) {
         if (!this.loaded) {
             return SoundEngine.PlayResult.NOT_STARTED;
         } else if (!instance.canPlaySound()) {
             return SoundEngine.PlayResult.NOT_STARTED;
         } else {
+            instance = net.minecraftforge.client.ForgeHooksClient.playSound(this, instance);
+            if (instance == null || !instance.canPlaySound())
+                return SoundEngine.PlayResult.NOT_STARTED;
             WeighedSoundEvents weighedsoundevents = instance.resolve(this.soundManager);
             Identifier identifier = instance.getIdentifier();
             if (weighedsoundevents == null) {
@@ -427,15 +_,18 @@
                         channel.setSelfPosition(vec3);
                         channel.setRelative(flag);
                     });
+                    SoundInstance soundinstance = instance;
                     if (!flag1) {
                         this.soundBuffers.getCompleteBuffer(sound.getPath()).thenAccept(soundBuffer -> channelaccess$channelhandle.execute(channel -> {
                             channel.attachStaticBuffer(soundBuffer);
                             channel.play();
+                            net.minecraftforge.client.event.ForgeEventFactoryClient.onPlaySoundSource(this, soundinstance, channel);
                         }));
                     } else {
                         this.soundBuffers.getStream(sound.getPath(), flag3).thenAccept(stream -> channelaccess$channelhandle.execute(channel -> {
                             channel.attachBufferStream(stream);
                             channel.play();
+                            net.minecraftforge.client.event.ForgeEventFactoryClient.onPlayStreamingSource(this, soundinstance, channel);
                         }));
                     }
 
