--- a/net/minecraft/world/level/storage/LevelStorageSource.java
+++ b/net/minecraft/world/level/storage/LevelStorageSource.java
@@ -571,6 +_,11 @@
             return dynamic;
         }
 
+        public CompoundTag getDataTagRaw(final boolean useFallback) throws IOException {
+            this.checkLock();
+            return LevelStorageSource.readLevelDataTagRaw(this.getDataFile(useFallback));
+        }
+
         public Dynamic<?> getUnfixedDataTag(final boolean useFallback) throws IOException {
             this.checkLock();
             Path path = this.getDataFile(useFallback);
@@ -590,6 +_,7 @@
             CompoundTag compoundtag = levelData.createTag(singleplayerUUID);
             CompoundTag compoundtag1 = new CompoundTag();
             compoundtag1.put("Data", compoundtag);
+            net.minecraftforge.common.ForgeHooks.writeAdditionalLevelSaveData(levelData, compoundtag1);
             this.saveLevelData(compoundtag1);
         }
 
@@ -616,6 +_,10 @@
 
         public Optional<Path> getIconFile() {
             return !this.lock.isValid() ? Optional.empty() : Optional.of(this.levelDirectory.iconFile());
+        }
+
+        public Path getWorldDir() {
+            return baseDir;
         }
 
         public void deleteLevel() throws IOException {
