--- a/net/minecraft/world/level/storage/LevelSummary.java
+++ b/net/minecraft/world/level/storage/LevelSummary.java
@@ -15,7 +_,7 @@
 import org.apache.commons.lang3.StringUtils;
 import org.jspecify.annotations.Nullable;
 
-public class LevelSummary implements Comparable<LevelSummary> {
+public class LevelSummary implements Comparable<LevelSummary>, net.minecraftforge.common.extensions.IForgeLevelSummary {
     public static final Component PLAY_WORLD = Component.translatable("selectWorld.select");
     public static final Component UPGRADE_AND_PLAY_WORLD = Component.translatable("selectWorld.upgrade_and_play");
     private final LevelSettings settings;
