--- a/net/minecraft/world/level/storage/PlayerDataStorage.java
+++ b/net/minecraft/world/level/storage/PlayerDataStorage.java
@@ -41,6 +_,7 @@
             Path path2 = path.resolve(player.getStringUUID() + ".dat");
             Path path3 = path.resolve(player.getStringUUID() + ".dat_old");
             Util.safeReplaceFile(path2, path1, path3);
+            net.minecraftforge.event.ForgeEventFactory.firePlayerSavingEvent(player, playerDir, player.getStringUUID());
         } catch (Exception exception) {
             LOGGER.warn("Failed to save player data for {}", player.getPlainTextName());
         }
@@ -83,5 +_,9 @@
             int i = NbtUtils.getDataVersion(tag);
             return DataFixTypes.PLAYER.updateToCurrentVersion(this.fixerUpper, tag, i);
         });
+    }
+
+    public File getPlayerDataFolder() {
+        return playerDir;
     }
 }
