--- a/net/minecraft/world/level/storage/SavedDataStorage.java
+++ b/net/minecraft/world/level/storage/SavedDataStorage.java
@@ -122,6 +_,10 @@
             }
 
             int version = NbtUtils.getDataVersion(tag, 1343);
+            // Forge: Allow the data fixer to be null, leaving the modder responsible for keeping track of their own data formats
+            if (type == null)
+                var8 = tag;
+            else
             var8 = type.update(this.fixerUpper, tag, version, newVersion);
         }
 
