--- a/net/minecraft/world/level/storage/SavedDataStorage.java
+++ b/net/minecraft/world/level/storage/SavedDataStorage.java
@@ -122,6 +_,10 @@
             }
 
             int i = NbtUtils.getDataVersion(compoundtag, 1343);
+            // Forge: Allow the data fixer to be null, leaving the modder responsible for keeping track of their own data formats
+            if (type == null)
+                compoundtag1 = compoundtag;
+            else
             compoundtag1 = type.update(this.fixerUpper, compoundtag, i, newVersion);
         }
 
