--- a/net/minecraft/client/renderer/block/BlockStateModelSet.java
+++ b/net/minecraft/client/renderer/block/BlockStateModelSet.java
@@ -28,4 +_,13 @@
     public Material.Baked getParticleMaterial(final BlockState blockState) {
         return this.get(blockState).particleMaterial();
     }
+
+    private final it.unimi.dsi.fastutil.objects.Object2BooleanOpenHashMap<BlockState> isInSolidLayer = new it.unimi.dsi.fastutil.objects.Object2BooleanOpenHashMap<>();
+    public boolean isInSolidLayer(final BlockState state) {
+        if (isInSolidLayer.containsKey(state))
+            return isInSolidLayer.getBoolean(state);
+        var ret = net.minecraftforge.client.ForgeHooksClient.isModelInSolidLayer(this.modelByState.get(state));
+        isInSolidLayer.put(state, ret);
+        return ret;
+    }
 }
