--- a/net/minecraft/client/renderer/block/ModelBlockRenderer.java
+++ b/net/minecraft/client/renderer/block/ModelBlockRenderer.java
@@ -49,6 +_,12 @@
         return !cutoutLeaves && blockState.getBlock() instanceof LeavesBlock;
     }
 
+
+    /** @deprecated Use ModelData and Layer variant */
+    public void tesselateBlock(final BlockQuadOutput output, final float x, final float y, final float z, final BlockAndTintGetter level, final BlockPos pos, final BlockState blockState, final BlockStateModel model, final long seed) {
+        tesselateBlock(output, x, y, z, level, pos, blockState, model, seed, net.minecraftforge.client.model.data.ModelData.EMPTY);
+    }
+
     public void tesselateBlock(
         final BlockQuadOutput output,
         final float x,
@@ -58,14 +_,15 @@
         final BlockPos pos,
         final BlockState blockState,
         final BlockStateModel model,
-        final long seed
+        final long seed,
+        net.minecraftforge.client.model.data.ModelData modelData
     ) {
         this.random.setSeed(seed);
-        model.collectParts(this.random, this.parts);
+        model.collectParts(this.random, this.parts, modelData);
         if (!this.parts.isEmpty()) {
             try {
                 Vec3 vec3 = blockState.getOffset(pos);
-                if (this.ambientOcclusion && blockState.getLightEmission() == 0 && this.parts.getFirst().useAmbientOcclusion()) {
+                if (this.ambientOcclusion && blockState.getLightEmission(level, pos) == 0 && this.parts.getFirst().useAmbientOcclusion()) {
                     this.tesselateAmbientOcclusion(output, x + (float)vec3.x, y + (float)vec3.y, z + (float)vec3.z, this.parts, level, blockState, pos);
                 } else {
                     this.tesselateFlat(output, x + (float)vec3.x, y + (float)vec3.y, z + (float)vec3.z, this.parts, level, blockState, pos);
