--- a/net/minecraft/network/chat/contents/TranslatableContents.java
+++ b/net/minecraft/network/chat/contents/TranslatableContents.java
@@ -136,6 +_,11 @@
                 j = l;
             }
 
+            if (j == 0) {
+                // Forge has some special formatting handlers defined in ForgeI18n, use those if no %s replacements present.
+                j = net.minecraftforge.internal.TextComponentMessageFormatHandler.handle(this, decomposedParts, this.args, template);
+            }
+
             if (j < template.length()) {
                 String s3 = template.substring(j);
                 if (s3.indexOf(37) != -1) {
