--- a/net/minecraft/network/chat/contents/TranslatableContents.java
+++ b/net/minecraft/network/chat/contents/TranslatableContents.java
@@ -135,6 +_,11 @@
                 current = end;
             }
 
+            if (current == 0) {
+                // Forge has some special formatting handlers defined in ForgeI18n, use those if no %s replacements present.
+                current = net.minecraftforge.internal.TextComponentMessageFormatHandler.handle(this, decomposedParts, this.args, template);
+            }
+
             if (current < template.length()) {
                 String tail = template.substring(current);
                 if (tail.indexOf(37) != -1) {
