--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -298,7 +_,7 @@
 
     public static <S extends MinecraftServer> S spin(final Function<Thread, S> factory) {
         AtomicReference<S> atomicreference = new AtomicReference<>();
-        Thread thread = new Thread(() -> atomicreference.get().runServer(), "Server thread");
+        Thread thread = new Thread(net.minecraftforge.fml.util.thread.SidedThreadGroups.SERVER, () -> atomicreference.get().runServer(), "Server thread");
         thread.setUncaughtExceptionHandler((t, e) -> LOGGER.error("Uncaught exception in server thread", e));
         if (Runtime.getRuntime().availableProcessors() > 4) {
             thread.setPriority(8);
@@ -442,6 +_,7 @@
         this.scoreboard.load(this.savedDataStorage.computeIfAbsent(ScoreboardSaveData.TYPE).getData());
         this.commandStorage = new CommandStorage(this.savedDataStorage);
         this.stopwatches = this.savedDataStorage.computeIfAbsent(Stopwatches.TYPE);
+        net.minecraftforge.event.ForgeEventFactory.onLevelLoad(levels.get(Level.OVERWORLD));
         if (!serverleveldata.isInitialized()) {
             try {
                 setInitialSpawn(serverlevel, serverleveldata, worldoptions.generateBonusChest(), flag, this.levelLoadListener);
@@ -476,6 +_,7 @@
                     this, this.executor, this.storageSource, derivedleveldata, resourcekey1, entry.getValue(), flag, j, ImmutableList.of(), false
                 );
                 this.levels.put(resourcekey1, serverlevel1);
+                net.minecraftforge.event.ForgeEventFactory.onLevelLoad(serverlevel1);
             } else {
                 serverlevel1 = serverlevel;
             }
@@ -498,6 +_,7 @@
             levelData.setSpawn(LevelData.RespawnData.of(level.dimension(), BlockPos.ZERO.above(80), 0.0F, 0.0F));
         } else {
             ServerChunkCache serverchunkcache = level.getChunkSource();
+            if (net.minecraftforge.event.ForgeEventFactory.onCreateWorldSpawn(level, levelData)) return;
             ChunkPos chunkpos = ChunkPos.containing(serverchunkcache.randomState().sampler().findSpawnPosition());
             levelLoadListener.start(LevelLoadListener.Stage.PREPARE_GLOBAL_SPAWN, 0);
             levelLoadListener.updateFocus(level.dimension(), chunkpos);
@@ -687,6 +_,7 @@
         for (ServerLevel serverlevel2 : this.getAllLevels()) {
             if (serverlevel2 != null) {
                 try {
+                    net.minecraftforge.event.ForgeEventFactory.onLevelUnload(serverlevel2);
                     serverlevel2.close();
                 } catch (IOException ioexception1) {
                     LOGGER.error("Exception closing the level", (Throwable)ioexception1);
@@ -734,9 +_,11 @@
                 throw new IllegalStateException("Failed to initialize server");
             }
 
+            net.minecraftforge.server.ServerLifecycleHooks.handleServerStarted(this);
             this.nextTickTimeNanos = Util.getNanos();
             this.statusIcon = this.loadStatusIcon().orElse(null);
             this.status = this.buildServerStatus();
+            resetStatusCache(status);
 
             while (this.running) {
                 long i;
@@ -786,6 +_,8 @@
                 this.isReady = true;
                 JvmProfiler.INSTANCE.onServerTick(this.smoothedTickTimeMillis);
             }
+            net.minecraftforge.server.ServerLifecycleHooks.handleServerStopping(this);
+            net.minecraftforge.server.ServerLifecycleHooks.expectServerStopped(); // Forge: Has to come before MinecraftServer#onServerCrash to avoid race conditions
         } catch (Throwable throwable2) {
             LOGGER.error("Encountered an unexpected exception", throwable2);
             CrashReport crashreport = constructOrExtractCrashReport(throwable2);
@@ -797,6 +_,7 @@
                 LOGGER.error("We were unable to save this crash report to disk.");
             }
 
+            net.minecraftforge.server.ServerLifecycleHooks.expectServerStopped(); // Forge: Has to come before MinecraftServer#onServerCrash to avoid race conditions
             this.onServerCrash(crashreport);
         } finally {
             try {
@@ -805,6 +_,7 @@
             } catch (Throwable throwable) {
                 LOGGER.error("Exception stopping the server", throwable);
             } finally {
+                net.minecraftforge.server.ServerLifecycleHooks.handleServerStopped(this);
                 this.onServerExit();
             }
         }
@@ -985,12 +_,14 @@
             }
         }
 
+        net.minecraftforge.event.ForgeEventFactory.onPreServerTick(haveTime, this);
         this.tickCount++;
         this.tickRateManager.tick();
         this.tickChildren(haveTime);
         if (i - this.lastServerStatus >= STATUS_EXPIRE_TIME_NANOS) {
             this.lastServerStatus = i;
             this.status = this.buildServerStatus();
+            resetStatusCache(status);
         }
 
         this.ticksUntilAutosave--;
@@ -1008,6 +_,7 @@
         this.smoothedTickTimeMillis = this.smoothedTickTimeMillis * 0.8F + (float)k / (float)TimeUtil.NANOSECONDS_PER_MILLISECOND * 0.19999999F;
         this.logTickMethodTime(i);
         profilerfiller.pop();
+        net.minecraftforge.event.ForgeEventFactory.onPostServerTick(haveTime, this);
     }
 
     protected void processPacketsAndTick(final boolean sprinting) {
@@ -1069,7 +_,8 @@
             Optional.of(serverstatus$players),
             Optional.of(ServerStatus.Version.current()),
             Optional.ofNullable(this.statusIcon),
-            this.enforceSecureProfile()
+            this.enforceSecureProfile(),
+            Optional.of(new net.minecraftforge.network.ServerStatusPing())
         );
     }
 
@@ -1114,9 +_,11 @@
         profilerfiller.push("levels");
         this.updateEffectiveRespawnData();
 
-        for (ServerLevel serverlevel : this.getAllLevels()) {
+        for (ServerLevel serverlevel : this.getWorldArray()) {
+            long tickStart = Util.getNanos();
             profilerfiller.push(() -> serverlevel + " " + serverlevel.dimension().identifier());
             profilerfiller.push("tick");
+            net.minecraftforge.event.ForgeEventFactory.onPreLevelTick(serverlevel, haveTime);
 
             try {
                 serverlevel.tick(haveTime);
@@ -1125,9 +_,11 @@
                 serverlevel.fillReportDetails(crashreport);
                 throw new ReportedException(crashreport);
             }
+            net.minecraftforge.event.ForgeEventFactory.onPostLevelTick(serverlevel, haveTime);
 
             profilerfiller.pop();
             profilerfiller.pop();
+            perWorldTickTimes.computeIfAbsent(serverlevel.dimension(), k -> new long[100])[this.tickCount % 100] = Util.getNanos() - tickStart;
         }
 
         profilerfiller.popPush("connection");
@@ -1136,7 +_,7 @@
         this.playerList.tick();
         profilerfiller.popPush("debugSubscribers");
         this.debugSubscribers.tick();
-        if (this.tickRateManager.runsNormally()) {
+        if (net.minecraftforge.gametest.ForgeGameTestHooks.isGametestEnabled() && this.tickRateManager.runsNormally()) {
             profilerfiller.popPush("gameTests");
             GameTestTicker.SINGLETON.tick();
         }
@@ -1222,7 +_,7 @@
     }
 
     public String getServerModName() {
-        return "vanilla";
+        return net.minecraftforge.internal.BrandingControl.getBranding();
     }
 
     public ServerClockManager clockManager() {
@@ -1561,6 +_,7 @@
                     this.functionManager.replaceLibrary(this.resources.managers.getFunctionLibrary());
                     this.structureTemplateManager.onResourceManagerReload(this.resources.resourceManager);
                     this.fuelValues = FuelValues.vanillaBurnTimes(this.registries.compositeAccess(), this.worldData.enabledFeatures());
+                    this.getPlayerList().getPlayers().forEach(this.getPlayerList()::sendPlayerPermissionLevel); //Forge: Fix newly added/modified commands not being sent to the client when commands reload.
                 },
                 this
             );
@@ -1574,12 +_,15 @@
     public static WorldDataConfiguration configurePackRepository(
         final PackRepository packRepository, final WorldDataConfiguration initialDataConfig, final boolean initMode, final boolean safeMode
     ) {
+        net.minecraftforge.resource.ResourcePackLoader.loadResourcePacks(packRepository, false);
         DataPackConfig datapackconfig = initialDataConfig.dataPacks();
         FeatureFlagSet featureflagset = initMode ? FeatureFlagSet.of() : initialDataConfig.enabledFeatures();
         FeatureFlagSet featureflagset1 = initMode ? FeatureFlags.REGISTRY.allFlags() : initialDataConfig.enabledFeatures();
         packRepository.reload();
+        DataPackConfig.DEFAULT.addModPacks(net.minecraftforge.common.ForgeHooks.getModPacks());
+        datapackconfig.addModPacks(net.minecraftforge.common.ForgeHooks.getModPacks());
         if (safeMode) {
-            return configureRepositoryWithSelection(packRepository, List.of("vanilla"), featureflagset, false);
+            return configureRepositoryWithSelection(packRepository, net.minecraftforge.common.ForgeHooks.getModPacksWithVanilla(), featureflagset, false);
         } else {
             Set<String> set = Sets.newLinkedHashSet();
 
@@ -2235,6 +_,48 @@
 
     public ServerLinks serverLinks() {
         return ServerLinks.EMPTY;
+    }
+
+    private static final com.google.gson.Gson GSON = new com.google.gson.Gson();
+    private String cachedServerStatus; // FORGE: cache the server status json in case a client spams requests
+    private void resetStatusCache(ServerStatus status) {
+       this.cachedServerStatus = GSON.toJson(ServerStatus.CODEC.encodeStart(com.mojang.serialization.JsonOps.INSTANCE, status).result().orElseThrow());
+    }
+
+    public String getStatusJson() {
+       return cachedServerStatus;
+    }
+
+    private Map<ResourceKey<Level>, long[]> perWorldTickTimes = Maps.newIdentityHashMap();
+    @Nullable
+    public long[] getTickTime(ResourceKey<Level> dim) {
+        return perWorldTickTimes.get(dim);
+    }
+
+    @Deprecated //Forge Internal use Only, You can screw up a lot of things if you mess with this map.
+    public synchronized Map<ResourceKey<Level>, ServerLevel> forgeGetWorldMap() {
+        return this.levels;
+    }
+
+    private int worldArrayMarker = 0;
+    private int worldArrayLast = -1;
+    private ServerLevel[] worldArray;
+
+    @Deprecated //Forge Internal use Only, use to protect against concurrent modifications in the world tick loop.
+    public synchronized void markWorldsDirty() {
+        worldArrayMarker++;
+    }
+
+    private ServerLevel[] getWorldArray() {
+        if (worldArrayMarker == worldArrayLast && worldArray != null)
+           return worldArray;
+        worldArray = this.levels.values().stream().toArray(ServerLevel[]::new);
+        worldArrayLast = worldArrayMarker;
+        return worldArray;
+    }
+
+    public MinecraftServer.ReloadableResources getServerResources() {
+        return resources;
     }
 
     protected int pauseWhenEmptySeconds() {
