--- a/net/minecraft/server/PlayerAdvancements.java
+++ b/net/minecraft/server/PlayerAdvancements.java
@@ -173,6 +_,7 @@
             this.unregisterListeners(holder);
             this.progressChanged.add(holder);
             flag = true;
+            net.minecraftforge.event.ForgeEventFactory.onAdvancementGrant(this.player, holder, advancementprogress, criterion);
             if (!flag1 && advancementprogress.isDone()) {
                 holder.value().rewards().grant(this.player);
                 holder.value().display().ifPresent(display -> {
@@ -180,6 +_,7 @@
                         this.playerList.broadcastSystemMessage(display.getType().createAnnouncement(holder, this.player), false);
                     }
                 });
+                net.minecraftforge.event.ForgeEventFactory.onAdvancementEarned(this.player, holder);
             }
         }
 
@@ -198,6 +_,7 @@
             this.registerListeners(advancement);
             this.progressChanged.add(advancement);
             flag = true;
+            net.minecraftforge.event.ForgeEventFactory.onAdvancementRevoke(this.player, advancement, advancementprogress, criterion);
         }
 
         if (flag1 && !advancementprogress.isDone()) {
