--- a/net/minecraft/server/ReloadableServerResources.java
+++ b/net/minecraft/server/ReloadableServerResources.java
@@ -30,6 +_,7 @@
     private final ServerFunctionLibrary functionLibrary;
     private final List<Registry.PendingTags<?>> postponedTags;
     private final List<DataComponentInitializers.PendingComponents<?>> newComponents;
+    private final net.minecraftforge.common.crafting.conditions.ICondition.IContext context;
 
     private ReloadableServerResources(
         final LayeredRegistryAccess<RegistryLayer> fullLayers,
@@ -43,9 +_,11 @@
         this.fullRegistryHolder = new ReloadableServerRegistries.Holder(fullLayers.compositeAccess());
         this.postponedTags = postponedTags;
         this.newComponents = newComponents;
-        this.recipes = new RecipeManager(loadingContext);
         this.commands = new Commands(commandSelection, CommandBuildContext.simple(loadingContext, enabledFeatures));
-        this.advancements = new ServerAdvancementManager(loadingContext);
+        // Forge: Create context object and pass it to the recipe manager.
+        this.context = new net.minecraftforge.common.crafting.conditions.ConditionContext(this.postponedTags);
+        this.recipes = new RecipeManager(loadingContext, this.context);
+        this.advancements = new ServerAdvancementManager(loadingContext, this.context);
         this.functionLibrary = new ServerFunctionLibrary(functionCompilationPermissions, this.commands.getDispatcher());
     }
 
@@ -101,7 +_,7 @@
                             );
                             return SimpleReloadInstance.create(
                                     resourceManager,
-                                    reloadableserverresources.listeners(),
+                                    net.minecraftforge.event.ForgeEventFactory.onResourceReload(reloadableserverresources, fullRegistries.lookupWithUpdatedTags(), reloadableserverresources.listeners()),
                                     backgroundExecutor,
                                     mainThreadExecutor,
                                     DATA_RELOAD_INITIAL_TASK,
@@ -117,5 +_,14 @@
     public void updateComponentsAndStaticRegistryTags() {
         this.postponedTags.forEach(Registry.PendingTags::apply);
         this.newComponents.forEach(DataComponentInitializers.PendingComponents::apply);
+    }
+
+    /**
+     * Exposes the current condition context for usage in other reload listeners.<br>
+     * This is not useful outside the reloading stage.
+     * @return The condition context for the currently active reload.
+     */
+    public net.minecraftforge.common.crafting.conditions.ICondition.IContext getConditionContext() {
+        return this.context;
     }
 }
