--- a/net/minecraft/server/ServerAdvancementManager.java
+++ b/net/minecraft/server/ServerAdvancementManager.java
@@ -22,6 +_,10 @@
     private final AdvancementTree tree;
 
     public ServerAdvancementManager(final HolderLookup.Provider registries) {
+        this(registries, net.minecraftforge.common.crafting.conditions.ICondition.IContext.EMPTY);
+    }
+
+    public ServerAdvancementManager(final HolderLookup.Provider registries, net.minecraftforge.common.crafting.conditions.ICondition.IContext context) {
         HolderLookup.RegistryLookup<Advancement> advancements = registries.lookupOrThrow(Registries.ADVANCEMENT);
         Builder<Identifier, AdvancementHolder> builder = ImmutableMap.builder();
         advancements.listElements().forEach(advancement -> {
