--- a/net/minecraft/server/ServerAdvancementManager.java
+++ b/net/minecraft/server/ServerAdvancementManager.java
@@ -27,6 +_,10 @@
     private final HolderLookup.Provider registries;
 
     public ServerAdvancementManager(final HolderLookup.Provider registries) {
+        this(registries, net.minecraftforge.common.crafting.conditions.ICondition.IContext.EMPTY);
+    }
+
+    public ServerAdvancementManager(final HolderLookup.Provider registries, net.minecraftforge.common.crafting.conditions.ICondition.IContext context) {
         super(registries, Advancement.CODEC, Registries.ADVANCEMENT);
         this.registries = registries;
     }
