--- a/net/minecraft/world/entity/animal/equine/AbstractHorse.java
+++ b/net/minecraft/world/entity/animal/equine/AbstractHorse.java
@@ -140,6 +_,7 @@
         }
 
         this.addBehaviourGoals();
+        this.itemHandler = net.minecraftforge.common.util.LazyOptional.of(() -> new net.minecraftforge.items.wrapper.InvWrapper(this.inventory));
     }
 
     protected void addBehaviourGoals() {
@@ -274,6 +_,11 @@
 
     @Override
     public boolean causeFallDamage(double p_454114_, float p_458836_, DamageSource p_450439_) {
+        var event = net.minecraftforge.event.ForgeEventFactory.onLivingFall(this, p_454114_, p_458836_);
+        if (event == null) return false;
+        p_454114_ = event.getDistance();
+        p_458836_ = event.getDamageMultiplier();
+
         if (p_454114_ > 1.0) {
             this.playSound(SoundEvents.HORSE_LAND, 0.4F, 1.0F);
         }
@@ -339,9 +_,9 @@
     protected void playStepSound(BlockPos p_457686_, BlockState p_459257_) {
         if (!p_459257_.liquid()) {
             BlockState blockstate = this.level().getBlockState(p_457686_.above());
-            SoundType soundtype = p_459257_.getSoundType();
+            SoundType soundtype = p_459257_.getSoundType(level(), p_457686_, this);
             if (blockstate.is(Blocks.SNOW)) {
-                soundtype = blockstate.getSoundType();
+                soundtype = blockstate.getSoundType(level(), p_457686_.above(), this);
             }
 
             if (this.isVehicle() && this.canGallop) {
@@ -773,6 +_,7 @@
             float f1 = Mth.cos(this.getYRot() * (float) (Math.PI / 180.0));
             this.setDeltaMovement(this.getDeltaMovement().add(-0.4F * f * p_452386_, 0.0, 0.4F * f1 * p_452386_));
         }
+        net.minecraftforge.common.ForgeHooks.onLivingJump(this);
     }
 
     protected void playJumpSound() {
@@ -1013,6 +_,26 @@
 
     public boolean hasInventoryChanged(Container p_457588_) {
         return this.inventory != p_457588_;
+    }
+
+    private net.minecraftforge.common.util.LazyOptional<?> itemHandler = null;
+
+    @Override
+    public <T> net.minecraftforge.common.util.LazyOptional<T> getCapability(net.minecraftforge.common.capabilities.Capability<T> capability, net.minecraft.core.@Nullable Direction facing) {
+        if (capability == net.minecraftforge.common.capabilities.ForgeCapabilities.ITEM_HANDLER && itemHandler != null && this.isAlive()) {
+            return itemHandler.cast();
+        }
+        return super.getCapability(capability, facing);
+    }
+
+    @Override
+    public void invalidateCaps() {
+        super.invalidateCaps();
+        if (itemHandler != null) {
+           var oldHandler = itemHandler;
+           itemHandler = null;
+           oldHandler.invalidate();
+        }
     }
 
     public int getAmbientStandInterval() {
