--- a/net/minecraft/world/entity/animal/equine/Llama.java
+++ b/net/minecraft/world/entity/animal/equine/Llama.java
@@ -369,6 +_,10 @@
 
     @Override
     public boolean causeFallDamage(final double fallDistance, final float damageModifier, final DamageSource damageSource) {
+        var event = net.minecraftforge.event.ForgeEventFactory.onLivingFall(this, fallDistance, damageModifier);
+        if (event == null) return false;
+        fallDistance = event.getDistance();
+        damageModifier = event.getDamageMultiplier();
         int i = this.calculateFallDamage(fallDistance, damageModifier);
         if (i <= 0) {
             return false;
