--- a/net/minecraft/world/entity/animal/equine/SkeletonTrapGoal.java
+++ b/net/minecraft/world/entity/animal/equine/SkeletonTrapGoal.java
@@ -31,6 +_,13 @@
     @Override
     public void tick() {
         ServerLevel level = (ServerLevel)this.horse.level();
+        // Forge: Trigger the trap in a tick task to avoid crashes when mods add goals to skeleton horses
+        // (MC-206338/Forge PR #7509)
+        level.getServer().schedule(level.getServer().wrapRunnable(() -> this.convert(level)));
+    }
+
+    private void convert(ServerLevel level) {
+        if (!this.horse.isAlive()) return;
         DifficultyInstance difficulty = level.getCurrentDifficultyAt(this.horse.blockPosition());
         this.horse.setTrap(false);
         this.horse.setTamed(true);
