--- a/net/minecraft/client/renderer/FirstPersonHandsAndItemsRenderer.java
+++ b/net/minecraft/client/renderer/FirstPersonHandsAndItemsRenderer.java
@@ -352,6 +_,7 @@
             if (state.handRenderSelection.renderMainHand) {
                 float mainHandAttack = attackHand == InteractionHand.MAIN_HAND ? attackValue : 0.0F;
                 float mainhandInverseArmHeight = state.mainHandSwapScale * (1.0F - Mth.lerp(partialTicks, state.oldMainHandHeight, state.mainHandHeight));
+                if (!net.minecraftforge.client.ForgeHooksClient.renderSpecificFirstPersonHand(InteractionHand.MAIN_HAND, poseStack, submitNodeCollector, playerState.avatarRenderState.lightCoords, partialTicks, xRot, mainHandAttack, mainhandInverseArmHeight, state.mainHandItem))
                 this.submitArmWithItem(
                     playerState,
                     state,
@@ -370,6 +_,7 @@
             if (state.handRenderSelection.renderOffHand) {
                 float offHandAttack = attackHand == InteractionHand.OFF_HAND ? attackValue : 0.0F;
                 float offhandInverseArmHeight = state.offHandSwapScale * (1.0F - Mth.lerp(partialTicks, state.oldOffHandHeight, state.offHandHeight));
+                if (!net.minecraftforge.client.ForgeHooksClient.renderSpecificFirstPersonHand(InteractionHand.OFF_HAND, poseStack, submitNodeCollector, playerState.avatarRenderState.lightCoords, partialTicks, xRot, offHandAttack, offhandInverseArmHeight, state.offHandItem))
                 this.submitArmWithItem(
                     playerState,
                     state,
@@ -418,7 +_,7 @@
                     } else {
                         this.renderOneHandedMap(poseStack, submitNodeCollector, lightCoords, inverseArmHeight, arm, attack, itemStack, playerState, state);
                     }
-                } else if (itemStack.is(Items.CROSSBOW)) {
+                } else if (itemStack.getItem() instanceof CrossbowItem) {
                     this.applyItemArmTransform(poseStack, arm, inverseArmHeight);
                     boolean charged = CrossbowItem.isCharged(itemStack);
                     boolean isRightArm = arm == HumanoidArm.RIGHT;
@@ -457,6 +_,7 @@
                 } else {
                     boolean isRightArm = arm == HumanoidArm.RIGHT;
                     int invert = isRightArm ? 1 : -1;
+                    if (!net.minecraftforge.client.extensions.common.IClientItemExtensions.of(itemStack).applyForgeHandTransform(poseStack, minecraft.player, arm, itemStack, partialTicks, inverseArmHeight, attack)) // FORGE: Allow items to define custom arm animation
                     if (avatarRenderState.isUsingItem && state.useItemRemainingTicks > 0 && avatarRenderState.useItemHand == hand) {
                         ItemUseAnimation useAnimation = itemStack.getUseAnimation();
                         if (!useAnimation.hasCustomArmTransform()) {
