--- a/net/minecraft/client/renderer/GameRenderer.java
+++ b/net/minecraft/client/renderer/GameRenderer.java
@@ -235,6 +_,9 @@
             case null:
             default:
                 this.clearPostEffect();
+                var rl = cameraEntity == null ? null : net.minecraftforge.client.EntitySpectatorShaderManager.get(cameraEntity.getType());
+                if (rl != null)
+                    this.setPostEffect(rl);
         }
     }
 
@@ -246,7 +_,13 @@
     public void processBlurEffect() {
         PostChain postchain = this.minecraft.getShaderManager().getPostChain(BLUR_POST_CHAIN_ID, LevelTargetBundle.MAIN_TARGETS);
         if (postchain != null) {
+            // TODO: [Forge][Rendering] Check if this blend management is needed anymnore with the new render pipeline changes
+            boolean wasBlendEnabled = com.mojang.blaze3d.opengl.GlStateManager._isBlendEnabled();
+            if (wasBlendEnabled)
+                com.mojang.blaze3d.opengl.GlStateManager._disableBlend();
             postchain.process(this.minecraft.getMainRenderTarget(), this.resourcePool);
+            if (wasBlendEnabled)
+                com.mojang.blaze3d.opengl.GlStateManager._enableBlend();
         }
     }
 
@@ -507,7 +_,7 @@
             profilerfiller.push("screen");
 
             try {
-                this.minecraft.screen.extractRenderStateWithTooltipAndSubtitles(guigraphicsextractor, i, j, deltaTracker.getGameTimeDeltaTicks());
+                net.minecraftforge.client.ForgeHooksClient.drawScreen(this.minecraft.screen, guigraphicsextractor, i, j, deltaTracker.getRealtimeDeltaTicks());
             } catch (Throwable throwable) {
                 CrashReport crashreport1 = CrashReport.forThrowable(throwable, "Rendering screen");
                 CrashReportCategory crashreportcategory1 = crashreport1.addCategory("Screen render details");
@@ -672,6 +_,10 @@
                         Registry<Block> registry = this.minecraft.level.registryAccess().lookupOrThrow(Registries.BLOCK);
                         flag = !itemstack.isEmpty()
                             && (itemstack.canBreakBlockInAdventureMode(blockinworld) || itemstack.canPlaceOnBlockInAdventureMode(blockinworld));
+                        if (!flag && blockstate.hasBlockEntity()) {
+                            var blockEntity = this.minecraft.level.getBlockEntity(blockpos);
+                            flag = blockEntity != null && blockEntity.hasCustomOutlineRendering();
+                        }
                     }
                 }
             }
@@ -696,6 +_,9 @@
         if (optionsrenderstate.bobView) {
             this.bobView(camerarenderstate, posestack);
         }
+
+        var cameraSetup = net.minecraftforge.client.event.ForgeEventFactoryClient.fireComputeCameraAngles(this, this.mainCamera, f);
+        this.mainCamera.setRotation(cameraSetup.getYaw(), cameraSetup.getPitch(), cameraSetup.getRoll());
 
         matrix4f.mul(posestack.last().pose());
         float f2 = optionsrenderstate.screenEffectScale;
