--- a/net/minecraft/client/renderer/LevelRenderer.java
+++ b/net/minecraft/client/renderer/LevelRenderer.java
@@ -279,6 +_,7 @@
             }
         }
 
+        net.minecraftforge.client.FramePassManager.insertForgePasses(frame, targets, this.levelRenderState); // Forge: Modded passes are inserted here.
         profiler.popPush("executeFrameGraph");
         frame.execute(resourceAllocator, new FrameGraphBuilder.Inspector() {
             @Override
@@ -335,7 +_,8 @@
         levelRenderState.blockBreakingRenderStates.clear();
         this.submitTransientBlocks(poseStack, submitNodeCollector, levelRenderState);
         levelRenderState.particlesRenderState.submit(submitNodeCollector, levelRenderState.cameraRenderState);
-        if (renderOutline) {
+        var hasCustomOutline = this.levelRenderState.blockOutlineRenderState != null && this.levelRenderState.blockOutlineRenderState.customRenderer() != null;
+        if (renderOutline || hasCustomOutline) {
             this.submitBlockOutline(poseStack, this.submitNodeStorage, levelRenderState);
         }
 
@@ -1075,6 +_,10 @@
     private void submitBlockOutline(final PoseStack poseStack, final SubmitNodeCollector submitNodeCollector, final LevelRenderState levelRenderState) {
         BlockOutlineRenderState state = levelRenderState.blockOutlineRenderState;
         if (state != null) {
+            if (state.customRenderer() != null) {
+                state.customRenderer().render(submitNodeCollector, poseStack, levelRenderState);
+                return;
+            }
             Vec3 cameraPos = levelRenderState.cameraRenderState.pos;
             BlockPos pos = state.pos();
             poseStack.pushPose();
@@ -1331,6 +_,14 @@
 
     public SectionOcclusionGraph sectionOcclusionGraph() {
         return this.sectionOcclusionGraph;
+    }
+
+    public WeatherEffectRenderer getWeatherEffects() {
+        return this.weatherEffectRenderer;
+    }
+
+    public void setWeatherEffects(WeatherEffectRenderer value) {
+        this.weatherEffectRenderer = value;
     }
 
     public Gizmos.TemporaryCollection collectPerFrameRenderThreadGizmos() {
