--- a/net/minecraft/client/renderer/ScreenEffectRenderer.java
+++ b/net/minecraft/client/renderer/ScreenEffectRenderer.java
@@ -62,19 +_,25 @@
         Player player = this.minecraft.player;
         if (isFirstPerson && !isSleeping) {
             if (!player.noPhysics) {
-                BlockState blockState = getViewBlockingState(player);
+                var overlay = getOverlayBlock(player);
+                BlockState blockState = overlay == null ? null : overlay.getLeft();
                 if (blockState != null) {
+                    if (!net.minecraftforge.client.ForgeHooksClient.renderBlockOverlay(player, poseStack, net.minecraftforge.client.event.RenderBlockScreenEffectEvent.OverlayType.BLOCK, overlay.getLeft(), overlay.getRight()))
                     renderTex(this.minecraft.getModelManager().getBlockStateModelSet().getParticleMaterial(blockState).sprite(), poseStack, this.bufferSource);
                 }
             }
 
             if (!this.minecraft.player.isSpectator()) {
                 if (this.minecraft.player.isEyeInFluid(FluidTags.WATER)) {
+                    if (!net.minecraftforge.client.ForgeHooksClient.renderWaterOverlay(player, poseStack))
                     renderWater(this.minecraft, poseStack, this.bufferSource);
+                } else if (!player.getEyeInFluidType().isAir()) {
+                    net.minecraftforge.client.extensions.common.IClientFluidTypeExtensions.of(player.getEyeInFluidType()).renderOverlay(this.minecraft, poseStack, this.bufferSource);
                 }
 
                 if (this.minecraft.player.isOnFire()) {
                     TextureAtlasSprite fireSprite = this.sprites.get(ModelBakery.FIRE_1);
+                    if (!net.minecraftforge.client.ForgeHooksClient.renderFireOverlay(player, poseStack))
                     renderFire(poseStack, this.bufferSource, fireSprite);
                 }
             }
@@ -124,6 +_,11 @@
     }
 
     private static @Nullable BlockState getViewBlockingState(final Player player) {
+        var ret = getOverlayBlock(player);
+        return ret == null ? null : ret.getLeft();
+    }
+
+    private static org.apache.commons.lang3.tuple.@Nullable Pair<BlockState, BlockPos> getOverlayBlock(Player player) {
         BlockPos.MutableBlockPos testPos = new BlockPos.MutableBlockPos();
 
         for (int i = 0; i < 8; i++) {
@@ -133,7 +_,7 @@
             testPos.set(xo, yo, zo);
             BlockState blockState = player.level().getBlockState(testPos);
             if (blockState.getRenderShape() != RenderShape.INVISIBLE && blockState.isViewBlocking(player.level(), testPos)) {
-                return blockState;
+                return org.apache.commons.lang3.tuple.Pair.of(blockState, testPos.immutable());
             }
         }
 
@@ -161,6 +_,10 @@
     }
 
     private static void renderWater(final Minecraft minecraft, final PoseStack poseStack, final MultiBufferSource bufferSource) {
+        renderFluid(minecraft, poseStack, bufferSource, UNDERWATER_LOCATION);
+    }
+
+    public static void renderFluid(Minecraft minecraft, PoseStack poseStack, MultiBufferSource bufferSource, Identifier texture) {
         BlockPos pos = BlockPos.containing(minecraft.player.getX(), minecraft.player.getEyeY(), minecraft.player.getZ());
         float br = Lightmap.getBrightness(minecraft.player.level().dimensionType(), minecraft.player.level().getMaxLocalRawBrightness(pos));
         int color = ARGB.colorFromFloat(0.1F, br, br, br);
@@ -173,7 +_,7 @@
         float uo = -minecraft.player.getYRot() / 64.0F;
         float vo = minecraft.player.getXRot() / 64.0F;
         Matrix4f pose = poseStack.last().pose();
-        VertexConsumer builder = bufferSource.getBuffer(RenderTypes.blockScreenEffect(UNDERWATER_LOCATION));
+        VertexConsumer builder = bufferSource.getBuffer(RenderTypes.blockScreenEffect(texture));
         builder.addVertex(pose, -1.0F, -1.0F, -0.5F).setUv(4.0F + uo, 4.0F + vo).setColor(color);
         builder.addVertex(pose, 1.0F, -1.0F, -0.5F).setUv(0.0F + uo, 4.0F + vo).setColor(color);
         builder.addVertex(pose, 1.0F, 1.0F, -0.5F).setUv(0.0F + uo, 0.0F + vo).setColor(color);
