--- a/net/minecraft/client/renderer/ScreenEffectRenderer.java
+++ b/net/minecraft/client/renderer/ScreenEffectRenderer.java
@@ -40,6 +_,9 @@
         PoseStack poseStack = new PoseStack();
         AvatarRenderState avatarRenderState = playerRenderState.avatarRenderState;
         if (!cameraRenderState.entityRenderState.isSleeping && playerRenderState.hasPlayer && avatarRenderState != null) {
+            if (playerRenderState.customBlockOverlay != null)
+                playerRenderState.customBlockOverlay.render(poseStack, submitNodeCollector);
+            else
             if (playerRenderState.blockOverlay != null) {
                 submitBlockSprite(
                     playerRenderState.blockOverlay.atlasLocation(),
@@ -54,10 +_,16 @@
             }
 
             if (cameraRenderState.isFirstPerson && !avatarRenderState.isSpectator) {
+                if (playerRenderState.customFluidOverlay != null)
+                    playerRenderState.customFluidOverlay.render(poseStack, submitNodeCollector);
+                else
                 if (playerRenderState.waterOverlay != null) {
                     submitWater(playerRenderState.waterOverlay, poseStack, submitNodeCollector);
                 }
 
+                if (playerRenderState.customFireOverlay != null)
+                    playerRenderState.customFireOverlay.render(poseStack, submitNodeCollector);
+                else
                 if (playerRenderState.isOnFire) {
                     TextureAtlasSprite fireSprite = this.sprites.get(ModelBakery.FIRE_1);
                     submitFire(poseStack, submitNodeCollector, fireSprite);
@@ -116,9 +_,12 @@
     }
 
     private static void submitWater(final PlayerRenderState.WaterOverlay waterOverlay, final PoseStack poseStack, final SubmitNodeCollector submitNodeCollector) {
+        submitFluid(waterOverlay, poseStack, submitNodeCollector, UNDERWATER_LOCATION);
+    }
+    public static void submitFluid(final PlayerRenderState.WaterOverlay waterOverlay, final PoseStack poseStack, final SubmitNodeCollector submitNodeCollector, Identifier texture) {
         submitNodeCollector.submitCustomGeometry(
             poseStack,
-            RenderTypes.blockScreenEffect(UNDERWATER_LOCATION),
+            RenderTypes.blockScreenEffect(texture),
             (pose, builder) -> {
                 float uvSize = 4.0F;
                 buildQuad(
