--- a/net/minecraft/client/renderer/SpriteCoordinateExpander.java
+++ b/net/minecraft/client/renderer/SpriteCoordinateExpander.java
@@ -17,12 +_,14 @@
 
     @Override
     public VertexConsumer addVertex(final float x, final float y, final float z) {
-        return this.delegate.addVertex(x, y, z);
+        this.delegate.addVertex(x, y, z);
+        return this; // Forge: Fix MC-263524 not working with chained methods
     }
 
     @Override
     public VertexConsumer setColor(final int r, final int g, final int b, final int a) {
-        return this.delegate.setColor(r, g, b, a);
+        this.delegate.setColor(r, g, b, a);
+        return this; // Forge: Fix MC-263524 not working with chained methods
     }
 
     @Override
@@ -32,22 +_,26 @@
 
     @Override
     public VertexConsumer setUv(final float u, final float v) {
-        return this.delegate.setUv(this.sprite.getU(u), this.sprite.getV(v));
+        this.delegate.setUv(this.sprite.getU(u), this.sprite.getV(v));
+        return this; // Forge: Fix MC-263524 not working with chained methods
     }
 
     @Override
     public VertexConsumer setUv1(final int u, final int v) {
-        return this.delegate.setUv1(u, v);
+        this.delegate.setUv1(u, v);
+        return this; // Forge: Fix MC-263524 not working with chained methods
     }
 
     @Override
     public VertexConsumer setUv2(final int u, final int v) {
-        return this.delegate.setUv2(u, v);
+        this.delegate.setUv2(u, v);
+        return this; // Forge: Fix MC-263524 not working with chained methods
     }
 
     @Override
     public VertexConsumer setNormal(final float x, final float y, final float z) {
-        return this.delegate.setNormal(x, y, z);
+        this.delegate.setNormal(x, y, z);
+        return this; // Forge: Fix MC-263524 not working with chained methods
     }
 
     @Override
