--- a/net/minecraft/client/renderer/SubmitNodeCollection.java
+++ b/net/minecraft/client/renderer/SubmitNodeCollection.java
@@ -179,6 +_,12 @@
     }
 
     @Override
+    public void submitBreakingBlockModel(final PoseStack poseStack, final BlockStateModel model, final long seed, final int progress, final net.minecraftforge.client.model.data.ModelData data) {
+        this.wasUsed = true;
+        this.breakingBlockModelSubmits.add(new SubmitNodeStorage.BreakingBlockModelSubmit(poseStack.last().copy(), model, seed, progress, data));
+    }
+
+    @Override
     public void submitItem(
         final PoseStack poseStack,
         final ItemDisplayContext displayContext,
