--- a/com/mojang/blaze3d/platform/Window.java
+++ b/com/mojang/blaze3d/platform/Window.java
@@ -114,9 +_,10 @@
         Monitor initialMonitor = monitorManager.getMonitor(SDLVideo.SDL_GetPrimaryDisplay());
         this.windowedWidth = this.width = allowedWindowMinSize(displayData.width(), 320);
         this.windowedHeight = this.height = allowedWindowMinSize(displayData.height(), 240);
-        this.handle = this.createWindow(backend, this.width, this.height, title);
+        this.handle = net.minecraftforge.fml.loading.ImmediateWindowHandler.setupMinecraftWindow(this.width, this.height, title, () -> backend);
         this.setWindowMaxSize(maximumSize, maximumSize);
         MacosUtil.disableCloseWindowMenuItem();
+        if (!net.minecraftforge.fml.loading.ImmediateWindowHandler.positionWindow(Optional.ofNullable(initialMonitor), w->this.width = this.windowedWidth = w, h->this.height = this.windowedHeight = h, x->this.x = this.windowedX = x, y->this.y = this.windowedY = y)) {
         if (initialMonitor != null) {
             this.windowedX = this.x = initialMonitor.x() + (initialMonitor.w() - this.width) / 2;
             this.windowedY = this.y = initialMonitor.y() + (initialMonitor.h() - this.height) / 2;
@@ -132,6 +_,7 @@
                 this.windowedY = this.y = actualY.get(0);
             }
         }
+        }
 
         this.setMode();
         this.refreshFramebufferSize();
@@ -148,6 +_,10 @@
     }
 
     private long createWindow(final GpuBackend backend, final int width, final int height, final String title) {
+        return createWindowStatic(backend, width, height, title);
+    }
+
+    public static long createWindowStatic(final GpuBackend backend, final int width, final int height, final String title) {
         long flags = 8224L;
         long windowHandle = backend.createWindow(title, width, height, 8224L);
         if (windowHandle == 0L) {
