--- a/net/minecraft/data/registries/RegistriesDatapackGenerator.java
+++ b/net/minecraft/data/registries/RegistriesDatapackGenerator.java
@@ -18,10 +_,21 @@
 public class RegistriesDatapackGenerator implements DataProvider {
     private final PackOutput output;
     private final CompletableFuture<HolderLookup.Provider> registries;
+    private final java.util.function.Predicate<String> namespacePredicate;
 
     public RegistriesDatapackGenerator(final PackOutput output, final CompletableFuture<HolderLookup.Provider> registries) {
+        this(output, registries, null);
+    }
+
+    public RegistriesDatapackGenerator(final PackOutput output, final CompletableFuture<HolderLookup.Provider> registries, @org.jetbrains.annotations.Nullable java.util.Set<String> modIds) {
         this.registries = registries;
         this.output = output;
+        this.namespacePredicate = modIds == null ? namespace -> true : modIds::contains;
+    }
+
+    /** Forge: Call this to get the registry holder lookup provider that includes elements added via {@link net.minecraftforge.common.data.DatapackBuiltinEntriesProvider} */
+    public CompletableFuture<HolderLookup.Provider> getRegistryProvider() {
+       return registries;
     }
 
     @Override
@@ -31,8 +_,7 @@
                 access -> {
                     DynamicOps<JsonElement> registryOps = access.createSerializationContext(JsonOps.INSTANCE);
                     return CompletableFuture.allOf(
-                        RegistryDataLoader.WORLDGEN_REGISTRIES
-                            .stream()
+                        RegistryDataLoader.getWorldGenAndDimensionStream()
                             .flatMap(v -> this.dumpRegistryCap(cache, access, registryOps, (RegistryDataLoader.RegistryData<?>)v).stream())
                             .toArray(CompletableFuture[]::new)
                     );
@@ -50,11 +_,16 @@
                     PackOutput.PathProvider pathProvider = this.output.createRegistryElementsPathProvider(registryKey);
                     return CompletableFuture.allOf(
                         registry.listElements()
+                            .filter(holder -> shouldDump(holder.key()))
                             .<CompletableFuture<?>>map(e -> dumpValue(pathProvider.json(e.key().identifier()), cache, writeOps, v.elementCodec(), e.value()))
                             .toArray(CompletableFuture[]::new)
                     );
                 }
             );
+    }
+
+    protected boolean shouldDump(ResourceKey<?> key) {
+        return this.namespacePredicate.test(key.identifier().getNamespace());
     }
 
     private static <E> CompletableFuture<?> dumpValue(
