--- a/net/minecraft/data/registries/RegistryPatchGenerator.java
+++ b/net/minecraft/data/registries/RegistryPatchGenerator.java
@@ -21,7 +_,7 @@
             parent -> {
                 RegistryAccess.Frozen registryaccess$frozen = RegistryAccess.fromRegistryOfRegistries(BuiltInRegistries.REGISTRY);
                 Cloner.Factory cloner$factory = new Cloner.Factory();
-                RegistryDataLoader.WORLDGEN_REGISTRIES.forEach(registryData -> registryData.runWithArguments(cloner$factory::addCodec));
+                RegistryDataLoader.getWorldGenAndDimensionStream().forEach(registryData -> registryData.runWithArguments(cloner$factory::addCodec));
                 RegistrySetBuilder.PatchedRegistries registrysetbuilder$patchedregistries = packBuilder.buildPatch(
                     registryaccess$frozen, parent, cloner$factory
                 );
