--- a/com/mojang/blaze3d/opengl/GlDevice.java
+++ b/com/mojang/blaze3d/opengl/GlDevice.java
@@ -129,7 +_,12 @@
         final int depthOrLayers,
         final int mipLevels
     ) {
-        return this.createTexture(this.debugLabels.exists() && label != null ? label.get() : null, usage, format, width, height, depthOrLayers, mipLevels);
+        return this.createTexture(label, usage, format, width, height, depthOrLayers, mipLevels, false);
+    }
+
+    @Override
+    public GpuTexture createTexture(final @Nullable Supplier<String> label, @GpuTexture.Usage final int usage, final TextureFormat format, final int width, final int height, final int depthOrLayers, final int mipLevels, final boolean stencil ) {
+        return this.createTexture(this.debugLabels.exists() && label != null ? label.get() : null, usage, format, width, height, depthOrLayers, mipLevels, stencil);
     }
 
     @Override
@@ -142,6 +_,11 @@
         final int depthOrLayers,
         final int mipLevels
     ) {
+        return this.createTexture(label, usage, format, width, height, depthOrLayers, mipLevels, false);
+    }
+
+    @Override
+    public GpuTexture createTexture(@Nullable String label, @GpuTexture.Usage final int usage, final TextureFormat format, final int width, final int height, final int depthOrLayers, final int mipLevels, boolean stencil) {
         GlStateManager.clearGlErrors();
         int i = GlStateManager._genTexture();
         if (label == null) {
@@ -169,14 +_,14 @@
             for (int k : GlConst.CUBEMAP_TARGETS) {
                 for (int l = 0; l < mipLevels; l++) {
                     GlStateManager._texImage2D(
-                        k, l, GlConst.toGlInternalId(format), width >> l, height >> l, 0, GlConst.toGlExternalId(format), GlConst.toGlType(format), null
+                        k, l, GlConst.toGlInternalId(format, stencil), width >> l, height >> l, 0, GlConst.toGlExternalId(format, stencil), GlConst.toGlType(format, stencil), null
                     );
                 }
             }
         } else {
             for (int i1 = 0; i1 < mipLevels; i1++) {
                 GlStateManager._texImage2D(
-                    j, i1, GlConst.toGlInternalId(format), width >> i1, height >> i1, 0, GlConst.toGlExternalId(format), GlConst.toGlType(format), null
+                    j, i1, GlConst.toGlInternalId(format, stencil), width >> i1, height >> i1, 0, GlConst.toGlExternalId(format, stencil), GlConst.toGlType(format, stencil), null
                 );
             }
         }
@@ -187,7 +_,7 @@
         } else if (j1 != 0) {
             throw new IllegalStateException("OpenGL error " + j1);
         } else {
-            GlTexture gltexture = new GlTexture(usage, label, format, width, height, depthOrLayers, mipLevels, i);
+            GlTexture gltexture = new GlTexture(usage, label, format, width, height, depthOrLayers, mipLevels, i, stencil);
             this.debugLabels.applyLabel(gltexture);
             return gltexture;
         }
