--- a/com/mojang/blaze3d/opengl/GlStateManager.java
+++ b/com/mojang/blaze3d/opengl/GlStateManager.java
@@ -86,6 +_,11 @@
         }
     }
 
+    public static boolean _isBlendEnabled() {
+        RenderSystem.assertOnRenderThread();
+        return BLEND.mode.enabled;
+    }
+
     public static void _disableBlend() {
         RenderSystem.assertOnRenderThread();
         BLEND.mode.disable();
@@ -380,9 +_,17 @@
         }
     }
 
+    /* Stores the last values sent into glMultiTexCoord2f */
+    public static float lastBrightnessX = 0.0f;
+    public static float lastBrightnessY = 0.0f;
+
     public static void _texParameter(final int target, final int name, final int value) {
         RenderSystem.assertOnRenderThread();
         GL11.glTexParameteri(target, name, value);
+        if (target == GL13.GL_TEXTURE1) {
+            lastBrightnessX = name;
+            lastBrightnessY = value;
+        }
     }
 
     public static int _getTexLevelParameter(final int target, final int level, final int name) {
