--- a/com/mojang/blaze3d/opengl/GlTexture.java
+++ b/com/mojang/blaze3d/opengl/GlTexture.java
@@ -28,8 +_,13 @@
         final int mipLevels,
         final int id
     ) {
+        this(usage, label, format, width, height, depthOrLayers, mipLevels, id, false);
+    }
+
+    protected GlTexture(int usage, String label, TextureFormat format, int width, int height, int depthOrLayers, int mipLevels, int id, boolean stencil) {
         super(usage, label, format, width, height, depthOrLayers, mipLevels);
         this.id = id;
+        this.stencilEnabled = stencil;
     }
 
     @Override
@@ -96,5 +_,12 @@
         if (this.closed && this.views == 0) {
             this.destroyImmediately();
         }
+    }
+
+    private final boolean stencilEnabled;
+
+    @Override
+    public boolean isStencilEnabled() {
+        return this.stencilEnabled;
     }
 }
