--- a/net/minecraft/resources/RegistryOps.java
+++ b/net/minecraft/resources/RegistryOps.java
@@ -71,6 +_,22 @@
             .forGetter(var0 -> null);
     }
 
+    public static <E> com.mojang.serialization.MapCodec<HolderLookup.RegistryLookup<E>> retrieveRegistryLookup(final ResourceKey<? extends Registry<? extends E>> registryKey) {
+       return ExtraCodecs.retrieveContext(ops -> {
+           if (!(ops instanceof RegistryOps<?> registryOps))
+              return DataResult.error(() -> "Not a registry ops");
+
+           var lookup = registryOps.lookupProvider.lookup(registryKey);
+           if (lookup.isEmpty())
+               return DataResult.error(() -> "Unknown Registry: " + registryKey);
+
+           if (!(lookup.get() instanceof HolderLookup.RegistryLookup<E> registryLookup))
+               return DataResult.<HolderLookup.RegistryLookup<E>>error(() -> "Found holder getter but was not a registry lookup for " + registryKey);
+
+           return DataResult.success(registryLookup, registryLookup.registryLifecycle());
+       });
+    }
+
     public static <E, O> RecordCodecBuilder<O, Holder.Reference<E>> retrieveElement(final ResourceKey<E> key) {
         ResourceKey<? extends Registry<E>> registryKey = ResourceKey.createRegistryKey(key.registry());
         return ExtraCodecs.retrieveContext(
