--- a/net/minecraft/world/level/chunk/ChunkAccess.java
+++ b/net/minecraft/world/level/chunk/ChunkAccess.java
@@ -316,10 +_,14 @@
 
     @Override
     public final void findBlockLightSources(final BiConsumer<BlockPos, BlockState> consumer) {
-        this.findBlocks(state -> state.getLightEmission() != 0, consumer);
+        this.findBlocks((state, pos) -> state.getLightEmission(this, pos) != 0, consumer);
     }
 
     public void findBlocks(final Predicate<BlockState> predicate, final BiConsumer<BlockPos, BlockState> consumer) {
+        findBlocks((state, pos) -> predicate.test(state), consumer);
+    }
+
+    public void findBlocks(final java.util.function.BiPredicate<BlockState, BlockPos> predicate, final BiConsumer<BlockPos, BlockState> consumer) {
         BlockStateConsumer blockStateConsumer = (pos, state) -> {
             consumer.accept(pos, state);
             return Continuation.CONTINUE;
@@ -328,7 +_,7 @@
 
         for (int sectionY = this.getMinSectionY(); sectionY <= this.getMaxSectionY(); sectionY++) {
             LevelChunkSection section = this.getSection(this.getSectionIndexFromSectionY(sectionY));
-            if (section.maybeHas(predicate)) {
+            if (section.maybeHas(state -> predicate.test(state, BlockPos.ZERO))) {
                 BlockPos origin = SectionPos.of(this.chunkPos, sectionY).origin();
                 BlockScanUtils.findBlocksInSection(section, origin, 0, 0, 0, 15, 15, 15, predicate, blockStateConsumer);
             }
@@ -479,5 +_,9 @@
     }
 
     public record PackedTicks(List<SavedTick<Block>> blocks, List<SavedTick<Fluid>> fluids) {
+    }
+
+    public net.minecraft.world.level.@Nullable LevelAccessor getWorldForge() {
+        return null;
     }
 }
