--- a/net/minecraft/world/level/chunk/ChunkAccess.java
+++ b/net/minecraft/world/level/chunk/ChunkAccess.java
@@ -318,23 +_,28 @@
 
     @Override
     public final void findBlockLightSources(final BiConsumer<BlockPos, BlockState> consumer) {
-        this.findBlocks(state -> state.getLightEmission() != 0, consumer);
+        this.findBlocks((state, pos) -> state.getLightEmission(this, pos) != 0, consumer);
     }
 
     public void findBlocks(final Predicate<BlockState> predicate, final BiConsumer<BlockPos, BlockState> consumer) {
+        findBlocks((state, pos) -> predicate.test(state), consumer);
+    }
+
+    public void findBlocks(final java.util.function.BiPredicate<BlockState, BlockPos> predicate, final BiConsumer<BlockPos, BlockState> consumer) {
         BlockPos.MutableBlockPos blockpos$mutableblockpos = new BlockPos.MutableBlockPos();
 
         for (int i = this.getMinSectionY(); i <= this.getMaxSectionY(); i++) {
             LevelChunkSection levelchunksection = this.getSection(this.getSectionIndexFromSectionY(i));
-            if (levelchunksection.maybeHas(predicate)) {
+            if (levelchunksection.maybeHas(state -> predicate.test(state, BlockPos.ZERO))) {
                 BlockPos blockpos = SectionPos.of(this.chunkPos, i).origin();
 
                 for (int j = 0; j < 16; j++) {
                     for (int k = 0; k < 16; k++) {
                         for (int l = 0; l < 16; l++) {
                             BlockState blockstate = levelchunksection.getBlockState(l, j, k);
-                            if (predicate.test(blockstate)) {
-                                consumer.accept(blockpos$mutableblockpos.setWithOffset(blockpos, l, j, k), blockstate);
+                            blockpos$mutableblockpos.setWithOffset(blockpos, l, j, k);
+                            if (predicate.test(blockstate, blockpos$mutableblockpos.immutable())) {
+                                consumer.accept(blockpos$mutableblockpos, blockstate);
                             }
                         }
                     }
@@ -493,5 +_,9 @@
     }
 
     public record PackedTicks(List<SavedTick<Block>> blocks, List<SavedTick<Fluid>> fluids) {
+    }
+
+    public net.minecraft.world.level.@Nullable LevelAccessor getWorldForge() {
+        return null;
     }
 }
