--- a/net/minecraft/world/level/chunk/PalettedContainer.java
+++ b/net/minecraft/world/level/chunk/PalettedContainer.java
@@ -49,13 +_,13 @@
     private static <T, C extends PalettedContainerRO<T>> Codec<C> codec(
         final Codec<T> elementCodec, final Strategy<T> strategy, final T defaultValue, final PalettedContainerRO.Unpacker<T, C> unpacker
     ) {
-        return RecordCodecBuilder.<PalettedContainerRO.PackedData>create(
+        return RecordCodecBuilder.<PalettedContainerRO.PackedData<T>>create(
                 i -> i.group(
                         elementCodec.mapResult(ExtraCodecs.orElsePartial(defaultValue))
                             .listOf()
                             .fieldOf("palette")
-                            .forGetter(PalettedContainerRO.PackedData::paletteEntries),
-                        Codec.LONG_STREAM.lenientOptionalFieldOf("data").forGetter(PalettedContainerRO.PackedData::storage)
+                            .<PalettedContainerRO.PackedData<T>>forGetter(PalettedContainerRO.PackedData::paletteEntries),
+                        Codec.LONG_STREAM.lenientOptionalFieldOf("data").<PalettedContainerRO.PackedData<T>>forGetter(PalettedContainerRO.PackedData::storage)
                     )
                     .apply(i, PalettedContainerRO.PackedData::new)
             )
