--- a/net/minecraft/client/gui/components/Button.java
+++ b/net/minecraft/client/gui/components/Button.java
@@ -33,6 +_,12 @@
         this.createNarration = p_259552_;
     }
 
+    // Forge (GH-9238)
+    protected Button(Builder builder) {
+        this(builder.x, builder.y, builder.width, builder.height, builder.message, builder.onPress, builder.createNarration);
+        this.setTooltip(builder.tooltip);
+    }
+
     @Override
     public void onPress(InputWithModifiers p_424878_) {
         this.onPress.onPress(this);
@@ -96,9 +_,12 @@
         }
 
         public Button build() {
-            Button button = new Button.Plain(this.x, this.y, this.width, this.height, this.message, this.onPress, this.createNarration);
-            button.setTooltip(this.tooltip);
-            return button;
+            return build(Button.Plain::new);
+        }
+
+        // Forge (GH-9238)
+        public Button build(java.util.function.Function<Builder, Button> builder) {
+            return builder.apply(this);
         }
     }
 
@@ -118,6 +_,10 @@
             int p_455297_, int p_455388_, int p_459653_, int p_452847_, Component p_459347_, Button.OnPress p_453378_, Button.CreateNarration p_456163_
         ) {
             super(p_455297_, p_455388_, p_459653_, p_452847_, p_459347_, p_453378_, p_456163_);
+        }
+
+        protected Plain(Builder builder) {
+            super(builder);
         }
 
         @Override
