--- a/net/minecraft/data/worldgen/BootstrapContextAccess.java
+++ b/net/minecraft/data/worldgen/BootstrapContextAccess.java
@@ -9,6 +_,8 @@
 public interface BootstrapContextAccess {
     <S> HolderGetter<S> lookup(ResourceKey<? extends Registry<? extends S>> key);
 
+    default <S> java.util.Optional<net.minecraft.core.HolderLookup.RegistryLookup<S>> registryLookup(ResourceKey<? extends Registry<? extends S>> registry) { return java.util.Optional.empty(); }
+
     @Deprecated
     <S> Stream<Holder.Reference<S>> listContextElements(ResourceKey<? extends Registry<? extends S>> key);
 }
