--- a/net/minecraft/world/entity/EntityEquipment.java
+++ b/net/minecraft/world/entity/EntityEquipment.java
@@ -49,7 +_,7 @@
         for (Entry<EquipmentSlot, ItemStack> entry : this.items.entrySet()) {
             ItemStack item = entry.getValue();
             if (!item.isEmpty()) {
-                item.inventoryTick(owner.level(), owner, entry.getKey());
+                item.inventoryTick(owner.level(), owner, entry.getKey(), -1);
             }
         }
     }
@@ -69,5 +_,9 @@
 
     public void clear() {
         this.items.replaceAll((s, v) -> ItemStack.EMPTY);
+    }
+
+    public int size() {
+        return this.items.size();
     }
 }
