--- a/net/minecraft/world/entity/SpawnPlacementTypes.java
+++ b/net/minecraft/world/entity/SpawnPlacementTypes.java
@@ -28,7 +_,7 @@
                 BlockPos above = blockPos.above();
                 BlockPos below = blockPos.below();
                 BlockState belowState = level.getBlockState(below);
-                return !belowState.isValidSpawn(level, below, type)
+                return !belowState.isValidSpawn(level, below, this, type)
                     ? false
                     : this.isValidEmptySpawnBlock(level, blockPos, type) && this.isValidEmptySpawnBlock(level, above, type);
             } else {
