--- a/net/minecraft/client/gui/components/debug/DebugEntryLookingAtEntity.java
+++ b/net/minecraft/client/gui/components/debug/DebugEntryLookingAtEntity.java
@@ -29,6 +_,7 @@
         if (entity != null) {
             list.add(ChatFormatting.UNDERLINE + "Targeted Entity");
             list.add(entity.typeHolder().getRegisteredName());
+            entity.getType().builtInRegistryHolder().tags().forEach(t -> list.add("#" + t.location()));
         }
 
         displayer.addToGroup(GROUP, list);
