--- a/net/minecraft/data/Main.java
+++ b/net/minecraft/data/Main.java
@@ -78,14 +_,17 @@
         OptionSpec<Void> allOption = parser.accepts("all", "Include all generators");
         OptionSpec<String> outputOption = parser.accepts("output", "Output folder").withRequiredArg().defaultsTo("generated");
         OptionSpec<String> inputOption = parser.accepts("input", "Input folder").withRequiredArg();
+        var loader = net.minecraftforge.data.loading.DatagenModLoader.setup(parser, false);
         OptionSet optionSet = parser.parse(args);
-        if (!optionSet.has(helpOption) && optionSet.hasOptions()) {
+        if (!optionSet.has(helpOption) && optionSet.hasOptions() && loader.hasArgs(optionSet)) {
             Path output = Paths.get(outputOption.value(optionSet));
             boolean allOptions = optionSet.has(allOption);
             boolean server = allOptions || optionSet.has(serverOption);
             boolean dev = allOptions || optionSet.has(devOption);
             boolean reports = allOptions || optionSet.has(reportsOption);
             Collection<Path> input = optionSet.valuesOf(inputOption).stream().map(x$0 -> Paths.get(x$0)).toList();
+            if (!loader.run(optionSet, output, input, server, allOptions, dev, reports))
+                return;
             DataGenerator generator = new DataGenerator.Cached(output, SharedConstants.getCurrentVersion(), true);
             addServerDefinitionProviders(generator, server, reports);
             addServerConverters(generator, input, server, dev);
