--- a/net/minecraft/client/resources/model/cuboid/UnbakedCuboidGeometry.java
+++ b/net/minecraft/client/resources/model/cuboid/UnbakedCuboidGeometry.java
@@ -29,8 +_,20 @@
         final ModelState modelState,
         final ModelDebugName name
     ) {
-        QuadCollection.Builder quadcollection$builder = new QuadCollection.Builder();
+        var builder = new QuadCollection.Builder();
+        bake(builder, elements, textures, modelBaker, modelState, name, id -> modelBaker.materials().resolveSlot(textures, id, name));
+        return builder.build();
+    }
 
+    public static void bake(
+        final QuadCollection.Builder builder,
+        final List<CuboidModelElement> elements,
+        final TextureSlots textures,
+        final ModelBaker modelBaker,
+        final ModelState modelState,
+        final ModelDebugName name,
+        final java.util.function.Function<String, Material.Baked> materialMapper //Forge: Allow overwriting textures
+    ) {
         for (CuboidModelElement cuboidmodelelement : elements) {
             boolean flag = true;
             boolean flag1 = true;
@@ -63,7 +_,7 @@
                         case Z -> flag2;
                     };
                     if (flag3) {
-                        Material.Baked material$baked = modelBaker.materials().resolveSlot(textures, cuboidface.texture(), name);
+                        Material.Baked material$baked = materialMapper.apply(cuboidface.texture());
                         BakedQuad bakedquad = FaceBakery.bakeQuad(
                             modelBaker,
                             vector3fc,
@@ -77,9 +_,9 @@
                             cuboidmodelelement.lightEmission()
                         );
                         if (cuboidface.cullForDirection() == null) {
-                            quadcollection$builder.addUnculledFace(bakedquad);
+                            builder.addUnculledFace(bakedquad);
                         } else {
-                            quadcollection$builder.addCulledFace(
+                            builder.addCulledFace(
                                 Direction.rotate(modelState.transformation().getMatrix(), cuboidface.cullForDirection()), bakedquad
                             );
                         }
@@ -87,7 +_,5 @@
                 }
             }
         }
-
-        return quadcollection$builder.build();
     }
 }
