--- a/net/minecraft/world/level/levelgen/structure/templatesystem/StructureTemplate.java
+++ b/net/minecraft/world/level/levelgen/structure/templatesystem/StructureTemplate.java
@@ -256,6 +_,10 @@
         return transform(pos, settings.getMirror(), settings.getRotation(), settings.getRotationPivot());
     }
 
+    public static Vec3 transformedVec3d(StructurePlaceSettings placementIn, Vec3 pos) {
+        return transform(pos, placementIn.getMirror(), placementIn.getRotation(), placementIn.getRotationPivot());
+    }
+
     public boolean placeInWorld(
         final ServerLevelAccessor level,
         final BlockPos position,
@@ -282,7 +_,7 @@
                 int l = Integer.MIN_VALUE;
                 int i1 = Integer.MIN_VALUE;
                 int j1 = Integer.MIN_VALUE;
-                List<StructureTemplate.StructureBlockInfo> list4 = processBlockInfos(level, position, referencePos, settings, list);
+                List<StructureTemplate.StructureBlockInfo> list4 = processBlockInfos(level, position, referencePos, settings, list, this);
 
                 try (ProblemReporter.ScopedCollector problemreporter$scopedcollector = new ProblemReporter.ScopedCollector(LOGGER)) {
                     for (StructureTemplate.StructureBlockInfo structuretemplate$structureblockinfo : list4) {
@@ -406,10 +_,10 @@
                             position,
                             settings.getMirror(),
                             settings.getRotation(),
-                            settings.getRotationPivot(),
                             boundingbox,
                             settings.shouldFinalizeEntities(),
-                            problemreporter$scopedcollector
+                            problemreporter$scopedcollector,
+                            settings
                         );
                     }
                 }
@@ -458,12 +_,21 @@
         );
     }
 
+    /**
+     * @deprecated Forge: Use {@link #processBlockInfos(ServerLevelAccessor, BlockPos, BlockPos, StructurePlaceSettings, List, StructureTemplate)} instead.
+     */
+    @Deprecated
+    public static List<StructureTemplate.StructureBlockInfo> processBlockInfos(final ServerLevelAccessor level, final BlockPos position, final BlockPos referencePos, final StructurePlaceSettings settings, final List<StructureTemplate.StructureBlockInfo> blockInfoList) {
+        return processBlockInfos(level, position, referencePos, settings, blockInfoList, null);
+    }
+
     public static List<StructureTemplate.StructureBlockInfo> processBlockInfos(
         final ServerLevelAccessor level,
         final BlockPos position,
         final BlockPos referencePos,
         final StructurePlaceSettings settings,
-        final List<StructureTemplate.StructureBlockInfo> blockInfoList
+        List<StructureTemplate.StructureBlockInfo> blockInfoList,
+        @Nullable StructureTemplate template
     ) {
         List<StructureTemplate.StructureBlockInfo> list = new ArrayList<>();
         List<StructureTemplate.StructureBlockInfo> list1 = new ArrayList<>();
@@ -479,7 +_,7 @@
 
             while (structuretemplate$structureblockinfo1 != null && iterator.hasNext()) {
                 structuretemplate$structureblockinfo1 = iterator.next()
-                    .processBlock(level, position, referencePos, structuretemplate$structureblockinfo, structuretemplate$structureblockinfo1, settings);
+                    .processBlock(level, position, referencePos, structuretemplate$structureblockinfo, structuretemplate$structureblockinfo1, settings, template);
             }
 
             if (structuretemplate$structureblockinfo1 != null) {
@@ -500,17 +_,17 @@
         final BlockPos position,
         final Mirror mirror,
         final Rotation rotation,
-        final BlockPos pivot,
         final @Nullable BoundingBox boundingBox,
         final boolean finalizeEntities,
-        final ProblemReporter problemReporter
+        final ProblemReporter problemReporter,
+        final StructurePlaceSettings placementIn
     ) {
-        for (StructureTemplate.StructureEntityInfo structuretemplate$structureentityinfo : this.entityInfoList) {
-            BlockPos blockpos = transform(structuretemplate$structureentityinfo.blockPos, mirror, rotation, pivot).offset(position);
+        var entities = processEntityInfos(this, level, position, placementIn, this.entityInfoList);
+        for (StructureTemplate.StructureEntityInfo structuretemplate$structureentityinfo : entities) {
+            BlockPos blockpos = structuretemplate$structureentityinfo.blockPos; // FORGE: Position will have already been transformed by processEntityInfos
             if (boundingBox == null || boundingBox.isInside(blockpos)) {
                 CompoundTag compoundtag = structuretemplate$structureentityinfo.nbt.copy();
-                Vec3 vec3 = transform(structuretemplate$structureentityinfo.pos, mirror, rotation, pivot);
-                Vec3 vec31 = vec3.add(position.getX(), position.getY(), position.getZ());
+                Vec3 vec31 = structuretemplate$structureentityinfo.pos; // FORGE: Position will have already been transformed by processEntityInfos
                 ListTag listtag = new ListTag();
                 listtag.add(DoubleTag.valueOf(vec31.x));
                 listtag.add(DoubleTag.valueOf(vec31.y));
@@ -532,6 +_,30 @@
             }
         }
     }
+
+    public static List<StructureTemplate.StructureEntityInfo> processEntityInfos(
+        @Nullable StructureTemplate template,
+        final LevelAccessor level,
+        final BlockPos targetPos,
+        final StructurePlaceSettings settings,
+        final List<StructureTemplate.StructureEntityInfo> entities
+    ) {
+        var list = new ArrayList<StructureTemplate.StructureEntityInfo>();
+        for (var entityInfo : entities) {
+           var pos = transformedVec3d(settings, entityInfo.pos).add(Vec3.atLowerCornerOf(targetPos));
+           var blockpos = calculateRelativePosition(settings, entityInfo.blockPos).offset(targetPos);
+           var info = new StructureTemplate.StructureEntityInfo(pos, blockpos, entityInfo.nbt);
+           for (var proc : settings.getProcessors()) {
+              info = proc.processEntity(level, targetPos, entityInfo, info, settings, template);
+              if (info == null)
+                 break;
+           }
+           if (info != null)
+              list.add(info);
+        }
+        return list;
+     }
+
 
     private static Optional<Entity> createEntityIgnoreException(final ProblemReporter reporter, final ServerLevelAccessor level, final CompoundTag tag) {
         try {
