--- a/net/minecraft/world/level/levelgen/structure/templatesystem/StructureProcessor.java
+++ b/net/minecraft/world/level/levelgen/structure/templatesystem/StructureProcessor.java
@@ -7,6 +_,7 @@
 import org.jspecify.annotations.Nullable;
 
 public abstract class StructureProcessor {
+    /** @deprecated Use variant with StructureTemplate argument */
     public StructureTemplate.@Nullable StructureBlockInfo processBlock(
         final LevelReader level,
         final BlockPos targetPosition,
@@ -18,6 +_,18 @@
         return processedBlockInfo;
     }
 
+    public StructureTemplate.@Nullable StructureBlockInfo processBlock(
+        final LevelReader level,
+        final BlockPos targetPosition,
+        final BlockPos referencePos,
+        final StructureTemplate.StructureBlockInfo originalBlockInfo,
+        final StructureTemplate.StructureBlockInfo processedBlockInfo,
+        final StructurePlaceSettings settings,
+        final @Nullable StructureTemplate template
+    ) {
+       return processBlock(level, targetPosition, referencePos, originalBlockInfo, processedBlockInfo, settings);
+    }
+
     protected abstract StructureProcessorType<?> getType();
 
     public List<StructureTemplate.StructureBlockInfo> finalizeProcessing(
@@ -29,5 +_,20 @@
         final StructurePlaceSettings settings
     ) {
         return processedBlockInfoList;
+    }
+
+    /**
+     * Use this method to process entities from a structure in much the same way as
+     * blocks, parameters are analogous.
+     */
+    public StructureTemplate.@Nullable StructureEntityInfo processEntity(
+        final LevelReader level,
+        final BlockPos targetPosition,
+        final StructureTemplate.StructureEntityInfo originalEntityInfo,
+        final StructureTemplate.StructureEntityInfo processedEntityInfo,
+        final StructurePlaceSettings settings,
+        final @Nullable StructureTemplate template
+    ) {
+       return processedEntityInfo;
     }
 }
