--- a/net/minecraft/world/level/levelgen/structure/templatesystem/StructureTemplateManager.java
+++ b/net/minecraft/world/level/levelgen/structure/templatesystem/StructureTemplateManager.java
@@ -77,6 +_,7 @@
         }
 
         builder.add(this.resourceManagerSource);
+        builder.add(net.minecraftforge.common.ForgeHooks.emptyStructureSource());
         this.sources = builder.build();
     }
 
